--- Makefile.in.orig	2006-04-11 11:13:59.000000000 -0700
+++ Makefile.in	2006-04-11 11:25:58.000000000 -0700
@@ -31,7 +31,7 @@
 ccache:
 	test -z "$(ENABLE_CCACHE)" || (cd $(CCACHE) && $(MAKE))
 
-libfiles: $(srcdir)/Lib/swigwarn.swg
+libfiles: Lib/swigwarn.swg
 
 # Files required just for the tarball
 maintainer: libfiles
@@ -415,7 +415,7 @@
 # Note: Generated into the source tree rather than build tree
 #####################################################################
 
-$(srcdir)/Lib/swigwarn.swg: $(srcdir)/Source/Include/swigwarn.h
+Lib/swigwarn.swg: $(srcdir)/Source/Include/swigwarn.h
 	mkdir -p Lib
 	echo "/* Automatically generated file containing all the swig warning codes.  */" > $@
 	echo "/* Do not modify this file by hand, change 'Source/Include/swigwarn.h'  */" >> $@
