--- tmac/doc-common.orig	2005-03-17 23:53:17.000000000 -0800
+++ tmac/doc-common	2012-10-13 22:10:42.000000000 -0700
@@ -418,8 +418,12 @@
 .ds doc-operating-system-NetBSD-1.2c  1.2C
 .ds doc-operating-system-NetBSD-1.2d  1.2D
 .ds doc-operating-system-NetBSD-1.2e  1.2E
+.ds doc-operating-system-NetBSD-1.2.1 1.2.1
 .ds doc-operating-system-NetBSD-1.3   1.3
 .ds doc-operating-system-NetBSD-1.3a  1.3A
+.ds doc-operating-system-NetBSD-1.3.1 1.3.1
+.ds doc-operating-system-NetBSD-1.3.2 1.3.2
+.ds doc-operating-system-NetBSD-1.3.3 1.3.3
 .ds doc-operating-system-NetBSD-1.4   1.4
 .ds doc-operating-system-NetBSD-1.4.1 1.4.1
 .ds doc-operating-system-NetBSD-1.4.2 1.4.2
@@ -431,10 +435,27 @@
 .ds doc-operating-system-NetBSD-1.6   1.6
 .ds doc-operating-system-NetBSD-1.6.1 1.6.1
 .ds doc-operating-system-NetBSD-1.6.2 1.6.2
+.ds doc-operating-system-NetBSD-1.6.3 1.6.3
 .ds doc-operating-system-NetBSD-2.0   2.0
 .ds doc-operating-system-NetBSD-2.0.1 2.0.1
 .ds doc-operating-system-NetBSD-2.0.2 2.0.2
+.ds doc-operating-system-NetBSD-2.0.3 2.0.3
 .ds doc-operating-system-NetBSD-2.1   2.1
+.ds doc-operating-system-NetBSD-3.0   3.0
+.ds doc-operating-system-NetBSD-3.0.1 3.0.1
+.ds doc-operating-system-NetBSD-3.0.2 3.0.2
+.ds doc-operating-system-NetBSD-3.0.3 3.0.3
+.ds doc-operating-system-NetBSD-3.1   3.1
+.ds doc-operating-system-NetBSD-3.1.1 3.1.1
+.ds doc-operating-system-NetBSD-4.0   4.0
+.ds doc-operating-system-NetBSD-4.0.1 4.0.1
+.ds doc-operating-system-NetBSD-5.0   5.0
+.ds doc-operating-system-NetBSD-5.0.1 5.0.1
+.ds doc-operating-system-NetBSD-5.0.2 5.0.2
+.ds doc-operating-system-NetBSD-5.1   5.1
+.ds doc-operating-system-NetBSD-5.1.1 5.1.1
+.ds doc-operating-system-NetBSD-5.1.2 5.1.2
+.ds doc-operating-system-NetBSD-6.0   6.0
 .
 .ds doc-operating-system-FreeBSD-1.0     1.0
 .ds doc-operating-system-FreeBSD-1.1     1.1
@@ -472,11 +493,100 @@
 .ds doc-operating-system-FreeBSD-4.8     4.8
 .ds doc-operating-system-FreeBSD-4.9     4.9
 .ds doc-operating-system-FreeBSD-4.10    4.10
+.ds doc-operating-system-FreeBSD-4.11    4.11
 .ds doc-operating-system-FreeBSD-5.0     5.0
 .ds doc-operating-system-FreeBSD-5.1     5.1
 .ds doc-operating-system-FreeBSD-5.2     5.2
 .ds doc-operating-system-FreeBSD-5.2.1   5.2.1
 .ds doc-operating-system-FreeBSD-5.3     5.3
+.ds doc-operating-system-FreeBSD-5.4     5.4
+.ds doc-operating-system-FreeBSD-5.5     5.5
+.ds doc-operating-system-FreeBSD-6.0     6.0
+.ds doc-operating-system-FreeBSD-6.1     6.1
+.ds doc-operating-system-FreeBSD-6.2     6.2
+.ds doc-operating-system-FreeBSD-6.3     6.3
+.ds doc-operating-system-FreeBSD-6.4     6.4
+.ds doc-operating-system-FreeBSD-7.0     7.0
+.ds doc-operating-system-FreeBSD-7.1     7.1
+.ds doc-operating-system-FreeBSD-7.2     7.2
+.ds doc-operating-system-FreeBSD-7.3     7.3
+.ds doc-operating-system-FreeBSD-7.4     7.4
+.ds doc-operating-system-FreeBSD-8.0     8.0
+.ds doc-operating-system-FreeBSD-8.1     8.1
+.ds doc-operating-system-FreeBSD-8.2     8.2
+.ds doc-operating-system-FreeBSD-8.3     8.3
+.ds doc-operating-system-FreeBSD-9.0     9.0
+.ds doc-operating-system-FreeBSD-9.1     9.1
+.
+.ds doc-operating-system-Darwin-8.0.0 8.0.0
+.ds doc-operating-system-Darwin-8.1.0 8.1.0
+.ds doc-operating-system-Darwin-8.2.0 8.2.0
+.ds doc-operating-system-Darwin-8.3.0 8.3.0
+.ds doc-operating-system-Darwin-8.4.0 8.4.0
+.ds doc-operating-system-Darwin-8.5.0 8.5.0
+.ds doc-operating-system-Darwin-8.6.0 8.6.0
+.ds doc-operating-system-Darwin-8.7.0 8.7.0
+.ds doc-operating-system-Darwin-8.8.0 8.8.0
+.ds doc-operating-system-Darwin-8.9.0 8.9.0
+.ds doc-operating-system-Darwin-8.10.0 8.10.0
+.ds doc-operating-system-Darwin-8.11.0 8.11.0
+.ds doc-operating-system-Darwin-9.0.0 9.0.0
+.ds doc-operating-system-Darwin-9.1.0 9.1.0
+.ds doc-operating-system-Darwin-9.2.0 9.2.0
+.ds doc-operating-system-Darwin-9.3.0 9.3.0
+.ds doc-operating-system-Darwin-9.4.0 9.4.0
+.ds doc-operating-system-Darwin-9.5.0 9.5.0
+.ds doc-operating-system-Darwin-9.6.0 9.6.0
+.ds doc-operating-system-Darwin-9.7.0 9.7.0
+.ds doc-operating-system-Darwin-9.8.0 9.8.0
+.ds doc-operating-system-Darwin-10.0.0 10.0.0
+.ds doc-operating-system-Darwin-10.1.0 10.1.0
+.ds doc-operating-system-Darwin-10.2.0 10.2.0
+.ds doc-operating-system-Darwin-10.3.0 10.3.0
+.ds doc-operating-system-Darwin-10.4.0 10.4.0
+.ds doc-operating-system-Darwin-10.5.0 10.5.0
+.ds doc-operating-system-Darwin-10.6.0 10.6.0
+.ds doc-operating-system-Darwin-10.7.0 10.7.0
+.ds doc-operating-system-Darwin-10.8.0 10.8.0
+.ds doc-operating-system-Darwin-11.0.0 11.0.0
+.ds doc-operating-system-Darwin-11.1.0 11.1.0
+.ds doc-operating-system-Darwin-11.2.0 11.2.0
+.ds doc-operating-system-Darwin-11.3.0 11.3.0
+.ds doc-operating-system-Darwin-11.4.0 11.4.0
+.ds doc-operating-system-Darwin-11.5.0 11.5.0
+.ds doc-operating-system-Darwin-12.0.0 12.0.0
+.ds doc-operating-system-Darwin-12.1.0 12.1.0
+.ds doc-operating-system-Darwin-12.2.0 12.2.0
+.ds doc-operating-system-Darwin-13.0.0 13.0.0
+.
+.ds doc-operating-system-DragonFly-1.0 1.0
+.ds doc-operating-system-DragonFly-1.1 1.1
+.ds doc-operating-system-DragonFly-1.2 1.2
+.ds doc-operating-system-DragonFly-1.3 1.3
+.ds doc-operating-system-DragonFly-1.4 1.4
+.ds doc-operating-system-DragonFly-1.4.4 1.4.4
+.ds doc-operating-system-DragonFly-1.5 1.5
+.ds doc-operating-system-DragonFly-1.6 1.6
+.ds doc-operating-system-DragonFly-1.7 1.7
+.ds doc-operating-system-DragonFly-1.8 1.8
+.ds doc-operating-system-DragonFly-1.8.1 1.8.1
+.ds doc-operating-system-DragonFly-1.9 1.9
+.ds doc-operating-system-DragonFly-1.10 1.10
+.ds doc-operating-system-DragonFly-1.10.1 1.10.1
+.ds doc-operating-system-DragonFly-1.11 1.11
+.ds doc-operating-system-DragonFly-1.12 1.12
+.ds doc-operating-system-DragonFly-1.12.1 1.12.1
+.ds doc-operating-system-DragonFly-1.12.2 1.12.2
+.ds doc-operating-system-DragonFly-2.0 2.0
+.ds doc-operating-system-DragonFly-2.0.1 2.0.1
+.ds doc-operating-system-DragonFly-2.2 2.2
+.ds doc-operating-system-DragonFly-2.2.1 2.2.1
+.ds doc-operating-system-DragonFly-2.4 2.4
+.ds doc-operating-system-DragonFly-2.4.1 2.4.1
+.ds doc-operating-system-DragonFly-3.0 3.0
+.ds doc-operating-system-DragonFly-3.0.1 3.0.1
+.ds doc-operating-system-DragonFly-3.0.2 3.0.2
+.ds doc-operating-system-DragonFly-3.0.3 3.0.3
 .
 .de Os
 .  ie "\$1"" \
@@ -512,11 +614,19 @@
 .      el \
 .        tm mdoc warning: .Os: Unknown NetBSD version `\$2' (#\n[.c])
 .  \}\}
+.  el \{ .ie "\$1"Darwin" \{\
+.    ds doc-operating-system Darwin
+.    if \A\$2 \{\
+.      ie d doc-operating-system-Darwin-\$2 \
+.        as doc-operating-system \~\*[doc-operating-system-Darwin-\$2]
+.      el \
+.        tm mdoc warning: .Os: Unknown Darwin version `\$2' (#\n[.c])
+.  \}\}
 .  el \{\
 .    ds doc-operating-system \$1
 .    if !"\$2"" \
 .      as doc-operating-system " \$2
-.  \}\}\}\}\}
+.  \}\}\}\}\}\}
 ..
 .
 .
