diff --git a/httpd/modules/ssl/ssl_engine_init.c b/httpd/modules/ssl/ssl_engine_init.c
index c2ec048..9480fd6 100644
--- modules/ssl/ssl_engine_init.c
+++ modules/ssl/ssl_engine_init.c
@@ -1716,11 +1716,12 @@ static apr_status_t ssl_init_proxy_certs(server_rec *s,
 
     ap_assert(store != NULL); /* safe to assume always non-NULL? */
 
+#warning "No SSL_CTX_set_post_handshake_auth"
 #if OPENSSL_VERSION_NUMBER >= 0x1010100fL && !defined(LIBRESSL_VERSION_NUMBER)
     /* For OpenSSL >=1.1.1, turn on client cert support which is
      * otherwise turned off by default (by design).
      * https://github.com/openssl/openssl/issues/6933 */
-    SSL_CTX_set_post_handshake_auth(mctx->ssl_ctx, 1);
+//    SSL_CTX_set_post_handshake_auth(mctx->ssl_ctx, 1);
 #endif
     
     SSL_CTX_set_client_cert_cb(mctx->ssl_ctx,
