--- docs/conf/httpd.conf.in.orig	2007-02-28 16:05:22.000000000 -0800
+++ docs/conf/httpd.conf.in	2007-02-28 16:11:33.000000000 -0800
@@ -171,13 +171,27 @@
 # The following lines prevent .htaccess and .htpasswd files from being 
 # viewed by Web clients. 
 #
-<FilesMatch "^\.ht">
+<FilesMatch "^\.([Hh][Tt]|[Dd][Ss]_[Ss])">
     Order allow,deny
     Deny from all
     Satisfy All
 </FilesMatch>
 
 #
+# Apple specific filesystem protection.
+#
+<Files "rsrc">
+    Order allow,deny
+    Deny from all
+    Satisfy All
+</Files>
+<DirectoryMatch ".*\.\.namedfork">
+    Order allow,deny
+    Deny from all
+    Satisfy All
+</DirectoryMatch>
+
+#
 # ErrorLog: The location of the error log file.
 # If you do not specify an ErrorLog directive within a <VirtualHost>
 # container, error messages relating to that virtual host will be
