diff -u --exclude=CVS -r cvs-1.11.18.orig/src/wrapper.c cvs-1.11.18/src/wrapper.c
--- src/wrapper.c	2004-06-22 18:51:07.000000000 -0700
+++ src/wrapper.c	2004-12-07 16:26:23.000000000 -0800
@@ -420,7 +420,7 @@
 	    /* Before this is reenabled, need to address the problem in
 	       commit.c (see http://www.cvshome.org/docs/infowrapper.html).  */
 	    error (1, 0,
-		   "-t/-f wrappers not supported by this version of CVS");
+		   "-t/-f wrappers not supported by this version of CVS\nA legacy version of cvs with -t/-f wrapper support is available as: /usr/bin/ocvs.");
 
 	    if(e.fromcvsFilter)
 		free(e.fromcvsFilter);
@@ -434,7 +434,7 @@
 	    /* Before this is reenabled, need to address the problem in
 	       commit.c (see http://www.cvshome.org/docs/infowrapper.html).  */
 	    error (1, 0,
-		   "-t/-f wrappers not supported by this version of CVS");
+		   "-t/-f wrappers not supported by this version of CVS\nA legacy version of cvs with -t/-f wrapper support is available as: /usr/bin/ocvs.");
 
 	    if(e.tocvsFilter)
 		free(e.tocvsFilter);
