--- big5.c.orig	2009-11-09 15:24:27.000000000 -0800
+++ big5.c	2009-11-09 18:41:35.000000000 -0800
@@ -41,6 +41,8 @@ static char sccsid[] = "@(#)big5.c	8.1 (
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/locale/big5.c,v 1.18 2007/10/13 16:28:21 ache Exp $");
 
+#include "xlocale_private.h"
+
 #include <sys/types.h>
 #include <errno.h>
 #include <runetype.h>
@@ -49,33 +51,30 @@ __FBSDID("$FreeBSD: src/lib/libc/locale/
 #include <wchar.h>
 #include "mblocal.h"
 
-extern int __mb_sb_limit;
-
 static size_t	_BIG5_mbrtowc(wchar_t * __restrict, const char * __restrict,
-		    size_t, mbstate_t * __restrict);
-static int	_BIG5_mbsinit(const mbstate_t *);
+		    size_t, mbstate_t * __restrict, locale_t);
+static int	_BIG5_mbsinit(const mbstate_t *, locale_t);
 static size_t	_BIG5_wcrtomb(char * __restrict, wchar_t,
-		    mbstate_t * __restrict);
+		    mbstate_t * __restrict, locale_t);
 
 typedef struct {
 	wchar_t	ch;
 } _BIG5State;
 
-int
-_BIG5_init(_RuneLocale *rl)
+__private_extern__ int
+_BIG5_init(struct __xlocale_st_runelocale *xrl)
 {
 
-	__mbrtowc = _BIG5_mbrtowc;
-	__wcrtomb = _BIG5_wcrtomb;
-	__mbsinit = _BIG5_mbsinit;
-	_CurrentRuneLocale = rl;
-	__mb_cur_max = 2;
-	__mb_sb_limit = 128;
+	xrl->__mbrtowc = _BIG5_mbrtowc;
+	xrl->__wcrtomb = _BIG5_wcrtomb;
+	xrl->__mbsinit = _BIG5_mbsinit;
+	xrl->__mb_cur_max = 2;
+	xrl->__mb_sb_limit = 128;
 	return (0);
 }
 
 static int
-_BIG5_mbsinit(const mbstate_t *ps)
+_BIG5_mbsinit(const mbstate_t *ps, locale_t loc __unused)
 {
 
 	return (ps == NULL || ((const _BIG5State *)ps)->ch == 0);
@@ -91,7 +90,7 @@ _big5_check(u_int c)
 
 static size_t
 _BIG5_mbrtowc(wchar_t * __restrict pwc, const char * __restrict s, size_t n,
-    mbstate_t * __restrict ps)
+    mbstate_t * __restrict ps, locale_t loc __unused)
 {
 	_BIG5State *bs;
 	wchar_t wc;
@@ -151,7 +150,7 @@ _BIG5_mbrtowc(wchar_t * __restrict pwc, 
 }
 
 static size_t
-_BIG5_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps)
+_BIG5_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps, locale_t loc __unused)
 {
 	_BIG5State *bs;
 
