--- gb18030.c.bsdnew	2009-11-09 15:05:25.000000000 -0800
+++ gb18030.c	2009-11-09 15:50:31.000000000 -0800
@@ -32,6 +32,8 @@
 #include <sys/param.h>
 __FBSDID("$FreeBSD: src/lib/libc/locale/gb18030.c,v 1.8 2007/10/13 16:28:21 ache Exp $");
 
+#include "xlocale_private.h"
+
 #include <errno.h>
 #include <runetype.h>
 #include <stdlib.h>
@@ -39,35 +41,34 @@ __FBSDID("$FreeBSD: src/lib/libc/locale/
 #include <wchar.h>
 #include "mblocal.h"
 
-extern int __mb_sb_limit;
+#define GB18030_MB_CUR_MAX	4
 
 static size_t	_GB18030_mbrtowc(wchar_t * __restrict, const char * __restrict,
-		    size_t, mbstate_t * __restrict);
-static int	_GB18030_mbsinit(const mbstate_t *);
+		    size_t, mbstate_t * __restrict, locale_t);
+static int	_GB18030_mbsinit(const mbstate_t *, locale_t);
 static size_t	_GB18030_wcrtomb(char * __restrict, wchar_t,
-		    mbstate_t * __restrict);
+		    mbstate_t * __restrict, locale_t);
 
 typedef struct {
 	int	count;
 	u_char	bytes[4];
 } _GB18030State;
 
-int
-_GB18030_init(_RuneLocale *rl)
+__private_extern__ int
+_GB18030_init(struct __xlocale_st_runelocale *xrl)
 {
 
-	__mbrtowc = _GB18030_mbrtowc;
-	__wcrtomb = _GB18030_wcrtomb;
-	__mbsinit = _GB18030_mbsinit;
-	_CurrentRuneLocale = rl;
-	__mb_cur_max = 4;
-	__mb_sb_limit = 128;
+	xrl->__mbrtowc = _GB18030_mbrtowc;
+	xrl->__wcrtomb = _GB18030_wcrtomb;
+	xrl->__mbsinit = _GB18030_mbsinit;
+	xrl->__mb_cur_max = GB18030_MB_CUR_MAX;
+	xrl->__mb_sb_limit = 128;
 
 	return (0);
 }
 
 static int
-_GB18030_mbsinit(const mbstate_t *ps)
+_GB18030_mbsinit(const mbstate_t *ps, locale_t loc __unused)
 {
 
 	return (ps == NULL || ((const _GB18030State *)ps)->count == 0);
@@ -75,7 +76,7 @@ _GB18030_mbsinit(const mbstate_t *ps)
 
 static size_t
 _GB18030_mbrtowc(wchar_t * __restrict pwc, const char * __restrict s,
-    size_t n, mbstate_t * __restrict ps)
+    size_t n, mbstate_t * __restrict ps, locale_t loc __unused)
 {
 	_GB18030State *gs;
 	wchar_t wch;
@@ -95,7 +96,7 @@ _GB18030_mbrtowc(wchar_t * __restrict pw
 		pwc = NULL;
 	}
 
-	ncopy = MIN(MIN(n, MB_CUR_MAX), sizeof(gs->bytes) - gs->count);
+	ncopy = MIN(MIN(n, GB18030_MB_CUR_MAX), sizeof(gs->bytes) - gs->count);
 	memcpy(gs->bytes + gs->count, s, ncopy);
 	ocount = gs->count;
 	gs->count += ncopy;
@@ -158,7 +159,7 @@ ilseq:
 }
 
 static size_t
-_GB18030_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps)
+_GB18030_wcrtomb(char * __restrict s, wchar_t wc, mbstate_t * __restrict ps, locale_t loc __unused)
 {
 	_GB18030State *gs;
 	size_t len;
