Radar 5865058

Doing diffs in .:
--- ./binutils/cxxfilt.c.~1~	2005-11-17 08:58:27.000000000 -0800
+++ ./binutils/cxxfilt.c	2008-04-15 12:38:48.000000000 -0700
@@ -63,12 +63,12 @@ demangle_it (char *mangled_name)
   result = cplus_demangle (mangled_name + skip_first, flags);
 
   if (result == NULL)
-    printf (mangled_name);
+    printf ("%s",mangled_name);
   else
     {
       if (mangled_name[0] == '.')
 	putchar ('.');
-      printf (result);
+      printf ("%s",result);
       free (result);
     }
 }
--------------
