Radar 6018700

Doing diffs in .:
--- ./binutils/strings.c.~1~	2006-09-26 09:45:25.000000000 -0700
+++ ./binutils/strings.c	2008-06-18 12:12:32.000000000 -0700
@@ -106,7 +106,7 @@ typedef off64_t file_off;
 typedef off_t file_off;
 #define file_open(s,m) fopen(s, m)
 #endif
-#ifdef HAVE_STAT64
+#if defined(HAVE_STAT64) && ! defined(__APPLE__)
 typedef struct stat64 statbuf;
 #define file_stat(f,s) stat64(f, s)
 #else
--------------
