--- common.mk.orig	2016-04-17 21:57:25.000000000 -0700
+++ common.mk	2016-09-07 11:20:44.000000000 -0700
@@ -140,7 +140,7 @@ EXTMK_ARGS    =	$(SCRIPT_ARGS) --extensi
 		--make-flags="V=$(V) MINIRUBY='$(MINIRUBY)'" \
 		--gnumake=$(gnumake) --extflags="$(EXTLDFLAGS)" \
 		--
-INSTRUBY      =	$(SUDO) $(RUNRUBY) -r./$(arch)-fake $(srcdir)/tool/rbinstall.rb
+INSTRUBY      =	$(SUDO) SDKROOT=/ $(RUNRUBY) -r./$(arch)-fake $(srcdir)/tool/rbinstall.rb
 INSTRUBY_ARGS =	$(SCRIPT_ARGS) \
 		--data-mode=$(INSTALL_DATA_MODE) \
 		--prog-mode=$(INSTALL_PROG_MODE) \
