--- configure.orig	2018-12-24 22:28:23.000000000 -0800
+++ configure	2019-01-13 22:40:05.000000000 -0800
@@ -5978,13 +5978,7 @@ $as_echo_n "checking target architecture
 esac
     done
     target_archs="$universal_binary"
-    unset universal_binary
-    case "$target_archs" in #(
-  *,*) :
-    universal_binary=yes ;; #(
-  *) :
-    unset universal_archnames ;;
-esac
+    universal_binary=yes
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: $target_archs" >&5
     colorize_result "$target_archs" ; }
 
@@ -11761,20 +11755,15 @@ else
 _ACEOF
 if ac_fn_c_try_compile "$LINENO"; then :
 
-	# Check for potential -arch flags.  It is not universal unless
-	# there are at least two -arch flags with different values.
+	# Check for potential -arch flags.  It is then universal.
 	ac_arch=
 	ac_prev=
 	for ac_word in $CC $CFLAGS $CPPFLAGS $LDFLAGS; do
 	 if test -n "$ac_prev"; then
 	   case $ac_word in
 	     i?86 | x86_64 | ppc | ppc64)
-	       if test -z "$ac_arch" || test "$ac_arch" = "$ac_word"; then
-		 ac_arch=$ac_word
-	       else
 		 ac_cv_c_bigendian=universal
 		 break
-	       fi
 	       ;;
 	   esac
 	   ac_prev=
