--- lib/irb/init.original.rb	2019-05-01 18:04:05.000000000 -0700
+++ lib/irb/init.rb	2019-05-02 13:28:31.000000000 -0700
@@ -14,6 +14,14 @@
 
   # initialize config
   def IRB.setup(ap_path, argv: ::ARGV)
+    if STDOUT.tty?
+        print "\nWARNING: This version of ruby is included in macOS "\
+              "for compatibility with legacy software. "\
+              "\nIn future versions of "\
+              "macOS the ruby runtime will not be available by "\
+              "\ndefault, and may require you to install an "\
+              "additional package.\n\n"
+    end
     IRB.init_config(ap_path)
     IRB.init_error
     IRB.parse_opts(argv: argv)
