--- ./ext/digest/sha1/extconf.rb.orig	2012-11-26 16:58:52.000000000 -0800
+++ ./ext/digest/sha1/extconf.rb	2013-02-25 12:08:37.000000000 -0800
@@ -17,7 +17,7 @@ if !with_config("bundled-sha1") &&
     have_library("crypto") && OpenSSL.check_func("SHA1_Transform", "openssl/sha.h")
   $objs << "sha1ossl.#{$OBJEXT}"
 else
-  $objs << "sha1.#{$OBJEXT}"
+  $objs << "sha1cc.#{$OBJEXT}"
 end
 
 have_header("sys/cdefs.h")

--- ext/digest/sha1/sha1init.c~	2012-02-25 12:37:12.000000000 -0800
+++ ext/digest/sha1/sha1init.c	2013-02-27 23:12:52.000000000 -0800
@@ -5,7 +5,7 @@
 #if defined(HAVE_OPENSSL_SHA_H)
 #include "sha1ossl.h"
 #else
-#include "sha1.h"
+#include "sha1cc.h"
 #endif
 
 static const rb_digest_metadata_t sha1 = {
