--- ./ext/dlgest/md5/extconf.rb.orig	2013-02-25 11:42:40.000000000 -0800
+++ ./ext/digest/md5/extconf.rb	2013-02-25 12:05:37.000000000 -0800
@@ -18,7 +18,7 @@ if !with_config("bundled-md5") &&
   $objs << "md5ossl.#{$OBJEXT}"
 
 else
-  $objs << "md5.#{$OBJEXT}"
+  $objs << "md5cc.#{$OBJEXT}"
 end
 
 have_header("sys/cdefs.h")
--- ext/digest/md5/md5init.c~	2012-02-25 12:37:12.000000000 -0800
+++ ext/digest/md5/md5init.c	2013-02-27 23:14:50.000000000 -0800
@@ -5,7 +5,7 @@
 #if defined(HAVE_OPENSSL_MD5_H)
 #include "md5ossl.h"
 #else
-#include "md5.h"
+#include "md5cc.h"
 #endif
 
 static const rb_digest_metadata_t md5 = {
