--- ./ext/openssl/extconf.rb~	2012-12-19 23:42:56.000000000 -0800
+++ ./ext/openssl/extconf.rb	2013-02-25 13:18:39.000000000 -0800
@@ -58,9 +58,6 @@ end
 unless have_header("openssl/conf_api.h")
   raise "OpenSSL 0.9.6 or later required."
 end
-unless OpenSSL.check_func("SSL_library_init()", "openssl/ssl.h")
-  raise "Ignore OpenSSL broken by Apple.\nPlease use another openssl. (e.g. using `configure --with-openssl-dir=/path/to/openssl')"
-end
 
 Logging::message "=== Checking for OpenSSL features... ===\n"
 have_func("ERR_peek_last_error")
