--- ext/digest/md5/extconf.rb.orig	2006-03-20 14:55:31.000000000 +0100
+++ ext/digest/md5/extconf.rb	2006-03-20 14:55:01.000000000 +0100
@@ -9,13 +9,13 @@
 
 dir_config("openssl")
 
-if !with_config("bundled-md5") &&
-    have_library("crypto") && have_header("openssl/md5.h")
-  $objs << "md5ossl.#{$OBJEXT}"
+#if !with_config("bundled-md5") &&
+#    have_library("crypto") && have_header("openssl/md5.h")
+#  $objs << "md5ossl.#{$OBJEXT}"
 
-else
+#else
   $objs << "md5.#{$OBJEXT}"
-end
+#end
 
 have_header("sys/cdefs.h")
 
