--- ext/digest/sha1/extconf.rb.old	2007-03-14 16:28:41.000000000 +0100
+++ ext/digest/sha1/extconf.rb	2007-03-14 16:28:48.000000000 +0100
@@ -10,12 +10,12 @@
 
 dir_config("openssl")
 
-if !with_config("bundled-sha1") &&
-    have_library("crypto") && have_header("openssl/sha.h")
-  $objs << "sha1ossl.#{$OBJEXT}"
-else
+#if !with_config("bundled-sha1") &&
+#    have_library("crypto") && have_header("openssl/sha.h")
+#  $objs << "sha1ossl.#{$OBJEXT}"
+#else
   $objs << "sha1.#{$OBJEXT}"
-end
+#end
 
 have_header("sys/cdefs.h")
 
