--- lib/rdoc/ri/ri_paths.rb.orig	2007-09-25 01:26:30.000000000 +0200
+++ lib/rdoc/ri/ri_paths.rb	2007-09-25 01:26:47.000000000 +0200
@@ -44,7 +44,7 @@
 
     begin
       require 'rubygems'
-      GEMDIRS = Dir["#{Gem.path}/doc/*/ri"]
+      GEMDIRS = Gem.path.to_a.map { |p| Dir["#{p}/doc/*/ri"] }.flatten
       GEMDIRS.each { |path| RI::Paths::PATH << path }
     rescue LoadError
       GEMDIRS = nil
