--- configure.orig	2005-08-31 12:51:55.000000000 -0700
+++ configure	2005-08-31 12:53:31.000000000 -0700
@@ -15332,9 +15332,11 @@
 _ACEOF
 
 else
-    arch="${target_cpu}-${target_os}"
+    arch="universal-${target_os}"
     cat >>confdefs.h <<_ACEOF
 #define RUBY_PLATFORM "${arch}"
+#define RUBY_THIN_ARCHLIB "${RUBY_LIB_PATH}/powerpc-${target_os}"
+#define RUBY_SITE_THIN_ARCHLIB "${RUBY_SITE_LIB_PATH2}/powerpc-${target_os}"
 _ACEOF
 
 fi
