--- process.c.old	2007-11-16 14:19:21.000000000 +0100
+++ process.c	2007-11-16 14:20:55.000000000 +0100
@@ -96,7 +96,7 @@
 #undef HAVE_SETRGID
 #endif
 
-#if defined(__MacOS_X__) || defined(__bsdi__)
+#if defined(__bsdi__)
 #define BROKEN_SETREUID 1
 #define BROKEN_SETREGID 1
 #endif
@@ -108,7 +108,7 @@
 #define setregid ruby_setregid
 #endif
 
-#if defined(HAVE_44BSD_SETUID) || defined(__MacOS_X__)
+#if defined(HAVE_44BSD_SETUID)
 #if !defined(USE_SETREUID) && !defined(BROKEN_SETREUID)
 #define OBSOLETE_SETREUID 1
 #endif
