--- rbconfig.rb.orig	2007-09-17 14:30:50.000000000 +0200
+++ rbconfig.rb	2007-09-17 14:33:48.000000000 +0200
@@ -209,7 +209,8 @@ module RbConfig
   CONFIG["infodir"] = "$(DESTDIR)/usr/share/info"
   CONFIG["docdir"] = "$(datarootdir)/doc/$(PACKAGE)"
   CONFIG["oldincludedir"] = "/usr/include"
-  CONFIG["includedir"] = "$(prefix)/include"
+  DEFAULTSDK = ENV['SDKROOT'] || %x{xcode-select --print-path >/dev/null 2>&1 && xcrun --sdk macosx --show-sdk-path 2>/dev/null}.chomp
+  CONFIG["includedir"] = DEFAULTSDK + "$(prefix)/include"
   CONFIG["localstatedir"] = "$(prefix)/var"
   CONFIG["sharedstatedir"] = "$(prefix)/com"
   CONFIG["sysconfdir"] = "$(DESTDIR)/Library/Ruby/Site"
@@ -263,3 +263,6 @@ module RbConfig
 end
 RbConfig = Config # compatibility for ruby-1.9
 CROSS_COMPILING = nil unless defined? CROSS_COMPILING
+RUBY_FRAMEWORK = true
+RUBY_FRAMEWORK_VERSION = RbConfig::CONFIG['ruby_version']
+APPLE_GEM_HOME = File.join(RbConfig::CONFIG['libdir'], 'ruby/gems', RbConfig::CONFIG['ruby_version'])
