--- rbconfig.rb.orig	2007-09-17 14:30:50.000000000 +0200
+++ rbconfig.rb	2007-09-17 14:33:48.000000000 +0200
@@ -8,6 +8,14 @@
 
   TOPDIR = File.dirname(__FILE__).chomp!("/lib/ruby/1.8/universal-darwin9.0")
   DESTDIR = '' unless defined? DESTDIR
+  ARCHFLAGS = 
+    if e = ENV['ARCHFLAGS']
+      e
+    elsif e = ENV['RC_ARCHS']
+      e.split.map { |a| "-arch #{a}" }.join(' ')
+    else
+      '-arch i386 -arch x86_64'
+    end
   CONFIG = {}
   CONFIG["DESTDIR"] = DESTDIR
   CONFIG["INSTALL"] = '/usr/bin/install -c'
@@ -176,3 +184,6 @@
 end
 RbConfig = Config # compatibility for ruby-1.9
 CROSS_COMPILING = nil unless defined? CROSS_COMPILING
+RUBY_FRAMEWORK = true
+RUBY_FRAMEWORK_VERSION = Config::CONFIG['ruby_version']
+APPLE_GEM_HOME = File.join(Config::CONFIG['libdir'], 'ruby/gems', Config::CONFIG['ruby_version'])
