--- rbconfig.rb.orig	2007-09-17 14:30:50.000000000 +0200
+++ rbconfig.rb	2007-09-17 14:33:48.000000000 +0200
@@ -8,6 +8,14 @@
 
   TOPDIR = File.dirname(__FILE__).chomp!("/lib/ruby/1.8/universal-darwin9.0")
   DESTDIR = '' unless defined? DESTDIR
+  ARCHFLAGS = 
+    if e = ENV['ARCHFLAGS']
+      e
+    elsif e = ENV['RC_ARCHS']
+      e.split.map { |a| "-arch #{a}" }.join(' ')
+    else
+      '-arch ppc -arch i386'
+    end
   CONFIG = {}
   CONFIG["DESTDIR"] = DESTDIR
   CONFIG["INSTALL"] = '/usr/bin/install -c'
@@ -66,8 +74,8 @@
   CONFIG["target_vendor"] = "apple"
   CONFIG["target_os"] = "darwin9.0"
   CONFIG["CC"] = "gcc"
-  CONFIG["CFLAGS"] = "    -g -Os -pipe -fno-common -DENABLE_DTRACE     -pipe  -fno-common  -pipe -fno-common"
-  CONFIG["LDFLAGS"] = "-L.                "
+  CONFIG["CFLAGS"] = "#{ARCHFLAGS} -Os -pipe -fno-common"
+  CONFIG["LDFLAGS"] = "-L. #{ARCHFLAGS}"
   CONFIG["CPPFLAGS"] = " $(DEFS)"
   CONFIG["OBJEXT"] = "o"
   CONFIG["CPP"] = "gcc -E"
@@ -99,7 +107,7 @@
   CONFIG["ARCH_FLAG"] = ""
   CONFIG["STATIC"] = ""
   CONFIG["CCDLFLAGS"] = " -fno-common"
-  CONFIG["LDSHARED"] = "cc     -pipe -bundle -undefined dynamic_lookup"
+  CONFIG["LDSHARED"] = "cc #{ARCHFLAGS} -pipe -bundle -undefined dynamic_lookup"
   CONFIG["DLEXT"] = "bundle"
   CONFIG["DLEXT2"] = ""
   CONFIG["LIBEXT"] = "a"
@@ -119,7 +127,7 @@
   CONFIG["RDOCTARGET"] = ""
   CONFIG["XCFLAGS"] = " -DRUBY_EXPORT"
   CONFIG["XLDFLAGS"] = ""
-  CONFIG["LIBRUBY_LDSHARED"] = "cc     -pipe -dynamiclib"
+  CONFIG["LIBRUBY_LDSHARED"] = "cc #{ARCHFLAGS} -pipe -dynamiclib"
   CONFIG["LIBRUBY_DLDFLAGS"] = "-install_name $(libdir)/lib$(RUBY_SO_NAME).$(MAJOR).dylib -current_version $(MAJOR).$(MINOR).$(TEENY) -compatibility_version $(MAJOR).$(MINOR)"
   CONFIG["rubyw_install_name"] = ""
   CONFIG["RUBYW_INSTALL_NAME"] = ""
@@ -176,3 +184,6 @@
 end
 RbConfig = Config # compatibility for ruby-1.9
 CROSS_COMPILING = nil unless defined? CROSS_COMPILING
+RUBY_FRAMEWORK = true
+RUBY_FRAMEWORK_VERSION = Config::CONFIG['ruby_version']
+APPLE_GEM_HOME = File.join(Config::CONFIG['libdir'], 'ruby/gems', Config::CONFIG['ruby_version'])
