diff --git configure configure
old mode 100755
new mode 100644
index 08ad869..03efd20
--- configure
+++ configure
@@ -7809,93 +7809,6 @@ rm -fr conf$$.dir
 }
 
 : "compiler section" && {
-save_CFLAGS="$CFLAGS"
-CFLAGS="$CFLAGS $rb_cv_warnflags"
-if test "${ac_c_werror_flag+set}"; then :
-
-  rb_c_werror_flag="$ac_c_werror_flag"
-
-else
-
-  unset rb_c_werror_flag
-
-fi
-ac_c_werror_flag=yes
-
-    { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether CFLAGS is valid" >&5
-$as_echo_n "checking whether CFLAGS is valid... " >&6; }
-    cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-int
-main ()
-{
-
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
-  { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-    $as_echo "${msg_result_yes}yes${msg_reset}" >&6 ; }
-else
-
-	{ $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
-    $as_echo "${msg_result_no}no${msg_reset}" >&6 ; }
-	as_fn_error $? "something wrong with CFLAGS=\"$CFLAGS\"" "$LINENO" 5
-
-
-fi
-rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
-
-    { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether LDFLAGS is valid" >&5
-$as_echo_n "checking whether LDFLAGS is valid... " >&6; }
-    {
-	mkdir tmp.$$.try_link &&
-	cd tmp.$$.try_link &&
-	cp ../confdefs.h . &&
-	echo '<?xml?><plist><dict><key>CFBundleIdentifier</key><string></string></dict></plist>' > Info.plist &&
-	:
-    } || as_fn_error $? "failed to make temporary directory" "$LINENO" 5
-    cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-int
-main ()
-{
-
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_link "$LINENO"; then :
-  { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-    $as_echo "${msg_result_yes}yes${msg_reset}" >&6 ; }
-else
-
-	cd .. && rm -fr tmp.$$.try_link
-	{ $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
-    $as_echo "${msg_result_no}no${msg_reset}" >&6 ; }
-	as_fn_error $? "something wrong with LDFLAGS=\"$LDFLAGS\"" "$LINENO" 5
-
-
-fi
-rm -f core conftest.err conftest.$ac_objext \
-    conftest$ac_exeext conftest.$ac_ext
-    cd .. && rm -fr tmp.$$.try_link
-
-CFLAGS="$save_CFLAGS"
-save_CFLAGS=
-if test "${rb_c_werror_flag+set}"; then :
-
-  ac_c_werror_flag="$rb_c_werror_flag"
-
-else
-
-  unset ac_c_werror_flag
-
-fi
-
 : ${RPATHFLAG=''}
 rpathflag=''
 if test x"${RPATHFLAG}" = x; then :
