diff --git ext/socket/extconf.rb ext/socket/extconf.rb
index 1cde64a..4c62647 100644
--- ext/socket/extconf.rb
+++ ext/socket/extconf.rb
@@ -671,7 +671,8 @@ EOS
 int t(struct in6_addr *addr) {return IN6_IS_ADDR_UNSPECIFIED(addr);}
 SRC
     print "fixing apple's netinet6/in6.h ..."; $stdout.flush
-    in6 = File.read("/usr/include/#{hdr}")
+    sdkroot = ENV["SDKROOT"] || ""
+    in6 = File.read(sdkroot + "/usr/include/#{hdr}")
     if in6.gsub!(/\*\(const\s+__uint32_t\s+\*\)\(const\s+void\s+\*\)\(&(\(\w+\))->s6_addr\[(\d+)\]\)/) do
         i, r = $2.to_i.divmod(4)
         if r.zero?
