diff --git tool/config.guess tool/config.guess
index 1972fda..3753fc1 100755
--- tool/config.guess
+++ tool/config.guess
@@ -132,7 +132,7 @@ if test -f /.attbin/uname ; then
 fi
 
 UNAME_MACHINE=$( (uname -m) 2>/dev/null) || UNAME_MACHINE=unknown
-UNAME_RELEASE=$( (uname -r) 2>/dev/null) || UNAME_RELEASE=unknown
+UNAME_RELEASE=$( (uname -r | cut -d . -f1) 2>/dev/null) || UNAME_RELEASE=unknown
 UNAME_SYSTEM=$( (uname -s) 2>/dev/null) || UNAME_SYSTEM=unknown
 UNAME_VERSION=$( (uname -v) 2>/dev/null) || UNAME_VERSION=unknown
 
