--- tool/mkconfig.rb.orig	2013-02-06 23:21:30.000000000 -0800
+++ tool/mkconfig.rb	2013-04-22 15:06:23.000000000 -0700
@@ -127,10 +127,6 @@ File.foreach "config.status" do |line|
       val = "arch_flag || #{val}" if universal
     when /^UNIVERSAL_ARCHNAMES$/
       universal, val = val, 'universal' if universal
-    when /^arch$/
-      if universal
-        val.sub!(/universal/, %q[#{arch && universal[/(?:\A|\s)#{Regexp.quote(arch)}=(\S+)/, 1] || '\&'}])
-      end
     end
     v = "  CONFIG[\"#{name}\"] #{eq} #{val}\n"
     if fast[name]
@@ -172,8 +168,7 @@ print "  TOPDIR = File.dirname(__FILE__)
 print "  DESTDIR = ", (drive ? "TOPDIR && TOPDIR[/\\A[a-z]:/i] || " : ""), "'' unless defined? DESTDIR\n"
 print <<'ARCH' if universal
   arch_flag = ENV['ARCHFLAGS'] || ((e = ENV['RC_ARCHS']) && e.split.uniq.map {|a| "-arch #{a}"}.join(' '))
-  arch = arch_flag && arch_flag[/\A\s*-arch\s+(\S+)\s*\z/, 1]
 ARCH
 print "  universal = #{universal}\n" if universal
 print "  CONFIG = {}\n"
 
