--- tool/config.guess.orig	2010-09-25 18:25:34.000000000 -0700
+++ tool/config.guess	2013-02-11 22:32:56.000000000 -0800
@@ -135,7 +135,7 @@ if (test -f /.attbin/uname) >/dev/null 2
 fi
 
 UNAME_MACHINE=`(uname -m) 2>/dev/null` || UNAME_MACHINE=unknown
-UNAME_RELEASE=`(uname -r) 2>/dev/null` || UNAME_RELEASE=unknown
+UNAME_RELEASE=`(uname -r | cut -d . -f1) 2>/dev/null` || UNAME_RELEASE=unknown
 UNAME_SYSTEM=`(uname -s) 2>/dev/null`  || UNAME_SYSTEM=unknown
 UNAME_VERSION=`(uname -v) 2>/dev/null` || UNAME_VERSION=unknown
 
