diff --git a/ruby-2.6.10/tool/mkconfig.rb b/ruby-2.6.10/tool/mkconfig.rb
index 4b032cf..32fec35 100755
--- tool/mkconfig.rb
+++ tool/mkconfig.rb
@@ -133,6 +133,10 @@ File.foreach "config.status" do |line|
       val = '"xcrun clang"'
     when /^CXX$/
       val = '"xcrun clang++"'
+    when /^host_cpu$/
+      val = 'arch'
+    when /^host$/
+      val = 'arch + "-" + CONFIG["host_vendor"] + "-" + CONFIG["host_os"]'
     end
     v = "  CONFIG[\"#{name}\"] #{eq} #{val}\n"
     if fast[name]
@@ -195,6 +199,7 @@ print <<'ARCH' if universal
   arch_flag = ENV['ARCHFLAGS'] || ((e = ENV['RC_ARCHS']) && e.split.uniq.map {|a| "-arch #{a}"}.join(' '))
 ARCH
 print "  universal = #{universal}\n" if universal
+print "  arch = 'arm64' == `uname -m`.strip ? 'aarch64' : 'x86_64'"
 print "  # The hash configurations stored.\n"
 print "  CONFIG = {}\n"
 print "  CONFIG[\"DESTDIR\"] = DESTDIR\n"
