diff -uNr ../sudo-1.6.8p9.orig/sudo.man.in ./sudo.man.in
--- ../sudo-1.6.8p9.orig/sudo.man.in	2005-06-19 13:05:34.000000000 -0700
+++ ./sudo.man.in	2006-09-28 12:34:06.000000000 -0700
@@ -151,7 +151,7 @@
 .IX Title "SUDO @mansectsu@"
 .TH SUDO @mansectsu@ "June 20, 2005" "1.6.8p9" "MAINTENANCE COMMANDS"
 .SH "NAME"
-sudo, sudoedit \- execute a command as another user
+sudo \- execute a command as another user
 .SH "SYNOPSIS"
 .IX Header "SYNOPSIS"
 \&\fBsudo\fR \fB\-K\fR | \fB\-L\fR | \fB\-V\fR | \fB\-h\fR | \fB\-k\fR | \fB\-l\fR | \fB\-v\fR
