--- bin/package+orig	2006-02-25 18:20:28.000000000 -0800
+++ bin/package	2006-02-25 18:20:48.000000000 -0800
@@ -1897,7 +1897,7 @@ main()
 			fi
 			split='
 '
-			a=`strings $a < /dev/null | sed -e 's/[^abcdefghijklmnopqrstuvwxyz0123456789]/ /g' -e 's/[ 	][ 	]*/\'"$split"'/g' | sed -e "/^${type}[0123456789]$/!d" -e "s/^${type}//" -e q`
+			a=`strings $a < /dev/null | sed -e 's/[^abcdefghijklmnopqrstuvwxyz0123456789]/ /g' -e "s/[ 	][ 	]*/'\"$split\"'/g" | sed -e "/^${type}[0123456789]$/!d" -e "s/^${type}//" -e q`
 			case $a in
 			[0123456789])	n=$a ;;
 			esac
