diff -up -r /SourceCache/ksh/ksh-11/ksh/src/lib/libast/string/strmatch.c ./src/lib/libast/string/strmatch.c
--- /SourceCache/ksh/ksh-11/ksh/src/lib/libast/string/strmatch.c	2006-07-22 21:09:35.000000000 -0700
+++ ./src/lib/libast/string/strmatch.c	2007-06-11 15:36:38.000000000 -0700
@@ -92,8 +92,13 @@ strgrpmatch(const char* b, const char* p
 			regcache(NiL, 0, NiL);
 		return 0;
 	}
-	if (!*p)
+	if (!*p) {
+		if (sub) {
+			sub[0] = 0;
+			sub[1] = 0;
+		}
 		return *b == 0;
+	}
 
 	/*
 	 * convert flags
