--- src/cmd/INIT/package.sh+orig	2006-02-25 18:53:20.000000000 -0800
+++ src/cmd/INIT/package.sh	2006-02-25 18:53:57.000000000 -0800
@@ -1915,7 +1915,7 @@ main()
 			fi
 			split='
 '
-			a=`strings $a < /dev/null | sed -e 's/[^abcdefghijklmnopqrstuvwxyz0123456789]/ /g' -e 's/[ 	][ 	]*/\'"$split"'/g' | sed -e "/^${type}[0123456789]$/!d" -e "s/^${type}//" -e q`
+			a=`strings $a < /dev/null | sed -e 's/[^abcdefghijklmnopqrstuvwxyz0123456789]/ /g' -e "s/[ 	][ 	]*/'\"$split\"'/g" | sed -e "/^${type}[0123456789]$/!d" -e "s/^${type}//" -e q`
 			case $a in
 			[0123456789])	n=$a ;;
 			esac
