--- ../../libxml2.pristine/libxml2-2.6.14/HTMLtree.c	2004-09-20 13:01:04.000000000 -0700
+++ HTMLtree.c	2004-11-29 10:08:47.000000000 -0800
@@ -287,7 +287,7 @@
  * output as <option selected>, as per XSLT 1.0 16.2 "HTML Output Method"
  *
  */
-static const char* htmlBooleanAttrs[] = {
+static const char* const htmlBooleanAttrs[] = {
   "checked", "compact", "declare", "defer", "disabled", "ismap",
   "multiple", "nohref", "noresize", "noshade", "nowrap", "readonly",
   "selected", NULL
