--- include/libxml/xmlversion.h.orig	2004-09-13 15:43:54.000000000 -0700
+++ include/libxml/xmlversion.h	2004-09-13 15:44:36.000000000 -0700
@@ -84,7 +84,7 @@
  * Whether the thread support is configured in
  */
 #if 1
-#if defined(_REENTRANT) || defined(__MT__) || (_POSIX_C_SOURCE - 0 >= 199506L)
+#if defined(_REENTRANT) || defined(__MT__) || (defined(_POSIX_C_SOURCE) && (_POSIX_C_SOURCE - 0 >= 199506L))
 #define LIBXML_THREAD_ENABLED
 #endif
 #endif
