--- xmlwriter.c.orig.c	2004-09-15 01:59:35.000000000 -0700
+++ xmlwriter.c	2004-11-29 10:31:23.000000000 -0800
@@ -1371,7 +1371,7 @@
 xmlOutputBufferWriteBase64(xmlOutputBufferPtr out, int len,
                            const unsigned char *data)
 {
-    static unsigned char dtable[64] =
+    static unsigned char const dtable[64] =
             {'A','B','C','D','E','F','G','H','I','J','K','L','M',
 	     'N','O','P','Q','R','S','T','U','V','W','X','Y','Z',
 	     'a','b','c','d','e','f','g','h','i','j','k','l','m',
@@ -1501,7 +1501,7 @@
 {
     int count;
     int sum;
-    static char hex[16] = 
+    static char const hex[16] = 
     	{'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
     int i;
 
