--- applesingle.py.orig	Wed Oct  8 15:38:34 2003
+++ applesingle.py	Wed Oct  8 15:40:32 2003
@@ -8,14 +8,14 @@
 verbose=0
 
 # File header format: magic, version, unused, number of entries
-AS_HEADER_FORMAT="ll16sh"
+AS_HEADER_FORMAT=">ll16sh"
 AS_HEADER_LENGTH=26
 # The flag words for AppleSingle
 AS_MAGIC=0x00051600
 AS_VERSION=0x00020000
 
 # Entry header format: id, offset, length
-AS_ENTRY_FORMAT="lll"
+AS_ENTRY_FORMAT=">lll"
 AS_ENTRY_LENGTH=12
 
 # The id values
