diff -ur a/libclamav/binhex.c b/libclamav/binhex.c
--- a/libclamav/binhex.c	2008-08-16 12:18:08.000000000 -0500
+++ b/libclamav/binhex.c	2008-11-11 22:48:19.000000000 -0600
@@ -158,7 +158,7 @@
 	line = NULL;
 
 	while(bytesleft > 0) {
-		int length = 0;
+		size_t length = 0;
 		char *ptr, *newline;
 
 		/*printf("%d: ", bytesleft);*/
diff -ur a/libclamav/mbox.c b/libclamav/mbox.c
--- a/libclamav/mbox.c	2008-09-01 12:30:40.000000000 -0500
+++ b/libclamav/mbox.c	2008-11-11 22:48:19.000000000 -0600
@@ -778,7 +778,7 @@
 
 				lastline = 0;
 				do {
-					int length = 0, datalen;
+					size_t length = 0, datalen;
 					char *newline, *equal;
 					unsigned char *bigbuf, *data;
 					unsigned char smallbuf[1024];
