--- unix/configure~	Thu May  4 12:49:38 2006
+++ unix/configure	Thu May  4 12:49:38 2006
@@ -5691,18 +5691,10 @@
 
   darwin* | rhapsody*)
-    case "$host_os" in
-    rhapsody* | darwin1.[012])
-      allow_undefined_flag='-undefined suppress'
-      ;;
-    *) # Darwin 1.3 on
-      allow_undefined_flag='-flat_namespace -undefined suppress'
-      ;;
-    esac
     # FIXME: Relying on posixy $() will cause problems for
     #        cross-compilation, but unfortunately the echo tests do not
     #        yet detect zsh echo's removal of \ escapes.  Also zsh mangles
     #	     `"' quotes if we put them in here... so don't!
-    archive_cmds='$CC -r -keep_private_externs -nostdlib -o ${lib}-master.o $libobjs && $CC $(test .$module = .yes && echo -bundle || echo -dynamiclib) $allow_undefined_flag -o $lib ${lib}-master.o $deplibs$linker_flags $(test .$module != .yes && echo -install_name $rpath/$soname $verstring)'
+    archive_cmds='$CC ${RC_CFLAGS} -r -keep_private_externs -nostdlib -o ${lib}-master.o $libobjs && $CC ${RC_CFLAGS} $(test .$module = .yes && echo -bundle || echo -dynamiclib) $allow_undefined_flag -o $lib ${lib}-master.o $deplibs$linker_flags $(test .$module != .yes && echo -install_name $soname $verstring)'
     # We need to add '_' to the symbols in $export_symbols first
     #archive_expsym_cmds="$archive_cmds"' && strip -s $export_symbols'
     hardcode_direct=yes
--- unix/Makefile.in~	Mon Jan 26 01:54:45 2004
+++ unix/Makefile.in	Wed Feb 14 03:12:03 2007
@@ -156,7 +156,7 @@
 libmk4tcl.la: mk4tcl.o mk4too.o $(LOBJS) $(LINK_SPECIAL_FILES)
 	$L --mode=link $(CXX) -o $@ $(CXX_SWITCHES) $(LIBTOOL_SHLIB_FLAGS) \
 		-avoid-version -rpath $(libdir) \
-		mk4tcl.lo mk4too.lo $(SHLOBJS) $(LINK_SPECIAL_FLAGS)
+		mk4tcl.lo mk4too.lo $(SHLOBJS) $(LINK_SPECIAL_FLAGS) -Wl,-exported_symbols_list -Wl,../Mk4tcl.exports
 
 Mk4py@MODULE_SUFFIX@: libmk4py.la
 	cp .libs/libmk4py@MODULE_SUFFIX@ $@

