diff -rupN xotcl-1.2.0/unix/Makefile.in xotcl/unix/Makefile.in
--- unix/Makefile.in	Tue Feb 17 10:32:01 2004
+++ unix/Makefile.in	Wed Jun  2 22:53:31 2004
@@ -269,12 +269,12 @@ mkinstalldirs = $(SHELL) $(src_config_di
 CPPFLAGS = @CPPFLAGS@
 LIBS = @LIBS@
 AR = ar
-CFLAGS=		@CFLAGS@ @CFLAGS_DEBUG@  \
-			-DXOTCLVERSION=\"$(VERSION)\"  \
+CFLAGS=		@CFLAGS@ @CFLAGS_DEBUG@
+EXTRA_CFLAGS=		-DXOTCLVERSION=\"$(VERSION)\"  \
 			-DXOTCLPATCHLEVEL=\"$(PATCHLEVEL)\"  \
 			$(DMALLOC_DEFINES)
-COMPILE = $(CC) $(DEFS_WITH_STUBS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
-COMPILE_NO_STUBS = $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+COMPILE = $(CC) $(DEFS_WITH_STUBS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) $(EXTRA_CFLAGS)
+COMPILE_NO_STUBS = $(CC) $(DEFS) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) $(EXTRA_CFLAGS)
 CCLD = $(CC)
 LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(LDFLAGS) -o $@
 
@@ -310,7 +310,6 @@ libraries: 
 	for dir in $(subdirs) ; do \
 	   if (cd $$dir; $(MAKE) $@) ; then true ; else exit 1 ; fi ; \
 	done; fi;
-	@$(TCLSH_PROG) $(src_lib_dir)/lib/make.xotcl -dir $(src_lib_dir) -all
 
 # use language reference as sample file to trigger generation of documentation files
 doc: $(src_doc_dir)/langRef-xotcl.html
@@ -582,7 +581,7 @@ distclean: clean
 # You should not have to modify this target.
 #========================================================================
 
-install-lib-binaries: $(DESTDIR)$(bindir) $(DESTDIR)$(libdir)
+install-lib-binaries: $(DESTDIR)$(bindir) $(DESTDIR)$(pkglibdir)
 	@list='$(lib_BINARIES)'; for p in $$list; do \
 	  if test -f $$p; then \
 	    ext=`echo $$p|sed -e "s/.*\.//"`; \
@@ -591,12 +590,12 @@ install-lib-binaries: $(DESTDIR)$(bindir
 	        $(INSTALL_PROGRAM) $$p $(DESTDIR)$(bindir)/$$p; \
 		lib=`basename $$p|sed -e 's/.[^.]*$$//'`.lib; \
 		if test -f $$lib; then \
-		    echo " $(INSTALL_PROGRAM) $$lib $(DESTDIR)$(libdir)/$$lib"; \
-	            $(INSTALL_PROGRAM) $$lib $(DESTDIR)$(libdir)/$$lib; \
+		    echo " $(INSTALL_PROGRAM) $$lib $(DESTDIR)$(pkglibdir)/$$lib"; \
+	            $(INSTALL_PROGRAM) $$lib $(DESTDIR)$(pkglibdir)/$$lib; \
 		fi; \
 	    else \
-		echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(libdir)/$$p"; \
-	        $(INSTALL_PROGRAM) $$p $(DESTDIR)$(libdir)/$$p; \
+		echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p"; \
+	        $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p; \
 	    fi; \
 	  fi; \
 	done
@@ -604,11 +603,11 @@ install-lib-binaries: $(DESTDIR)$(bindir
 	  if test -f $$p; then \
 	    stub=`echo $$p|sed -e "s/.*\(stub\).*/\1/"`; \
 	    if test "x$$stub" = "xstub"; then \
-		    echo " $(RANLIB_STUB) $(DESTDIR)$(libdir)/$$p"; \
-		    $(RANLIB_STUB) $(DESTDIR)$(libdir)/$$p; \
+		    echo " $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p"; \
+		    $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p; \
 	    else \
-		    echo " $(RANLIB) $(DESTDIR)$(libdir)/$$p"; \
-		    $(RANLIB) $(DESTDIR)$(libdir)/$$p; \
+		    echo " $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p"; \
+		    $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p; \
 	    fi; \
 	  fi; \
 	done
diff -rupN xotcl-1.2.0/unix/pkgIndex.unix.in xotcl/unix/pkgIndex.unix.in
--- unix/pkgIndex.unix.in	Fri Oct 24 07:57:34 2003
+++ unix/pkgIndex.unix.in	Wed Jun  2 22:51:08 2004
@@ -1,4 +1,4 @@
   package ifneeded XOTcl @VERSION@ [list load \
-    [file join $dir .. @xotcl_LIB_FILE@] XOTcl]
+    [file join $dir @xotcl_LIB_FILE@] XOTcl]
 
 
