--- Makefile.in	2008-05-19 10:11:49.000000000 -0700
+++ Makefile.in.new	2008-05-19 10:32:51.000000000 -0700
@@ -530,18 +530,19 @@
 clean-libtool:
 	-rm -rf .libs _libs
 
+# Skipping to fix B&I build failure
 elc-stamp: $(LISP)
 	@echo 'WARNING: Warnings can be ignored. :-)'
 	@rm -f elc-temp && touch elc-temp
-	if test "$(EMACS)" != no; then \
-	  set x; \
-	  list='$(LISP)'; for p in $$list; do \
-	    if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
-	    set x "$$@" "$$d$$p"; shift; \
-	  done; \
-	  shift; \
-	  EMACS="$(EMACS)" $(SHELL) $(elisp_comp) "$$@" || exit 1; \
-	else : ; fi
+#	if test "$(EMACS)" != no; then \
+#	  set x; \
+#	  list='$(LISP)'; for p in $$list; do \
+#	    if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \
+#	    set x "$$@" "$$d$$p"; shift; \
+#	  done; \
+#	  shift; \
+#	  EMACS="$(EMACS)" $(SHELL) $(elisp_comp) "$$@" || exit 1; \
+#	else : ; fi
 	@mv -f elc-temp $@
 $(am__ELCFILES): elc-stamp
 	@if test "$(EMACS)" != no && test ! -f $@; then \
