--- a/Makefile.PL	2010-03-16 08:49:25.000000000 -0600
+++ b/Makefile.PL	2010-05-21 11:53:09.000000000 -0600
@@ -155,6 +155,8 @@
         'sa-learn'         => '$(INST_MAN1DIR)/sa-learn.$(MAN1EXT)',
         'sa-update'        => '$(INST_MAN1DIR)/sa-update.$(MAN1EXT)',
         'sa-compile'       => '$(INST_MAN1DIR)/sa-compile.$(MAN1EXT)',
+        'sa-awl'           => '$(INST_MAN1DIR)/sa-awl.$(MAN1EXT)',
+        'sa-check_spamd'   => '$(INST_MAN1DIR)/sa-check_spamd.$(MAN1EXT)',
         'spamc/spamc.pod'  => '$(INST_MAN1DIR)/spamc.$(MAN1EXT)',
         'spamd/spamd'      => '$(INST_MAN1DIR)/spamd.$(MAN1EXT)',
     },

