--- tran.c~	2007-10-23 14:28:05.000000000 -0700
+++ tran.c	2008-11-12 12:17:57.000000000 -0800
@@ -378,8 +378,16 @@ static char *get_str_val(Cell *vp, char 
 		else
 			sprintf(s, *fmt, vp->fval);
 		vp->sval = tostring(s);
+#if 0
+		/*
+		  Cannot reuse the converted form unless you confirm
+		  that the current value of CONVFMT is the same as that
+		  which was used to convert the previous string. See
+		  conformance test awk.ex 233
+		*/
 		vp->tval &= ~DONTFREE;
 		vp->tval |= STR;
+#endif
 	}
 	   dprintf( ("getsval %p: %s = \"%s (%p)\", t=%o\n", vp, NN(vp->nval), vp->sval, vp->sval, vp->tval) );
 	return(vp->sval);
