--- subversion/libsvn_ra_svn/client.c.orig	2009-03-26 14:38:22.000000000 -0700
+++ subversion/libsvn_ra_svn/client.c	2009-03-26 14:38:47.000000000 -0700
@@ -447,6 +447,8 @@
   if (status != APR_SUCCESS)
     return svn_error_wrap_apr(status, _("Can't create tunnel"));
 
+  apr_pool_note_subprocess(pool, proc, APR_KILL_ALWAYS);
+
   /* APR pipe objects inherit by default.  But we don't want the
    * tunnel agent's pipes held open by future child processes
    * (such as other ra_svn sessions), so turn that off. */
