--- configure.orig	2007-04-02 20:32:58.000000000 -0700
+++ configure	2007-04-02 20:33:09.000000000 -0700
@@ -3942,7 +3942,7 @@
 
 # Either a space-separated list of allowable Neon versions, or "any" to
 # mean allow anything.
-NEON_ALLOWED_LIST="0.24.7 0.25.0 0.25.1 0.25.2 0.25.3 0.25.4 0.25.5 0.26.0 0.26.1 0.26.2"
+NEON_ALLOWED_LIST="0.24.7 0.25.0 0.25.1 0.25.2 0.25.3 0.25.4 0.25.5 0.26.0 0.26.1 0.26.2 0.26.3"
 NEON_LATEST_WORKING_VER=0.25.5
 NEON_URL="http://www.webdav.org/neon/neon-${NEON_LATEST_WORKING_VER}.tar.gz"
 # Check whether --enable-neon-version-check or --disable-neon-version-check was given.
