--- configure.orig	2010-12-01 13:06:27.000000000 -0800
+++ configure	2010-12-01 13:06:29.000000000 -0800
@@ -24084,7 +24084,7 @@
     #   packages/rpm/rhel-4/subversion.spec
     if test -n "$SWIG_VERSION" &&
        test "$SWIG_VERSION" -ge "103024" &&
-       test "$SWIG_VERSION" -le "103036"; then
+       test "$SWIG_VERSION" -le "103040"; then
       SWIG_SUITABLE=yes
     else
       SWIG_SUITABLE=no
