--- getenv.c.orig	Mon Apr 28 16:37:26 2003
+++ getenv.c	Sat May  3 14:29:07 2003
@@ -40,6 +40,7 @@
 #include <stdlib.h>
 #include <stddef.h>
 #include <string.h>
+#include <crt_externs.h>
 
 inline char *__findenv(const char *, int *);
 
@@ -57,7 +58,7 @@
 	const char *name;
 	int *offset;
 {
-	extern char **environ;
+	char **environ = *_NSGetEnviron();
 	int len, i;
 	const char *np;
 	char **p, *cp;
