--- Makefile.in.orig	2005-08-21 18:42:14.000000000 -0700
+++ Makefile.in	2006-05-26 16:53:52.000000000 -0700
@@ -2,7 +2,7 @@
 # [need allow_null_glob_expansion=true]
 # The parts behind % are removed when no message catalogs should be installed
 
-all:	Makefile conf_script source manhtml manpages %messages
+all:	Makefile conf_script source manpages %messages
 	@echo; echo 'You can now do "make install"'
 
 # On a machine with time before distribution time
@@ -45,9 +45,8 @@
 %gencat/gencat:
 %	cd gencat; $(MAKE)
 
-install: src/Makefile man/Makefile man2html/Makefile %msgs/Makefile
+install: src/Makefile man/Makefile %msgs/Makefile
 	cd src; $(MAKE) install
-	cd man2html; $(MAKE) install
 	cd man; $(MAKE) installsubdirs
 %	cd msgs; $(MAKE) install
 	@echo; echo 'Done.'
