--- src/man-getopt.c	2007-03-25 21:49:18.000000000 -0700
+++ src/man-getopt.c	2019-04-22 16:02:46.000000000 -0700
@@ -39,7 +39,7 @@
      exit(1);
 }
 
-static char short_opts[] = "B:C:H:xM:P:S:acdDfFhkKm:p:s:tvVwW?";
+static char short_opts[] = "B:C:H:xM:P:S:acdDfFhkKm:p:s:tvVwW?Lq";
 
 #ifndef NOGETOPT
 #undef _GNU_SOURCE
@@ -169,6 +169,11 @@
 		    fatal (INCOMPAT, "-w", "-t");
 	       print_where++;
 	       break;
+	  /* Silently ignore manpath -q and -L (3825529). */
+	  case 'L':
+	  case 'q':
+	       if (!strncmp(progname, "manpath", 7))
+		    break;
 	  case 'h':
 	  case '?':
 	  default:
