--- /tmp/jabberd-2.2.14/configure.ac	2011-05-31 15:11:38.000000000 -0700
+++ ./jabberd2/configure.ac	2011-06-22 20:13:00.000000000 -0700
@@ -615,6 +615,25 @@ fi
 AC_SUBST(SQLITE_LIBS)
 AM_CONDITIONAL(STORAGE_SQLITE, [test "x-$have_sqlite" = "x-yes"])
 
+# Apple Open Directory (Enables OD auth, SQLite for storage only).
+AC_ARG_ENABLE([apple],
+        AS_HELP_STRING([--enable-apple], [enable Apple OD auth, SQLite3 storage support (no)]),
+        [enable_apple=$enableval have_apple=no],
+        [enable_apple=no         have_apple=no])
+if test "x-$enable_sqlite" = "x-yes" ; then
+    AC_CHECK_HEADERS([sqlite3.h], [
+                AC_CHECK_LIB([sqlite3], [sqlite3_open], [
+                        have_apple=yes
+                        AC_DEFINE(STORAGE_APPLE, 1, [Define to 1 if you want to use Apple OD for auth.])
+                ])
+        ])
+        if test "x-$have_apple" != "x-yes" ; then
+                AC_MSG_ERROR([SQLite3 support requested, but headers/libraries not found.])
+        fi
+fi
+AC_SUBST(SQLITE_LIBS)
+AM_CONDITIONAL(STORAGE_SQLITE, [test "x-$have_apple" = "x-yes"])
+
 # Berkeley DB
 _save_libs="$LIBS"
 AC_ARG_ENABLE(db, AC_HELP_STRING([--enable-db], [enable Berkeley DB auth/reg/storage support (no)]),
