--- /tmp/jabberd-2.2.13/config.h.in	2011-02-23 08:24:50.000000000 -0800
+++ ./jabberd2/config.h.in	2011-02-24 16:25:21.000000000 -0800
@@ -585,6 +585,9 @@
 /* Define to 1 if you want to use SQLite 3 for storage. */
 #undef STORAGE_SQLITE
 
+/* Define to 1 if you want to use Apple OD for auth, SQLite3 for storage. */
+#undef STORAGE_APPLE
+
 /* Define to 1 if you want to get SX debug output with -D. */
 #undef SX_DEBUG
 
