--- /tmp/jabberd-2.1.24.1/configure.ac	2008-04-27 02:57:54.000000000 -0700
+++ ./jabberd2/configure.ac	2009-01-29 17:21:47.000000000 -0800
@@ -841,8 +841,8 @@ fi
 #
 # mio backend selection
 
-AC_ARG_ENABLE(mio, AC_HELP_STRING([--enable-mio=BACKENDS], [use one of BACKENDS to drive MIO (select poll epoll)]),
-              mio_check=$enableval, mio_check='epoll poll select')
+AC_ARG_ENABLE(mio, AC_HELP_STRING([--enable-mio=BACKENDS], [use one of BACKENDS to drive MIO (select poll kqueue epoll)]),
+              mio_check=$enableval, mio_check='epoll kqueue poll select')
 
 mio_backend=''
 for backend in $mio_check ; do
@@ -857,6 +857,15 @@ for backend in $mio_check ; do
             fi
             ;;
 
+        x-kqueue)
+            AC_CHECK_HEADERS(sys/event.h)
+            if test "x-$ac_cv_header_sys_event_h" = "x-yes" ; then
+                AC_CHECK_FUNCS(kqueue,[
+                    mio_backend='kqueue'
+                    AC_DEFINE(MIO_KQUEUE,1,[Define to 1 if you want to use 'kqueue' for non-blocking I/O.])])
+            fi
+            ;;
+
         x-poll)
             AC_CHECK_HEADERS(poll.h)
             if test "x-$ac_cv_header_poll_h" = "x-yes" ; then
