--- configure.orig	2006-11-29 16:13:55.000000000 -0800
+++ configure	2006-11-29 16:14:44.000000000 -0800
@@ -9117,13 +9117,13 @@
 
 
     if test x$enable_utf8 != xno; then
-	{ echo "$as_me:$LINENO: checking for get_wch in -lncursesw" >&5
-echo $ECHO_N "checking for get_wch in -lncursesw... $ECHO_C" >&6; }
-if test "${ac_cv_lib_ncursesw_get_wch+set}" = set; then
+	{ echo "$as_me:$LINENO: checking for get_wch in -lncurses" >&5
+echo $ECHO_N "checking for get_wch in -lncurses... $ECHO_C" >&6; }
+if test "${ac_cv_lib_ncurses_get_wch+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lncursesw  $LIBS"
+LIBS="-lncurses  $LIBS"
 cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
 _ACEOF
@@ -9180,22 +9180,22 @@
   ac_status=$?
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); }; }; then
-  ac_cv_lib_ncursesw_get_wch=yes
+  ac_cv_lib_ncurses_get_wch=yes
 else
   echo "$as_me: failed program was:" >&5
 sed 's/^/| /' conftest.$ac_ext >&5
 
-	ac_cv_lib_ncursesw_get_wch=no
+	ac_cv_lib_ncurses_get_wch=no
 fi
 
 rm -f core conftest.err conftest.$ac_objext \
       conftest$ac_exeext conftest.$ac_ext
 LIBS=$ac_check_lib_save_LIBS
 fi
-{ echo "$as_me:$LINENO: result: $ac_cv_lib_ncursesw_get_wch" >&5
-echo "${ECHO_T}$ac_cv_lib_ncursesw_get_wch" >&6; }
-if test $ac_cv_lib_ncursesw_get_wch = yes; then
-  CURSES_LIB="-lncursesw" CURSES_LIB_NAME=ncursesw CURSES_LIB_WIDE=yes
+{ echo "$as_me:$LINENO: result: $ac_cv_lib_ncurses_get_wch" >&5
+echo "${ECHO_T}$ac_cv_lib_ncurses_get_wch" >&6; }
+if test $ac_cv_lib_ncurses_get_wch = yes; then
+  CURSES_LIB="-lncurses" CURSES_LIB_NAME=ncurses CURSES_LIB_WIDE=yes
 fi
 
     fi
