--- cpio/cmdline.c.orig	2009-02-11 16:03:40.000000000 -0800
+++ cpio/cmdline.c	2009-02-11 16:03:44.000000000 -0800
@@ -50,7 +50,7 @@
 /*
  * Short options for cpio.  Please keep this sorted.
  */
-static const char *short_options = "0AaBC:F:O:cdE:f:H:hijLlmopR:rtuvW:yZz";
+static const char *short_options = "0AaBC:F:O:cdE:f:H:hI:ijLlmopR:rtuvW:yZz";
 
 /*
  * Long options for cpio.  Please keep this sorted.
