--- Makefile.in.orig	2009-02-25 20:44:54.000000000 -0800
+++ Makefile.in	2009-03-05 08:56:42.000000000 -0800
@@ -879,7 +879,7 @@
 	cd $(top_srcdir) && \
 	  $(AUTOMAKE) --foreign  Makefile
 .PRECIOUS: Makefile
-Makefile: $(srcdir)/Makefile.in $(top_builddir)/config.status
+Makefile:
 	@case '$?' in \
 	  *config.status*) \
 	    echo ' $(SHELL) ./config.status'; \
@@ -5647,12 +5647,15 @@
 # Building it automatically provides a sanity-check on libarchive_test_SOURCES
 # above.
 libarchive/test/list.h: Makefile
+	mkdir -p libarchive/test
 	cat $(top_srcdir)/libarchive/test/test_*.c | grep DEFINE_TEST > libarchive/test/list.h
 
 tar/test/list.h: Makefile
+	mkdir -p tar/test
 	cat $(top_srcdir)/tar/test/test_*.c | grep DEFINE_TEST > tar/test/list.h
 
 cpio/test/list.h: Makefile
+	mkdir -p cpio/test
 	cat $(top_srcdir)/cpio/test/test_*.c | grep DEFINE_TEST > cpio/test/list.h
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
