diff -ur ../pyOpenSSL-0.7/src/crypto/netscape_spki.c ./src/crypto/netscape_spki.c
--- ../pyOpenSSL-0.7/src/crypto/netscape_spki.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/crypto/netscape_spki.c	2008-07-14 15:59:16.000000000 -0700
@@ -9,6 +9,9 @@
 #define crypto_MODULE
 #include "crypto.h"
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: netscape_spki.c,v 1.1 2004/08/09 13:41:25 martin Exp $";
 
 
diff -ur ../pyOpenSSL-0.7/src/crypto/pkcs12.c ./src/crypto/pkcs12.c
--- ../pyOpenSSL-0.7/src/crypto/pkcs12.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/crypto/pkcs12.c	2008-07-14 15:59:04.000000000 -0700
@@ -14,6 +14,9 @@
 #define crypto_MODULE
 #include "crypto.h"
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: pkcs12.c,v 1.3 2003/01/09 17:08:32 martin Exp $";
 
 /* 
diff -ur ../pyOpenSSL-0.7/src/crypto/pkcs7.c ./src/crypto/pkcs7.c
--- ../pyOpenSSL-0.7/src/crypto/pkcs7.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/crypto/pkcs7.c	2008-07-14 15:58:11.000000000 -0700
@@ -11,6 +11,9 @@
 #define crypto_MODULE
 #include "crypto.h"
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: pkcs7.c,v 1.2 2002/07/09 12:55:13 martin Exp $";
 
 static char crypto_PKCS7_type_is_signed_doc[] = "\n\
diff -ur ../pyOpenSSL-0.7/src/crypto/x509store.c ./src/crypto/x509store.c
--- ../pyOpenSSL-0.7/src/crypto/x509store.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/crypto/x509store.c	2008-07-14 15:57:44.000000000 -0700
@@ -10,6 +10,9 @@
 #define crypto_MODULE
 #include "crypto.h"
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: x509store.c,v 1.9 2002/09/04 22:24:59 iko Exp $";
 
 static char crypto_X509Store_add_cert_doc[] = "\n\
diff -ur ../pyOpenSSL-0.7/src/rand/rand.c ./src/rand/rand.c
--- ../pyOpenSSL-0.7/src/rand/rand.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/rand/rand.c	2008-07-14 16:00:01.000000000 -0700
@@ -25,6 +25,9 @@
 See the file RATIONALE for a short explanation of why this module was written.\n\
 ";
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: rand.c,v 1.10 2002/07/08 11:06:01 martin Exp $";
 
 static char rand_add_doc[] = "\n\
diff -ur ../pyOpenSSL-0.7/src/util.c ./src/util.c
--- ../pyOpenSSL-0.7/src/util.c	2008-03-03 12:16:06.000000000 -0800
+++ ./src/util.c	2008-07-14 15:59:39.000000000 -0700
@@ -11,6 +11,9 @@
 #include <Python.h>
 #include "util.h"
 
+#if defined(__GNUC__) || defined(__INTEL_COMPILER)
+__attribute__((unused))
+#endif
 static char *CVSid = "@(#) $Id: util.c,v 1.5 2001/08/09 11:26:36 martin Exp $";
 
 
