diff -ur ../pyOpenSSL-0.7/src/crypto/crypto.c ./src/crypto/crypto.c
--- ../pyOpenSSL-0.7/src/crypto/crypto.c	2008-03-21 15:34:42.000000000 -0700
+++ ./src/crypto/crypto.c	2008-07-14 15:52:51.000000000 -0700
@@ -592,7 +592,7 @@
     if (py_pkey) {
 	py_pkey->initialized = 0;
     }
-    return py_pkey;
+    return (PyObject *)py_pkey;
 }
 
 static char crypto_X509Extension_doc[] = "\n\
diff -ur ../pyOpenSSL-0.7/src/crypto/x509.c ./src/crypto/x509.c
--- ../pyOpenSSL-0.7/src/crypto/x509.c	2008-03-25 12:23:37.000000000 -0700
+++ ./src/crypto/x509.c	2008-07-14 15:54:53.000000000 -0700
@@ -323,7 +323,7 @@
     if (py_pkey != NULL) {
 	py_pkey->only_public = 1;
     }
-    return py_pkey;
+    return (PyObject *)py_pkey;
 }
 
 static char crypto_X509_set_pubkey_doc[] = "\n\
diff -ur ../pyOpenSSL-0.7/src/crypto/x509req.c ./src/crypto/x509req.c
--- ../pyOpenSSL-0.7/src/crypto/x509req.c	2008-03-21 15:34:42.000000000 -0700
+++ ./src/crypto/x509req.c	2008-07-14 15:58:48.000000000 -0700
@@ -70,7 +70,7 @@
     if (py_pkey != NULL) {
 	py_pkey->only_public = 1;
     }
-    return py_pkey;
+    return (PyObject *)py_pkey;
 }
 
 static char crypto_X509Req_set_pubkey_doc[] = "\n\
