--- ltconfig.orig	Fri Aug  6 01:00:26 2004
+++ ltconfig	Fri Aug  6 01:00:45 2004
@@ -624,8 +624,8 @@
 old_postuninstall_cmds=
 
 if test -n "$RANLIB"; then
-  old_archive_cmds="$old_archive_cmds~\$RANLIB \$oldlib"
-  old_postinstall_cmds="\$RANLIB \$oldlib~$old_postinstall_cmds"
+  old_archive_cmds="$old_archive_cmds:\$RANLIB \$oldlib"
+  old_postinstall_cmds="\$RANLIB \$oldlib:$old_postinstall_cmds"
 fi
 
 # Source the script associated with the $tagname tag configuration.
