--- magic/Magdir/cafebabe.orig	2009-02-04 09:41:10.000000000 -0800
+++ magic/Magdir/cafebabe	2009-02-04 09:41:33.000000000 -0800
@@ -1,9 +1,9 @@
 #------------------------------------------------------------------------------
 # Cafe Babes unite!
 #
-# Since Java bytecode and Mach-O fat-files have the same magic number, the test
+# Since Java bytecode and Mach-O universal binaries have the same magic number, the test
 # must be performed in the same "magic" sequence to get both right.  The long
-# at offset 4 in a mach-O fat file tells the number of architectures; the short at
+# at offset 4 in a mach-O universal binary tells the number of architectures; the short at
 # offset 4 in a Java bytecode file is the JVM minor version and the
 # short at offset 6 is the JVM major version.  Since there are only 
 # only 18 labeled Mach-O architectures at current, and the first released 
@@ -19,9 +19,9 @@
 >>4	beshort		x       	\b%d
 
 0	belong		0xcafebabe
->4	belong		1		Mach-O fat file with 1 architecture
+>4	belong		1		Mach-O universal binary with 1 architecture
 >4	belong		>1
->>4	belong		<20		Mach-O fat file with %ld architectures
+>>4	belong		<20		Mach-O universal binary with %ld architectures
 
 0	belong		0xcafed00d	JAR compressed with pack200,
 >>5	byte		x		version %d.
--- magic/Magdir/mach.orig	2008-01-11 17:36:28.000000000 -0800
+++ magic/Magdir/mach	2008-01-11 17:38:17.000000000 -0800
@@ -15,7 +15,9 @@
 >12	lelong		7		dynamic linker
 >12	lelong		8		bundle
 >12	lelong		9		dynamically linked shared library stub
->12	lelong		>9
+>12	lelong		10		dSYM companion file
+>12	lelong		11		kext bundle
+>12	lelong		>11
 >>12	lelong		x		filetype=%ld
 >4	lelong		<0
 >>4	lelong		x		architecture=%ld
@@ -28,15 +30,16 @@
 >4	lelong		7		i386
 >4	lelong		8		mips
 >4	lelong		9		ns32532
->4	lelong		10		architecture=10
+>4	lelong		10		m98k
 >4	lelong		11		hppa
->4	lelong		12		acorn
+>4	lelong		12		arm
 >4	lelong		13		m88k
 >4	lelong		14		sparc
->4	lelong		15		i860-big
->4	lelong		16		i860
+>4	lelong		15		i860
+>4	lelong		16		alpha
 >4	lelong		17		rs6000
 >4	lelong		18		ppc
+>4	lelong		16777223	x86_64
 >4	lelong		16777234	ppc64
 >4	lelong		>16777234
 >>4	lelong		x		architecture=%ld
@@ -52,7 +55,9 @@
 >12	belong		7               dynamic linker
 >12	belong		8		bundle
 >12	belong		9		dynamically linked shared library stub
->12	belong		>9
+>12	belong		10		dSYM companion file
+>12	belong		11		kext bundle
+>12	belong		>11
 >>12	belong		x		filetype=%ld
 >4	belong		<0
 >>4	belong		x		architecture=%ld
@@ -70,15 +75,16 @@
 >4	belong		7		i386
 >4	belong		8		mips
 >4	belong		9		ns32532
->4	belong		10		architecture=10
+>4	belong		10		m98k
 >4	belong		11		hppa
->4	belong		12		acorn
+>4	belong		12		arm
 >4	belong		13		m88k
 >4	belong		14		sparc
->4	belong		15		i860-big
->4	belong		16		i860
+>4	belong		15		i860
+>4	belong		16		alpha
 >4	belong		17		rs6000
 >4	belong		18		ppc
+>4	belong		16777223	x86_64
 >4	belong		16777234	ppc64
 >4	belong		>16777234
 >>4	belong		x		architecture=%ld
