--- magic/Magdir/sun.orig	2004-08-06 14:23:08.000000000 -0700
+++ magic/Magdir/sun	2004-08-06 14:37:51.000000000 -0700
@@ -113,7 +113,7 @@
 # Microsoft ICM color profile
 36	string		acspMSFT	Microsoft ICM Color Profile
 # Sun KCMS
-36	string		acsp		Kodak Color Management System, ICC Profile
+36	string		acsp		ColorSync ICC Profile
 
 #---------------------------------------------------------------------------
 # The following entries have been tested by Duncan Laurie <duncan@sun.com> (a
