--- magic/Magdir/gnu.orig	2006-08-29 07:06:35.000000000 -0700
+++ magic/Magdir/gnu	2006-08-29 07:06:41.000000000 -0700
@@ -11,7 +11,7 @@
 >8	belong		x		%d messages
 # message catalogs, from Mitchum DSouza <m.dsouza@mrc-apu.cam.ac.uk>
 0	string		*nazgul*	Nazgul style compiled message catalog
->8	lelong		>0		\b, version %ld
+>8	belong		>0		\b, version %ld
 
 # GnuPG
 # The format is very similar to pgp
