--- magic/Makefile.in.orig	Fri Aug  6 01:46:01 2004
+++ magic/Makefile.in	Fri Aug  6 02:04:50 2004
@@ -493,7 +493,8 @@
 	$(top_builddir)/src/file -C -m magic
 
 magic.mime.mgc: magic.mime
-	$(top_builddir)/src/file -C -m $(srcdir)/magic.mime
+	cat $(srcdir)/magic.mime > magic.mime
+	$(top_builddir)/src/file -C -m magic.mime
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
