--- magic/Magdir/mach	2005-07-06 06:06:52.000000000 -0700
+++ magic/Magdir/mach	2007-03-14 00:46:47.000000000 -0700
@@ -1,9 +1,9 @@
 #------------------------------------------------------------------------------
 # mach file description
 #
-# Since Java bytecode and Mach-O fat-files have the same magic number the test
+# Since Java bytecode and Mach-O universal binaries have the same magic number the test
 # must be preformed in the same "magic" sequence to get both right.  The long
-# at offset 4 in a fat file tells the number of architectures.  The short at
+# at offset 4 in a universal binary tells the number of architectures.  The short at
 # offset 4 in a Java bytecode file is the compiler minor version and the
 # short at offset 6 is the compiler major version.  Since there are only 
 # only 18 labeled Mach-O architectures at current, and the first released 
@@ -15,13 +15,13 @@
 >4	belong		>19		compiled Java class data,
 >>6     beshort		x	        version %d.
 >>4     beshort		x       	\b%d
->4	belong		1		Mach-O fat file with 1 architecture
+>4	belong		1		Mach-O universal binary with 1 architecture
 >4	belong		>1
 # The following is necessary to support java class files.
->>4	belong		<20		Mach-O fat file with %ld architectures
-#>>4	belong		<0xffff		Mach-O fat file with %ld architectures
+>>4	belong		<20		Mach-O universal binary with %ld architectures
+#>>4	belong		<0xffff		Mach-O universal binary with %ld architectures
 #
-0	lelong&0xfeffffff	0xfeedface	Mach-O
+0	lelong&0xfffffffe	0xfeedface	Mach-O
 >0	byte		0xcf		64-bit
 >12	lelong		1		object
 >12	lelong		2		executable
@@ -45,15 +45,16 @@
 >4	lelong		7		i386
 >4	lelong		8		mips
 >4	lelong		9		ns32532
->4	lelong		10		architecture=10
+>4	lelong		10		m98k
 >4	lelong		11		hppa
->4	lelong		12		acorn
+>4	lelong		12		arm
 >4	lelong		13		m88k
 >4	lelong		14		sparc
->4	lelong		15		i860-big
->4	lelong		16		i860
+>4	lelong		15		i860
+>4	lelong		16		alpha
 >4	lelong		17		rs6000
 >4	lelong		18		ppc
+>4	lelong		16777223	x86_64
 >4	lelong		16777234	ppc64
 >4	lelong		>16777234
 >>4	lelong		x		architecture=%ld
@@ -87,15 +88,16 @@
 >4	belong		7		i386
 >4	belong		8		mips
 >4	belong		9		ns32532
->4	belong		10		architecture=10
+>4	belong		10		m98k
 >4	belong		11		hppa
->4	belong		12		acorn
+>4	belong		12		arm
 >4	belong		13		m88k
 >4	belong		14		sparc
->4	belong		15		i860-big
->4	belong		16		i860
+>4	belong		15		i860
+>4	belong		16		alpha
 >4	belong		17		rs6000
 >4	belong		18		ppc
+>4	belong		16777223	x86_64
 >4	belong		16777234	ppc64
 >4	belong		>16777234
 >>4	belong		x		architecture=%ld
