--- magic/Makefile.in.orig	2005-08-18 08:52:10.000000000 -0700
+++ magic/Makefile.in	2005-09-28 17:50:35.000000000 -0700
@@ -201,6 +201,7 @@
 Magdir/claris \
 Magdir/clipper \
 Magdir/cracklib \
+Magdir/misctools \
 Magdir/spec \
 Magdir/commands \
 Magdir/communications \
@@ -264,7 +265,6 @@
 Magdir/mime \
 Magdir/mips \
 Magdir/mirage \
-Magdir/misctools \
 Magdir/mkid \
 Magdir/mlssa \
 Magdir/mmdf \
@@ -551,7 +551,8 @@
 	$(FILE_COMPILE) -C -m magic
 
 magic.mime.mgc: magic.mime
-	$(FILE_COMPILE) -C -m $(srcdir)/magic.mime
+	cp $(srcdir)/magic.mime magic.mime
+	$(FILE_COMPILE) -C -m magic.mime
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
