--- magic/Magdir/commands.orig	2007-02-02 19:44:32.000000000 -0800
+++ magic/Magdir/commands	2007-02-02 19:46:00.000000000 -0800
@@ -27,8 +27,7 @@
 #
 0	string/b	#!\ /bin/awk		awk script text executable
 0	string/b	#!\ /usr/bin/awk	awk script text executable
-# update to distinguish from *.vcf files
-0	regex		BEGIN[[:space:]]*[{]	awk script text
+0	string		BEGIN			awk script text
 
 # AT&T Bell Labs' Plan 9 shell
 0	string/b	#!\ /bin/rc	Plan 9 rc shell script text executable
