--- magic/Makefile.in.orig	2007-02-02 19:26:42.000000000 -0800
+++ magic/Makefile.in	2007-02-02 19:26:48.000000000 -0800
@@ -216,6 +216,7 @@
 Magdir/dolby \
 Magdir/dump \
 Magdir/editors \
+Magdir/efi \
 Magdir/elf \
 Magdir/encore \
 Magdir/epoc \
--- /dev/null	2007-02-02 19:22:40.000000000 -0800
+++ magic/Magdir/efi	2007-02-02 19:25:43.000000000 -0800
@@ -0,0 +1,11 @@
+# Universal EFI binary
+0		lelong		0x0ef1fab9
+>4		lelong		1				Universal EFI binary with 1 architecture
+>>&0	lelong		7				\b, i386
+>>&0	lelong		0x01000007		\b, x86_64
+>4		lelong		2				Universal EFI binary with 2 architectures
+>>&0	lelong		7				\b, i386
+>>&0	lelong		0x01000007		\b, x86_64
+>>&20	lelong		7				\b, i386
+>>&20	lelong		0x01000007		\b, x86_64
+>4		lelong		>2				Universal EFI binary with %ld architectures
