--- magic/Magdir/apple.orig	2007-05-30 15:05:17.000000000 -0700
+++ magic/Magdir/apple	2007-05-30 15:05:25.000000000 -0700
@@ -189,3 +189,29 @@
 >0		byte		<5		\b
 >>13	byte		0x81	\b
 >>>14	uleshort	x		\b, system %hd
+
+#------------------------------------------------------------------------------
+# Keychain database files
+0       string          kych            Mac OS X Keychain File
+
+#------------------------------------------------------------------------------
+# Code Signing related file types
+0       belong          0xfade0c00      Mac OS X Code Requirement
+>8      belong          1                       (opExpr)
+>4      belong          x                       - %d bytes
+
+0       belong          0xfade0c01      Mac OS X Code Requirement Set
+>8      belong          >1                      containing %d items
+>4      belong          x                       - %d bytes
+
+0       belong          0xfade0c02      Mac OS X Code Directory
+>8      belong          x                       version %x
+>12     belong          >0                      flags 0x%x
+>4      belong          x                       - %d bytes
+
+0       belong          0xfade0cc0      Mac OS X Detached Code Signature (non-executable)
+>4      belong          x                       - %d bytes
+
+0       belong          0xfade0cc1      Mac OS X Detached Code Signature
+>8      belong          >1                      (%d elements)
+>4      belong          x                       - %d bytes
--- magic/Magdir/mips.orig	2007-05-30 16:04:23.000000000 -0700
+++ magic/Magdir/mips	2007-05-30 16:04:36.000000000 -0700
@@ -173,5 +173,3 @@
 4	belong	0x0000007d		GLF_BINARY_MSB_FIRST
 # GLS is OpenGL stream encoding; GLS is the successor of GLF
 0	string	glsBeginGLS(		GLS_TEXT
-4	belong	0x10000000		GLS_BINARY_LSB_FIRST
-4	belong	0x00000010		GLS_BINARY_MSB_FIRST
