--- src/file.h.orig	2005-09-01 12:12:38.000000000 -0700
+++ src/file.h	2005-09-01 12:13:06.000000000 -0700
@@ -70,11 +70,11 @@
 #endif
 #define MAXMAGIS 8192		/* max entries in /etc/magic */
 #define MAXDESC	64		/* max leng of text description */
-#define MAXstring 32		/* max leng of "string" types */
+#define MAXstring 256		/* max leng of "string" types */
 
 #define MAGICNO		0xF11E041C
 #define VERSIONNO	2
-#define FILE_MAGICSIZE	(32 * 4)
+#define FILE_MAGICSIZE	(32 * 11)
 
 #define	FILE_LOAD	0
 #define FILE_CHECK	1
@@ -265,6 +265,7 @@
 protected int file_printf(struct magic_set *, const char *, ...);
 protected int file_reset(struct magic_set *);
 protected int file_tryelf(struct magic_set *, int, const unsigned char *, size_t);
+protected int file_trymacho(struct magic_set *, int, const unsigned char *, size_t, const char *);
 protected int file_zmagic(struct magic_set *, int, const unsigned char *, size_t);
 protected int file_ascmagic(struct magic_set *, const unsigned char *, size_t);
 protected int file_is_tar(struct magic_set *, const unsigned char *, size_t);
