diff -Naur ../tcpdump-2005.06.16/Makefile.in ./Makefile.in
--- ../tcpdump-2005.06.16/Makefile.in	2005-05-20 14:02:29.000000000 -0700
+++ ./Makefile.in	2005-06-17 15:28:35.000000000 -0700
@@ -156,7 +156,7 @@
 	@rm -f $@
 	sed -e 's/.*/char version[] = "&";/' $(srcdir)/VERSION > $@
 
-install:
+install: tcpdump.o
 	[ -d $(DESTDIR)$(sbindir) ] || \
 	    (mkdir -p $(DESTDIR)$(sbindir); chmod 755 $(DESTDIR)$(sbindir))
 	$(INSTALL_PROGRAM) $(PROG) $(DESTDIR)$(sbindir)/$(PROG)
