Index: Makefile.in
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/Makefile.in,v
retrieving revision 1.15
diff -u -p -u -p -r1.15 Makefile.in
--- Makefile.in	13 Aug 2001 15:57:43 -0000	1.15
+++ Makefile.in	11 May 2004 01:17:06 -0000
@@ -102,6 +102,16 @@ install-libraries:
 		fi \
 	done;
 
+install-doc:
+	for dir in $(subdirs) ; do \
+		if test -d $$dir ; then \
+		    echo "$$dir: $(MAKE) $@" ;\
+		    if (cd $$dir; $(MAKE) $@) ; then true ; else exit 1 ; fi ; \
+		else \
+		    exit 1; \
+		fi \
+	done;
+
 clean:
 	for dir in $(subdirs) ; do \
 		if test -d $$dir ; then \
Index: tcl.m4
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/tcl.m4,v
retrieving revision 1.24
diff -u -p -u -p -r1.24 tcl.m4
--- tcl.m4	4 Apr 2003 20:57:42 -0000	1.24
+++ tcl.m4	11 May 2004 01:17:06 -0000
@@ -2591,6 +2591,7 @@ AC_DEFUN(TEA_MAKE_LIB, [
     # substituted.
     #--------------------------------------------------------------------
 
+    RANLIB_STUB="${RANLIB}"
     if test "${TEA_PLATFORM}" = "windows" ; then
 	if test "${SHARED_BUILD}" = "1" ; then
 	    # We force the unresolved linking of symbols that are really in
@@ -2632,6 +2633,7 @@ AC_DEFUN(TEA_MAKE_LIB, [
     AC_SUBST(MAKE_SHARED_LIB)
     AC_SUBST(MAKE_STATIC_LIB)
     AC_SUBST(MAKE_STUB_LIB)
+    AC_SUBST(RANLIB_STUB)
 ])
 
 #------------------------------------------------------------------------
Index: itcl/Makefile.in
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itcl/Makefile.in,v
retrieving revision 1.21
diff -u -p -u -p -r1.21 Makefile.in
--- itcl/Makefile.in	29 Apr 2004 17:29:11 -0000	1.21
+++ itcl/Makefile.in	11 May 2004 01:17:06 -0000
@@ -155,6 +155,7 @@ MAKE_STATIC_LIB	= @MAKE_STATIC_LIB@
 MAKE_STUB_LIB	= @MAKE_STUB_LIB@
 OBJEXT		= @OBJEXT@
 RANLIB		= @RANLIB@
+RANLIB_STUB	= @RANLIB_STUB@
 SHLIB_CFLAGS	= @SHLIB_CFLAGS@
 SHLIB_LD	= @SHLIB_LD@
 SHLIB_LDFLAGS	= @SHLIB_LDFLAGS@
@@ -179,6 +180,7 @@ EXTRA_PATH	= $(top_builddir):$(TCL_BIN_D
 TCLSH_ENV	= TCL_LIBRARY=`@CYGPATH@ $(TCL_SRC_DIR)/library` \
 		  ITCL_LIBRARY=`@CYGPATH@ $(srcdir)/library` \
 		  LD_LIBRARY_PATH="$(EXTRA_PATH):$(LD_LIBRARY_PATH)" \
+		  DYLD_LIBRARY_PATH="$(EXTRA_PATH):$(DYLD_LIBRARY_PATH)" \
 		  LIBPATH="$(EXTRA_PATH):${LIBPATH}" \
 		  SHLIB_PATH="$(EXTRA_PATH):${SHLIB_PATH}" \
 		  PATH="$(EXTRA_PATH):$(PATH)" \
@@ -298,7 +300,7 @@ $($(PACKAGE)_LIB_FILE): $($(PACKAGE)_OBJ
 $($(PACKAGE)stub_LIB_FILE): $($(PACKAGE)stub_OBJECTS)
 	-rm -f $($(PACKAGE)stub_LIB_FILE)
 	${MAKE_STUB_LIB}
-	$(RANLIB) $($(PACKAGE)stub_LIB_FILE)
+	$(RANLIB_STUB) $($(PACKAGE)stub_LIB_FILE)
 
 #========================================================================
 # We need to enumerate the list of .c to .o lines here.
@@ -367,8 +369,14 @@ install-lib-binaries:
 	  if test -f $$p; then \
 	    echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p; \
+	    stub=`echo $$p|sed -e "s/.*\(stub\).*/\1/"`; \
+	    if test "x$$stub" = "xstub"; then \
+	    echo " $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p"; \
+	    $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    else \
 	    echo " $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    fi; \
 	    ext=`echo $$p|sed -e "s/.*\.//"`; \
 	    if test "x$$ext" = "xdll"; then \
 		lib=`basename $$p|sed -e 's/.[^.]*$$//'`.lib; \
Index: itcl/configure
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itcl/configure,v
retrieving revision 1.37
diff -u -p -u -p -r1.37 configure
--- itcl/configure	4 Apr 2003 20:57:42 -0000	1.37
+++ itcl/configure	11 May 2004 01:17:06 -0000
@@ -3939,6 +3939,7 @@ fi
     # substituted.
     #--------------------------------------------------------------------
 
+    RANLIB_STUB="${RANLIB}"
     if test "${TEA_PLATFORM}" = "windows" ; then
 	if test "${SHARED_BUILD}" = "1" ; then
 	    # We force the unresolved linking of symbols that are really in
@@ -3980,6 +3981,7 @@ fi
     
     
     
+    
 
 
 #--------------------------------------------------------------------
@@ -4002,7 +4004,7 @@ fi
 
 
     echo $ac_n "checking for tclsh""... $ac_c" 1>&6
-echo "configure:4006: checking for tclsh" >&5
+echo "configure:4008: checking for tclsh" >&5
 
     if eval "test \"`echo '$''{'ac_cv_path_tclsh'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -4280,6 +4282,7 @@ s%@MAKE_LIB@%$MAKE_LIB%g
 s%@MAKE_SHARED_LIB@%$MAKE_SHARED_LIB%g
 s%@MAKE_STATIC_LIB@%$MAKE_STATIC_LIB%g
 s%@MAKE_STUB_LIB@%$MAKE_STUB_LIB%g
+s%@RANLIB_STUB@%$RANLIB_STUB%g
 s%@itclstub_LIB_FILE@%$itclstub_LIB_FILE%g
 s%@itcl_LIB_FILE@%$itcl_LIB_FILE%g
 s%@TCLSH_PROG@%$TCLSH_PROG%g
Index: itcl/generic/itcl_cmds.c
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itcl/generic/itcl_cmds.c,v
retrieving revision 1.23
diff -u -p -u -p -r1.23 itcl_cmds.c
--- itcl/generic/itcl_cmds.c	23 Dec 2003 06:58:27 -0000	1.23
+++ itcl/generic/itcl_cmds.c	11 May 2004 01:17:06 -0000
@@ -69,6 +69,13 @@ namespace eval ::itcl {\n\
             lappend dirs [file join $bindir .. .. library]\n\
             lappend dirs [file join $bindir .. .. itcl library]\n\
             lappend dirs [file join $bindir .. .. .. itcl library]\n\
+            # On MacOSX, check the directories in the tcl_pkgPath\n\
+            if {[string equal $::tcl_platform(platform) \"unix\"] && \
+                    [string equal $::tcl_platform(os) \"Darwin\"]} {\n\
+                foreach d $::tcl_pkgPath {\n\
+                    lappend dirs [file join $d itcl$version]\n\
+                }\n\
+            }\n\
         }\n\
         foreach i $dirs {\n\
             set library $i\n\
Index: itk/Makefile.in
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itk/Makefile.in,v
retrieving revision 1.22
diff -u -p -u -p -r1.22 Makefile.in
--- itk/Makefile.in	5 Apr 2002 17:26:05 -0000	1.22
+++ itk/Makefile.in	11 May 2004 01:17:06 -0000
@@ -142,6 +142,7 @@ MAKE_STATIC_LIB	= @MAKE_STATIC_LIB@
 MAKE_STUB_LIB	= @MAKE_STUB_LIB@
 OBJEXT		= @OBJEXT@
 RANLIB		= @RANLIB@
+RANLIB_STUB	= @RANLIB_STUB@
 SHLIB_CFLAGS	= @SHLIB_CFLAGS@
 SHLIB_LD	= @SHLIB_LD@
 SHLIB_LDFLAGS	= @SHLIB_LDFLAGS@
@@ -175,6 +176,7 @@ TCLSH_ENV	= TCL_LIBRARY=`@CYGPATH@ $(TCL
 		  ITCL_LIBRARY=`@CYGPATH@ @itcl_SRC_DIR@/library` \
 		  ITK_LIBRARY=`@CYGPATH@ $(srcdir)/library` \
 		  LD_LIBRARY_PATH="$(EXTRA_PATH):$(LD_LIBRARY_PATH)" \
+		  DYLD_LIBRARY_PATH="$(EXTRA_PATH):$(DYLD_LIBRARY_PATH)" \
 		  LIBPATH="$(EXTRA_PATH):${LIBPATH}" \
 		  SHLIB_PATH="$(EXTRA_PATH):${SHLIB_PATH}" \
 		  PATH="$(EXTRA_PATH):$(PATH)" \
@@ -297,7 +299,7 @@ $($(PACKAGE)_LIB_FILE): $($(PACKAGE)_OBJ
 $($(PACKAGE)stub_LIB_FILE): $($(PACKAGE)stub_OBJECTS)
 	-rm -f $($(PACKAGE)stub_LIB_FILE)
 	${MAKE_STUB_LIB}
-	$(RANLIB) $($(PACKAGE)stub_LIB_FILE)
+	$(RANLIB_STUB) $($(PACKAGE)stub_LIB_FILE)
 
 #========================================================================
 # We need to enumerate the list of .c to .o lines here.
@@ -360,8 +362,14 @@ install-lib-binaries:
 	  if test -f $$p; then \
 	    echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p; \
+	    stub=`echo $$p|sed -e "s/.*\(stub\).*/\1/"`; \
+	    if test "x$$stub" = "xstub"; then \
+	    echo " $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p"; \
+	    $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    else \
 	    echo " $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    fi; \
 	    ext=`echo $$p|sed -e "s/.*\.//"`; \
 	    if test "x$$ext" = "xdll"; then \
 		lib=`basename $$p|sed -e 's/.[^.]*$$//'`.lib; \
Index: itk/configure
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itk/configure,v
retrieving revision 1.41
diff -u -p -u -p -r1.41 configure
--- itk/configure	4 Apr 2003 20:57:43 -0000	1.41
+++ itk/configure	11 May 2004 01:17:06 -0000
@@ -4271,6 +4271,7 @@ fi
     # substituted.
     #--------------------------------------------------------------------
 
+    RANLIB_STUB="${RANLIB}"
     if test "${TEA_PLATFORM}" = "windows" ; then
 	if test "${SHARED_BUILD}" = "1" ; then
 	    # We force the unresolved linking of symbols that are really in
@@ -4312,6 +4313,7 @@ fi
     
     
     
+    
 
 
 #--------------------------------------------------------------------
@@ -4345,7 +4347,7 @@ fi
 
 
     echo $ac_n "checking for tclsh""... $ac_c" 1>&6
-echo "configure:4349: checking for tclsh" >&5
+echo "configure:4351: checking for tclsh" >&5
 
     if eval "test \"`echo '$''{'ac_cv_path_tclsh'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -4382,7 +4384,7 @@ fi
 
 
     echo $ac_n "checking for wish""... $ac_c" 1>&6
-echo "configure:4386: checking for wish" >&5
+echo "configure:4388: checking for wish" >&5
 
     if eval "test \"`echo '$''{'ac_cv_path_wish'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
@@ -4651,6 +4653,7 @@ s%@MAKE_LIB@%$MAKE_LIB%g
 s%@MAKE_SHARED_LIB@%$MAKE_SHARED_LIB%g
 s%@MAKE_STATIC_LIB@%$MAKE_STATIC_LIB%g
 s%@MAKE_STUB_LIB@%$MAKE_STUB_LIB%g
+s%@RANLIB_STUB@%$RANLIB_STUB%g
 s%@itkstub_LIB_FILE@%$itkstub_LIB_FILE%g
 s%@itk_LIB_FILE@%$itk_LIB_FILE%g
 s%@TCLSH_PROG@%$TCLSH_PROG%g
Index: itk/generic/itk_cmds.c
===================================================================
RCS file: /cvsroot/incrtcl/incrTcl/itk/generic/itk_cmds.c,v
retrieving revision 1.15
diff -u -p -u -p -r1.15 itk_cmds.c
--- itk/generic/itk_cmds.c	23 Dec 2003 06:57:54 -0000	1.15
+++ itk/generic/itk_cmds.c	11 May 2004 01:17:06 -0000
@@ -73,6 +73,13 @@ namespace eval ::itk {\n\
             lappend dirs [file join $bindir .. library]\n\
             lappend dirs [file join $bindir .. .. library]\n\
             lappend dirs [file join $bindir .. .. itk library]\n\
+            # On MacOSX, check the directories in the tcl_pkgPath\n\
+            if {[string equal $::tcl_platform(platform) \"unix\"] && \
+                    [string equal $::tcl_platform(os) \"Darwin\"]} {\n\
+                foreach d $::tcl_pkgPath {\n\
+                    lappend dirs [file join $d itk$version]\n\
+                }\n\
+            }\n\
         }\n\
         foreach i $dirs {\n\
             set library $i\n\
diff -rupN orig/incrTcl/config/installFile.tcl incrTcl/config/installFile.tcl
--- config/installFile.tcl	Thu Jan  1 10:00:00 1970
+++ config/installFile.tcl	Wed Oct 31 12:08:54 2001
@@ -0,0 +1,131 @@
+#!/bin/sh
+#
+# installFile.tcl - a Tcl version of install-sh
+#	that copies a file and preserves its permission bits.
+#	This also optimizes out installation of existing files
+#	that have the same size and time stamp as the source.
+#
+# \
+exec tclsh8.3 "$0" ${1+"$@"}
+
+set doCopy 0	;# Rename files instead of copy
+set doStrip 0	;# Strip the symbols from installed copy
+set verbose 0
+set src ""
+set dst ""
+
+# Process command line arguments, compatible with install-sh
+
+for {set i 0} {$i < $argc} {incr i} {
+    set arg [lindex $argv $i]
+    switch -- $arg {
+	-c {
+	    set doCopy 1
+	}
+	-m  {
+	    incr i
+	    # Assume UNIX standard "644", etc, so force Tcl to think octal
+	    set permissions 0[lindex $argv $i]
+	}
+	-o  {
+	    incr i
+	    set owner [lindex $argv $i]
+	}
+	-g  {
+	    incr i
+	    set group [lindex $argv $i]
+	}
+	-s {
+	    set doStrip 1
+	}
+	-v {
+	    set verbose 1
+	}
+	default {
+	    set src $arg
+	    incr i
+	    set dst [lindex $argv $i]
+	    break
+	}
+    }
+}
+if {[string length $src] == 0} {
+    puts stderr "$argv0: no input file specified"
+    exit 1
+}
+if {[string length $dst] == 0} {
+    puts stderr "$argv0: no destination file specified"
+    exit 1
+}
+
+# Compatibility with CYGNUS-style pathnames
+regsub {^/(cygdrive)?/(.)/(.*)} $src {\2:/\3} src
+regsub {^/(cygdrive)?/(.)/(.*)} $dst {\2:/\3} dst
+
+if {$verbose && $doStrip} {
+    puts stderr "Ignoring -s (strip) option for $dst"
+}
+if {[file isdirectory $dst]} {
+    set dst [file join $dst [file tail $src]]
+}
+
+# Temporary file name
+
+set dsttmp [file join [file dirname $dst] #inst.[pid]#]
+
+# Optimize out install if the file already exists
+
+set actions ""
+if {[file exists $dst] &&
+	([file mtime $src] == [file mtime $dst]) &&
+	([file size $src] == [file size $dst])} {
+
+    # Looks like the same file, so don't bother to copy.
+    # Set dsttmp in case we still need to tweak mode, group, etc.
+
+    set dsttmp $dst
+    lappend actions "already installed"
+} else {
+    if {"[file type $src]" == "link"} {
+	# Perfom a true copy.
+	set in  [open $src r]
+	set out [open $dsttmp w]
+	fcopy $in $out
+	close $in
+	close $out
+    } else {
+	file copy -force $src $dsttmp
+    }
+    lappend actions copied
+}
+
+# update the modification time of the target file
+file mtime $dsttmp [clock seconds]
+
+# At this point "$dsttmp" is installed, but might not have the
+# right permissions and may need to be renamed.
+
+
+foreach attrName {owner group permissions} {
+    upvar 0 $attrName attr
+
+    if {[info exists attr]} {
+	if {![catch {file attributes $dsttmp -$attrName} dstattr]} {
+
+	    # This system supports "$attrName" kind of attributes
+
+	    if {($attr != $dstattr)} {
+		file attributes $dsttmp -$attrName $attr
+		lappend actions "set $attrName to $attr"
+	    }
+	}
+    }
+}
+
+if {[string compare $dst $dsttmp] != 0} {
+    file rename -force $dsttmp $dst
+}
+if {$verbose} {
+    puts stderr "$dst: [join $actions ", "]"
+}
+exit 0
diff -rupN --exclude=CVS --exclude=.cvsignore orig/incrTcl/config/mkinstalldirs incrTcl/config/mkinstalldirs
--- config/mkinstalldirs	Thu Jan  1 10:00:00 1970
+++ config/mkinstalldirs	Wed Jul 21 23:58:15 2004
@@ -0,0 +1,40 @@
+#! /bin/sh
+# mkinstalldirs --- make directory hierarchy
+# Author: Noah Friedman <friedman@prep.ai.mit.edu>
+# Created: 1993-05-16
+# Public domain
+
+# $Id: incrTcl.diff,v 1.3 2006/04/17 20:47:04 bbyer Exp $
+
+errstatus=0
+
+for file
+do
+   set fnord `echo ":$file" | sed -ne 's/^:\//#/;s/^://;s/\// /g;s/^#/\//;p'`
+   shift
+
+   pathcomp=
+   for d
+   do
+     pathcomp="$pathcomp$d"
+     case "$pathcomp" in
+       -* ) pathcomp=./$pathcomp ;;
+     esac
+
+     if test ! -d "$pathcomp"; then
+        echo "mkdir $pathcomp"
+
+        mkdir "$pathcomp" || lasterr=$?
+
+        if test ! -d "$pathcomp"; then
+  	  errstatus=$lasterr
+        fi
+     fi
+
+     pathcomp="$pathcomp/"
+   done
+done
+
+exit $errstatus
+
+# mkinstalldirs ends here
diff -rupN --exclude=CVS --exclude=.cvsignore orig/incrTcl/tcl.m4 incrTcl/tcl.m4
--- tcl.m4	2003-04-04 12:57:42.000000000 -0800
+++ tcl.m4	2006-04-17 13:23:51.000000000 -0700
@@ -2575,7 +2575,7 @@
 	MAKE_STUB_LIB="\${STLIB_LD} -out:\[$]@ \$(\[$](PACKAGE)stub_OBJECTS)"
     else
 	MAKE_STATIC_LIB="\${STLIB_LD} \[$]@ \$(\[$](PACKAGE)_OBJECTS)"
-	MAKE_SHARED_LIB="\${SHLIB_LD} -o \[$]@ \$(\[$](PACKAGE)_OBJECTS) \${SHLIB_LDFLAGS} \${SHLIB_LD_LIBS}"
+	MAKE_SHARED_LIB="\${SHLIB_LD} -o \[$]@ \$(\[$](PACKAGE)_OBJECTS) \${SHLIB_LDFLAGS} \${SHLIB_LD_LIBS} \${TCL_SHLIB_LD_EXTRAS}"
 	MAKE_STUB_LIB="\${STLIB_LD} \[$]@ \$(\[$](PACKAGE)stub_OBJECTS)"
     fi
 
--- itk/Makefile.in.old	2006-07-10 18:20:36.000000000 -0700
+++ itk/Makefile.in	2006-07-10 18:21:48.000000000 -0700
@@ -259,6 +259,7 @@
 
 install-doc: doc
 	@mkdir -p $(DESTDIR)$(mandir)/mann
+	mv $(srcdir)/doc/Widget.n $(srcdir)/doc/itkWidget.n
 	@echo "Installing man pages in $(DESTDIR)$(mandir)"
 	@for i in $(srcdir)/doc/*.n; do \
 	    echo "Installing $$i"; \
--- itcl/configure~	2003-04-04 12:57:42.000000000 -0800
+++ itcl/configure	2007-01-30 17:11:28.000000000 -0800
@@ -1399,34 +1399,6 @@
 OBJEXT=$ac_cv_objext
 ac_objext=$ac_cv_objext
 
-    
-
-echo $ac_n "checking for executable suffix""... $ac_c" 1>&6
-echo "configure:1406: checking for executable suffix" >&5
-if eval "test \"`echo '$''{'ac_cv_exeext'+set}'`\" = set"; then
-  echo $ac_n "(cached) $ac_c" 1>&6
-else
-  if test "$CYGWIN" = yes || test "$MINGW32" = yes; then
-  ac_cv_exeext=.exe
-else
-  rm -f conftest*
-  echo 'int main () { return 0; }' > conftest.$ac_ext
-  ac_cv_exeext=
-  if { (eval echo configure:1416: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; }; then
-    for file in conftest.*; do
-      case $file in
-      *.c | *.o | *.obj) ;;
-      *) ac_cv_exeext=`echo $file | sed -e s/conftest//` ;;
-      esac
-    done
-  else
-    { echo "configure: error: installation or configuration problem: compiler cannot create executables." 1>&2; exit 1; }
-  fi
-  rm -f conftest*
-  test x"${ac_cv_exeext}" = x && ac_cv_exeext=no
-fi
-fi
-
 EXEEXT=""
 test x"${ac_cv_exeext}" != xno && EXEEXT=${ac_cv_exeext}
 echo "$ac_t""${ac_cv_exeext}" 1>&6

diff -c /Network/Servers/hills/Volumes/tienda/bbyer/work/tcl/tcl_ext/incrtcl/incrTcl/itk/configure\~ /Network/Servers/hills/Volumes/tienda/bbyer/work/tcl/tcl_ext/incrtcl/incrTcl/itk/configure
--- itk/configure~	2003-04-04 12:57:43.000000000 -0800
+++ itk/configure	2007-01-30 19:16:19.000000000 -0800
@@ -1405,34 +1405,6 @@
 OBJEXT=$ac_cv_objext
 ac_objext=$ac_cv_objext
 
-    
-
-echo $ac_n "checking for executable suffix""... $ac_c" 1>&6
-echo "configure:1412: checking for executable suffix" >&5
-if eval "test \"`echo '$''{'ac_cv_exeext'+set}'`\" = set"; then
-  echo $ac_n "(cached) $ac_c" 1>&6
-else
-  if test "$CYGWIN" = yes || test "$MINGW32" = yes; then
-  ac_cv_exeext=.exe
-else
-  rm -f conftest*
-  echo 'int main () { return 0; }' > conftest.$ac_ext
-  ac_cv_exeext=
-  if { (eval echo configure:1422: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; }; then
-    for file in conftest.*; do
-      case $file in
-      *.c | *.o | *.obj) ;;
-      *) ac_cv_exeext=`echo $file | sed -e s/conftest//` ;;
-      esac
-    done
-  else
-    { echo "configure: error: installation or configuration problem: compiler cannot create executables." 1>&2; exit 1; }
-  fi
-  rm -f conftest*
-  test x"${ac_cv_exeext}" = x && ac_cv_exeext=no
-fi
-fi
-
 EXEEXT=""
 test x"${ac_cv_exeext}" != xno && EXEEXT=${ac_cv_exeext}
 echo "$ac_t""${ac_cv_exeext}" 1>&6

