Index: Makefile.in
===================================================================
RCS file: /cvsroot/incrtcl/iwidgets/Makefile.in,v
retrieving revision 1.14
diff -u -u -p -r1.14 Makefile.in
--- Makefile.in	25 Sep 2002 23:12:59 -0000	1.14
+++ Makefile.in	26 Dec 2003 01:12:30 -0000
@@ -31,6 +31,7 @@ exec_prefix =	@exec_prefix@
 srcdir = @srcdir@
 bindir = @bindir@
 libdir = @libdir@
+mandir = @mandir@
 
 # The following definition can be set to non-null for special systems
 # like AFS with replication.  It allows the pathnames used for installation
@@ -49,7 +50,7 @@ LIB_INSTALL_DIR = $(INSTALL_ROOT)$(libdi
 BIN_INSTALL_DIR = $(INSTALL_ROOT)$(exec_prefix)/bin
 
 # Top-level directory in which to install manual entries:
-MAN_INSTALL_DIR = $(INSTALL_ROOT)$(prefix)/man/mann
+MAN_INSTALL_DIR = $(INSTALL_ROOT)$(mandir)/mann
 
 # Directory containing Tcl source code (for library used during test):
 TCL_SRC_DIR = @TCL_SRC_DIR@
