diff -ru ../openssh-4.5p1.orig/openbsd-compat/glob.c ./openbsd-compat/glob.c
--- ../openssh-4.5p1.orig/openbsd-compat/glob.c	2006-09-01 03:29:11.000000000 -0700
+++ ./openbsd-compat/glob.c	2007-08-02 22:40:35.000000000 -0700
@@ -48,7 +48,8 @@
 
 #if !defined(HAVE_GLOB) || !defined(GLOB_HAS_ALTDIRFUNC) || \
     !defined(GLOB_HAS_GL_MATCHC) || \
-    !defined(HAVE_DECL_GLOB_NOMATCH) || HAVE_DECL_GLOB_NOMATCH == 0
+    !defined(HAVE_DECL_GLOB_NOMATCH) || HAVE_DECL_GLOB_NOMATCH == 0 || \
+    defined(__APPLE_GLOBBING__)
 
 static long
 get_arg_max(void)
diff -ru ../openssh-4.5p1.orig/openbsd-compat/glob.h ./openbsd-compat/glob.h
--- ../openssh-4.5p1.orig/openbsd-compat/glob.h	2006-09-01 03:29:11.000000000 -0700
+++ ./openbsd-compat/glob.h	2007-08-02 22:40:28.000000000 -0700
@@ -39,7 +39,8 @@
 
 #if !defined(HAVE_GLOB_H) || !defined(GLOB_HAS_ALTDIRFUNC) || \
     !defined(GLOB_HAS_GL_MATCHC) || \
-    !defined(HAVE_DECL_GLOB_NOMATCH) || HAVE_DECL_GLOB_NOMATCH == 0
+    !defined(HAVE_DECL_GLOB_NOMATCH) || HAVE_DECL_GLOB_NOMATCH == 0 || \
+    defined(__APPLE_GLOBBING__)
 
 #ifndef _GLOB_H_
 #define	_GLOB_H_
