diff -uNr ../openssh-4.5p1.orig/sshconnect.c ./sshconnect.c
--- ../openssh-4.5p1.orig/sshconnect.c	2006-10-23 10:02:24.000000000 -0700
+++ ./sshconnect.c	2007-01-10 16:23:25.000000000 -0800
@@ -319,9 +319,9 @@
 	hints.ai_family = family;
 	hints.ai_socktype = SOCK_STREAM;
 	snprintf(strport, sizeof strport, "%u", port);
-	if ((gaierr = getaddrinfo(host, strport, &hints, &aitop)) != 0)
-		fatal("%s: %.100s: %s", __progname, host,
-		    gai_strerror(gaierr));
+	if ((gaierr = getaddrinfo(host, strport, &hints, &aitop)) != 0) {
+		fatal("%s: Error resolving hostname %.100s: %s", __progname, host, (EAI_SYSTEM == gaierr) ? strerror(errno) : gai_strerror(gaierr));
+	}
 
 	for (attempt = 0; attempt < connection_attempts; attempt++) {
 		if (attempt > 0) {
