diff -Naur ../openssh-4.4p1.orig/sshd_config ./sshd_config
--- ../openssh-4.4p1.orig/sshd_config	2006-07-23 21:06:47.000000000 -0700
+++ ./sshd_config	2006-10-20 20:14:13.000000000 -0700
@@ -52,7 +52,8 @@
 # Don't read the user's ~/.rhosts and ~/.shosts files
 #IgnoreRhosts yes
 
-# To disable tunneled clear text passwords, change to no here!
+# To disable tunneled clear text passwords, change to no here! Also,
+# remember to set the UsePAM setting to 'no'.
 #PasswordAuthentication yes
 #PermitEmptyPasswords no
 
@@ -78,7 +79,7 @@
 # If you just want the PAM account and session checks to run without
 # PAM authentication, then enable this but set PasswordAuthentication
 # and ChallengeResponseAuthentication to 'no'.
-#UsePAM no
+#UsePAM yes
 
 #AllowTcpForwarding yes
 #GatewayPorts no
