diff -Naur ../openssh-4.4p1.orig/sshd.0 ./sshd.0
--- ../openssh-4.4p1.orig/sshd.0	2006-09-26 04:03:46.000000000 -0700
+++ ./sshd.0	2006-10-19 18:32:10.000000000 -0700
@@ -526,8 +526,7 @@
 
 SEE ALSO
      scp(1), sftp(1), ssh(1), ssh-add(1), ssh-agent(1), ssh-keygen(1),
-     chroot(2), hosts_access(5), login.conf(5), moduli(5), sshd_config(5),
-     inetd(8), sftp-server(8)
+     chroot(2), hosts_access(5), sshd_config(5), sftp-server(8)
 
 AUTHORS
      OpenSSH is a derivative of the original and free ssh 1.2.12 release by
diff -Naur ../openssh-4.4p1.orig/sshd.8 ./sshd.8
--- ../openssh-4.4p1.orig/sshd.8	2006-08-29 18:07:01.000000000 -0700
+++ ./sshd.8	2006-10-19 18:32:40.000000000 -0700
@@ -828,10 +828,7 @@
 .Xr ssh-keygen 1 ,
 .Xr chroot 2 ,
 .Xr hosts_access 5 ,
-.Xr login.conf 5 ,
-.Xr moduli 5 ,
 .Xr sshd_config 5 ,
-.Xr inetd 8 ,
 .Xr sftp-server 8
 .Sh AUTHORS
 OpenSSH is a derivative of the original and free
diff -Naur ../openssh-4.4p1.orig/sshd_config.0 ./sshd_config.0
--- ../openssh-4.4p1.orig/sshd_config.0	2006-09-26 04:03:48.000000000 -0700
+++ ./sshd_config.0	2006-10-19 18:35:10.000000000 -0700
@@ -84,8 +84,7 @@
 
      ChallengeResponseAuthentication
              Specifies whether challenge-response authentication is allowed.
-             All authentication styles from login.conf(5) are supported.  The
-             default is ``yes''.
+             The default is ``yes''.
 
      Ciphers
              Specifies the ciphers allowed for protocol version 2.  Multiple
diff -Naur ../openssh-4.4p1.orig/sshd_config.5 ./sshd_config.5
--- ../openssh-4.4p1.orig/sshd_config.5	2006-08-29 18:06:34.000000000 -0700
+++ ./sshd_config.5	2006-10-19 18:35:45.000000000 -0700
@@ -169,9 +169,6 @@
 By default, no banner is displayed.
 .It Cm ChallengeResponseAuthentication
 Specifies whether challenge-response authentication is allowed.
-All authentication styles from
-.Xr login.conf 5
-are supported.
 The default is
 .Dq yes .
 .It Cm Ciphers
