diff -Naur /var/tmp/openssh4.2-testing/OpenSSH.roots/OpenSSH/openssh/servconf.c ./servconf.c
--- /var/tmp/openssh4.2-testing/OpenSSH.roots/OpenSSH/openssh/servconf.c	2005-09-27 12:33:06.000000000 -0700
+++ ./servconf.c	2005-09-27 14:06:44.000000000 -0700
@@ -112,7 +112,7 @@
 {
 	/* Portable-specific options */
 	if (options->use_pam == -1)
-		options->use_pam = 0;
+		options->use_pam = 1;
 
 	/* Standard Options */
 	if (options->protocol == SSH_PROTO_UNKNOWN)
diff -Naurdp ../openssh/sshd_config ./sshd_config
--- ../openssh/sshd_config	2005-09-28 16:11:01.000000000 -0700
+++ ./sshd_config	2005-09-28 17:23:55.000000000 -0700
@@ -77,7 +77,7 @@
 # "PermitRootLogin without-password". If you just want the PAM account and 
 # session checks to run without PAM authentication, then enable this but set 
 # ChallengeResponseAuthentication=no
-#UsePAM no
+#UsePAM yes 
 
 #AllowTcpForwarding yes
 #GatewayPorts no
