--- src/ppc/config.h	2008-05-09 17:11:18.000000000 -0700
+++ src.switch_libiconv/ppc/config.h	2008-05-09 17:20:46.000000000 -0700
@@ -612,10 +612,10 @@
 #define STDC_HEADERS 1
 
 /* Using GNU libiconv */
-/* #undef USE_LIBICONV_GNU */
+#define USE_LIBICONV_GNU 1
 
 /* Using a native implementation of iconv in a separate library */
-#define USE_LIBICONV_NATIVE 1
+/* #undef USE_LIBICONV_NATIVE 1 */
 
 /* using the system-supplied PCRE library */
 /* #undef USE_SYSTEM_PCRE */
