--- ./ext/session-old/session.c      2007-08-02 15:09:24.000000000 -0700
+++ ./ext/session/session.c   2007-08-02 17:01:06.000000000 -0700
@@ -918,6 +918,7 @@
 {
	smart_str ncookie = {0};
	char *date_fmt = NULL;
+	char *e_session_name, *e_id;
 
	if (SG(headers_sent)) {
		char *output_start_filename = php_get_output_start_filename(TSRMLS_C);
@@ -932,11 +933,19 @@
		return;
	}
 
+	/* URL encode session_name and id because they might be user supplied */
+	e_session_name = php_url_encode(PS(session_name), strlen(PS(session_name)), NULL);
+	e_id = php_url_encode(PS(id), strlen(PS(id)), NULL);
+
+
	smart_str_appends(&ncookie, COOKIE_SET_COOKIE);
-	smart_str_appends(&ncookie, PS(session_name));
+	smart_str_appends(&ncookie, e_session_name);
	smart_str_appendc(&ncookie, '=');
-	smart_str_appends(&ncookie, PS(id));
-	
+	smart_str_appends(&ncookie, e_id);
+
+	efree(e_session_name);
+	efree(e_id);
+
	if (PS(cookie_lifetime) > 0) {
		struct timeval tv;
		time_t t;
