diff --git a/Makefile.PL b/Makefile.PL
index 505d1df..1a17193 100644
--- a/Makefile.PL
+++ b/Makefile.PL
@@ -70,7 +70,7 @@ Expat_Not_Installed;
 
 if ( not $expat_libpath and $] >= 5.006001 and $^O ne 'MSWin32' ) {
     require ExtUtils::Liblist;    # Buggy before this
-    ($expat_libpath) = ExtUtils::Liblist->ext('-lexpat');
+    ($expat_libpath) = ExtUtils::Liblist->ext([ '-L@SDKROOT@/usr/lib -lexpat ' ]);
 }
 
 # Don't try to descend into Expat directory for testing
@@ -105,6 +105,7 @@ WriteMakefile1(
             repository => 'http://github.com/toddr/XML-Parser',
         },
     },
+    LIBS => [ '-L@SDKROOT@/usr/lib -lexpat ' ],
     TEST_REQUIRES => {
         'Test::More' => 0,
         'warnings'   => 0,
diff --git a/Expat/Makefile.PL b/Expat/Makefile.PL
index 505d1df..1a17193 100644
--- a/Expat/Makefile.PL
+++ b/Expat/Makefile.PL
@@ -2,7 +2,7 @@ use ExtUtils::MakeMaker;
 use Config;
 use English;
 
-my $libs = "-lexpat";
+my $libs = "-L@SDKROOT@/usr/lib -lexpat";
 my @extras = ();
 
 push(@extras, INC => "-I$expat_incpath") if $expat_incpath;
