diff -c /tmp/openmpi.roots/openmpi/openmpi/ompi/mca/io/romio/romio/adio/common/Makefile.in /tmp/Makefile.in
--- ompi/mca/io/romio/romio/adio/common/Makefile.in	Thu Jun 22 07:01:31 2006
+++ ompi/mca/io/romio/romio/adio/common/Makefile.in	Thu Jul 24 07:01:31 2006
@@ -56,7 +56,7 @@
 pkgdatadir = $(datadir)/@PACKAGE@
 pkglibdir = $(libdir)/@PACKAGE@
 pkgincludedir = $(includedir)/@PACKAGE@
-top_builddir = ../..
+top_builddir = @top_builddir@
 am__cd = CDPATH="$${ZSH_VERSION+.}$(PATH_SEPARATOR)" && cd
 INSTALL = @INSTALL@
 install_sh_DATA = $(install_sh) -c -m 644

Diff finished.  Mon Jul 24 15:51:52 2006
