--- ./host.defs	2014-09-03 12:16:39.000000000 -0700
+++ ./host.defs	2014-09-03 12:15:58.000000000 -0700
@@ -854,10 +854,11 @@
 newdef	: defined(APPLE) && defined(MACH)
 comment : OS X
 vendor	:						: "apple"
-hosttype: defined(i386)					: "intel-pc"
+hosttype: defined(i386)					: "intel-mac"
 hosttype: defined(ppc)					: "powermac"
 hosttype: defined(M_amd64)				: "amd"
 ostype	:						: "darwin"
+machtype: defined(__x86_64__)				: "x86_64"
 machtype: defined(i386)					: "i386"
 machtype: defined(M_amd64)				: "x86_64"
 machtype: defined(ppc)					: "powerpc"
