--- init/environment.old	2007-02-23 16:12:19.000000000 -0800
+++ init/environment	2007-02-19 23:58:17.000000000 -0800
@@ -0,0 +1,22 @@
+##
+# TCSH Expanded C-Shell INITIALIZATION FILE
+# Global environment setup
+#
+# Wilfredo Sanchez Jr. | tritan@mit.edu
+# July 09, 1992
+#
+# MIT Project Athena
+#
+# ORIGINAL SOURCES: /usr/athena/lib/init/cshrc (ATHENA REL 7.3P)
+##
+
+setenv ENV_SET			# avoid repeat
+
+umask 022			# files created are -rw-r--r--
+
+##
+# Read user's environment
+##
+if (-r "${tcsh_initdir}/environment.mine") then
+  source "${tcsh_initdir}/environment.mine"
+endif
