--- init/README.old	2007-02-23 16:12:10.000000000 -0800
+++ init/README	2007-02-19 23:58:17.000000000 -0800
@@ -0,0 +1,27 @@
+This directory contains some useful tcsh files.
+
+In order to use this configuration:
+
+	echo "source /usr/share/tcsh/examples/rc"     >> ~/.tcshrc
+	echo "source /usr/share/tcsh/examples/login"  >> ~/.login
+	echo "source /usr/share/tcsh/examples/logout" >> ~/.logout
+
+To do this system-wide, do the same instead to /etc/csh.cshrc, /etc/csh.login, and /etc/csh.logout.
+
+In order to customize tcsh:
+
+	mkdir -p ~/Library/init/tcsh
+
+and create the following files there as necessary:
+
+	aliases.mine		- shell aliases
+	completions.mine	- completions
+	environment.mine	- environment
+	rc.mine			- run commands
+	path			- command search path
+
+See the corresponding file in /usr/share/tcsh/examples for more information about the role of each file. You can easily extend or override the configuration provided by the default file. For example, you can add more aliases by adding the appropriate commands in aliases.mine.
+
+	-Fred
+	 Massachusetts Institute of Technology
+	 tritan@mit.edu
