--- Misc/python.man.orig	2012-04-10 08:32:09.000000000 -0700
+++ Misc/python.man	2012-08-12 23:01:19.000000000 -0700
@@ -440,6 +440,49 @@
 The integer must be a decimal number in the range [0,4294967295].  Specifying
 the value 0 will lead to the same hash values as when hash randomization is
 disabled.
+.SH INTERACTIVE INPUT EDITING AND HISTORY SUBSTITUTION
+The Python inteterpreter supports editing of the current input line and
+history substitution, similar to facilities found in the Korn shell and
+the GNU Bash shell.
+However, rather than being implemented using the GNU Readline library, this
+Python interpreter uses the BSD EditLine library
+.BR editline (3)
+with a GNU Readline emulation layer.
+.PP
+The
+.I readline
+module provides the access to the EditLine library, but there are a few major
+differences compared to a traditional implementation using the Readline library.
+The command language used in the preference files is that of EditLine,
+as described in
+.BR editrc (5)
+and not that used by the Readline library.
+This also means that the
+.IR parse_and_bind ()
+routines uses EditLine commands.
+And the preference file itself is
+.I ~/.editrc
+instead of
+.IR ~/.inputrc .
+.PP
+For example, the
+.I rlcompleter
+module, which defines a completion function for the
+.I readline
+modules, works correctly with the EditLine libraries, but needs to be
+initialized somewhat differently:
+.IP
+import rlcompleter
+.br
+import readline
+.br
+readline.parse_and_bind("bind ^I rl_complete")
+.PP
+For
+.I vi
+mode, one needs:
+.IP
+readline.parse_and_bind("bind -v")
 .SH AUTHOR
 The Python Software Foundation: http://www.python.org/psf
 .SH INTERNET RESOURCES
