--- setup.py.orig	2013-10-29 08:04:39.000000000 -0700
+++ setup.py	2015-06-17 15:29:07.000000000 -0700
@@ -44,7 +44,7 @@
     """
     Returns True if 'path' can be located in an OSX SDK
     """
-    return (path.startswith('/usr/') and not path.startswith('/usr/local')) or path.startswith('/System/')
+    return path.startswith('/usr/') or path.startswith('/System/')
 
 def find_file(filename, std_dirs, paths):
     """Searches for the directory where a given file is located,
@@ -691,6 +691,7 @@
                                depends = ['socketmodule.h']) )
         # Detect SSL support for the socket module (via _ssl)
         search_for_ssl_incs_in = [
+                              '/usr/local/include',
                               '/usr/local/ssl/include',
                               '/usr/contrib/ssl/include/'
                              ]
