--- ltmain.sh.orig	2004-07-27 16:30:45.000000000 -0700
+++ ltmain.sh	2004-07-27 16:32:04.000000000 -0700
@@ -2281,7 +2281,7 @@
 	    if test -f "$output_objdir/$soname-def"; then :
 	    else
 	      $show "extracting exported symbol list from \`$soname'"
-	      save_ifs="$IFS"; IFS='~'
+	      save_ifs="$IFS"; IFS=':'
 	      eval cmds=\"$extract_expsyms_cmds\"
 	      for cmd in $cmds; do
 		IFS="$save_ifs"
@@ -2294,7 +2294,7 @@
 	    # Create $newlib
 	    if test -f "$output_objdir/$newlib"; then :; else
 	      $show "generating import library for \`$soname'"
-	      save_ifs="$IFS"; IFS='~'
+	      save_ifs="$IFS"; IFS=':'
 	      eval cmds=\"$old_archive_from_expsyms_cmds\"
 	      for cmd in $cmds; do
 		IFS="$save_ifs"
@@ -3551,7 +3551,7 @@
 	    export_symbols="$output_objdir/$libname.exp"
 	    $run $rm $export_symbols
 	    eval cmds=\"$export_symbols_cmds\"
-	    save_ifs="$IFS"; IFS='~'
+	    save_ifs="$IFS"; IFS=':'
 	    for cmd in $cmds; do
 	      IFS="$save_ifs"
 	      if len=`expr "X$cmd" : ".*"` &&
@@ -3728,7 +3728,7 @@
 	      else
 		# All subsequent reloadable object files will link in
 		# the last one created.
-		eval concat_cmds=\"\$concat_cmds~$reload_cmds $objlist $last_robj\"
+		eval concat_cmds=\"\$concat_cmds:$reload_cmds $objlist $last_robj\"
 	      fi
 	      last_robj=$output_objdir/$save_output-${k}.$objext
 	      k=`expr $k + 1`
@@ -3749,7 +3749,7 @@
 	    $run $rm $export_symbols
 	    libobjs=$output
 	    # Append the command to create the export file.
-	    eval concat_cmds=\"\$concat_cmds~$export_symbols_cmds\"
+	    eval concat_cmds=\"\$concat_cmds:$export_symbols_cmds\"
           fi
 
 	  # Set up a command to remove the reloadale object files
@@ -3764,7 +3764,7 @@
 	  $echo "creating a temporary reloadable object file: $output"
 
 	  # Loop through the commands generated above and execute them.
-	  save_ifs="$IFS"; IFS='~'
+	  save_ifs="$IFS"; IFS=':'
 	  for cmd in $concat_cmds; do
 	    IFS="$save_ifs"
 	    $show "$cmd"
@@ -3791,9 +3791,9 @@
 
 	  # Append the command to remove the reloadable object files
 	  # to the just-reset $cmds.
-	  eval cmds=\"\$cmds~$rm $delfiles\"
+	  eval cmds=\"\$cmds:$rm $delfiles\"
 	fi
-	save_ifs="$IFS"; IFS='~'
+	save_ifs="$IFS"; IFS=':'
 	for cmd in $cmds; do
 	  IFS="$save_ifs"
 	  $show "$cmd"
@@ -3947,7 +3947,7 @@
 
       output="$obj"
       eval cmds=\"$reload_cmds\"
-      save_ifs="$IFS"; IFS='~'
+      save_ifs="$IFS"; IFS=':'
       for cmd in $cmds; do
 	IFS="$save_ifs"
 	$show "$cmd"
@@ -3983,7 +3983,7 @@
 	reload_objs="$libobjs $reload_conv_objs"
 	output="$libobj"
 	eval cmds=\"$reload_cmds\"
-	save_ifs="$IFS"; IFS='~'
+	save_ifs="$IFS"; IFS=':'
 	for cmd in $cmds; do
 	  IFS="$save_ifs"
 	  $show "$cmd"
@@ -4637,7 +4637,7 @@
     return xstrdup (path);
 
   /* prepend the current directory */
-  /* doesn't handle '~' */
+  /* doesn't handle ':' */
   if (getcwd (tmp, LT_PATHMAX) == NULL)
     lt_fatal ("getcwd failed");
   size = strlen(tmp) + 1 + strlen(path) + 1; /* +2 for '/' and '\0' */
@@ -5005,7 +5005,7 @@
 	      if test "$obj" = "$last_oldobj" ; then
 	        RANLIB=$save_RANLIB
 	      fi  
-	      test -z "$concat_cmds" || concat_cmds=$concat_cmds~
+	      test -z "$concat_cmds" || concat_cmds=$concat_cmds:
 	      eval concat_cmds=\"\${concat_cmds}$old_archive_cmds\"
 	      objlist=
 	    fi
@@ -5015,11 +5015,11 @@
 	  if test "X$oldobjs" = "X" ; then
 	    eval cmds=\"\$concat_cmds\"
 	  else
-	    eval cmds=\"\$concat_cmds~$old_archive_cmds\"
+	    eval cmds=\"\$concat_cmds:$old_archive_cmds\"
 	  fi
 	fi
       fi
-      save_ifs="$IFS"; IFS='~'
+      save_ifs="$IFS"; IFS=':'
       for cmd in $cmds; do
 	IFS="$save_ifs"
 	$show "$cmd"
@@ -5418,7 +5418,7 @@
 	  # Do each command in the postinstall commands.
 	  lib="$destdir/$realname"
 	  eval cmds=\"$postinstall_cmds\"
-	  save_ifs="$IFS"; IFS='~'
+	  save_ifs="$IFS"; IFS=':'
 	  for cmd in $cmds; do
 	    IFS="$save_ifs"
 	    $show "$cmd"
@@ -5646,7 +5646,7 @@
 
       # Do each command in the postinstall commands.
       eval cmds=\"$old_postinstall_cmds\"
-      save_ifs="$IFS"; IFS='~'
+      save_ifs="$IFS"; IFS=':'
       for cmd in $cmds; do
 	IFS="$save_ifs"
 	$show "$cmd"
@@ -5684,7 +5684,7 @@
 	if test -n "$finish_cmds"; then
 	  # Do each command in the finish commands.
 	  eval cmds=\"$finish_cmds\"
-	  save_ifs="$IFS"; IFS='~'
+	  save_ifs="$IFS"; IFS=':'
 	  for cmd in $cmds; do
 	    IFS="$save_ifs"
 	    $show "$cmd"
@@ -5961,7 +5961,7 @@
 	    if test -n "$library_names"; then
 	      # Do each command in the postuninstall commands.
 	      eval cmds=\"$postuninstall_cmds\"
-	      save_ifs="$IFS"; IFS='~'
+	      save_ifs="$IFS"; IFS=':'
 	      for cmd in $cmds; do
 		IFS="$save_ifs"
 		$show "$cmd"
@@ -5976,7 +5976,7 @@
 	    if test -n "$old_library"; then
 	      # Do each command in the old_postuninstall commands.
 	      eval cmds=\"$old_postuninstall_cmds\"
-	      save_ifs="$IFS"; IFS='~'
+	      save_ifs="$IFS"; IFS=':'
 	      for cmd in $cmds; do
 		IFS="$save_ifs"
 		$show "$cmd"
