--- jcr_elements.c	2007-11-10 02:37:38.000000000 -0800
+++ jcr_elements.c.new	2009-05-01 18:27:43.000000000 -0700
@@ -172,7 +172,7 @@
     if (rc != G_IO_STATUS_NORMAL) {
       if (gerr != NULL) {
         log_warn(JDBG, "Main Channel Error: rc=%d, %d '%s' '%s'", rc,
-          gerr->code, gerr->domain, gerr->message);
+          gerr->code, g_quark_to_string(gerr->domain), gerr->message);
         g_error_free(gerr);
       } else {
         log_warn(JDBG, "Main Channel Error: rc=%d", rc);
