--- Makefile.PL.orig	2018-04-17 10:29:26.000000000 -0700
+++ Makefile.PL	2020-04-12 17:38:34.000000000 -0700
@@ -11,13 +11,16 @@
 if ($args =~ /LIBDIR[ =](\S+)/) {
   $libdir = $1;
 }
-my $libs = "-lssl -lcrypto";
+
+my $OPENSSL_PREFIX = $ENV{OPENSSL_PREFIX};
+
+my $libs = "-lssl -lcrypto -L'.$OPENSSL_PREFIX.'/lib'";
 
 WriteMakefile(
   'NAME'	    => 'Crypt::OpenSSL::Random',
   'VERSION_FROM'    => 'Random.pm',
   'LIBS'            => $libdir ? [ "-L$libdir $libs" ] : [ openssl_lib_paths() . " $libs" ],
-  'INC'             => $incdir ? "-I$incdir" : openssl_inc_paths(),
+  'INC'             => $incdir ? "-I$incdir -I'.$OPENSSL_PREFIX.'/include'" : openssl_inc_paths(),
   'AUTHOR'          => 'Ian Robertson',
    (eval { ExtUtils::MakeMaker->VERSION('6.52'); 1 } ?
     ('CONFIGURE_REQUIRES' =>
