diff -urN ../openssl-0.9.7i.orig/doc/crypto/d2i_X509_REQ.pod ./doc/crypto/d2i_X509_REQ.pod
--- ../openssl-0.9.7i.orig/doc/crypto/d2i_X509_REQ.pod	2002-10-09 10:14:41.000000000 -0700
+++ ./doc/crypto/d2i_X509_REQ.pod	2006-09-27 18:45:05.000000000 -0700
@@ -15,8 +15,8 @@
  X509_REQ *d2i_X509_REQ_bio(BIO *bp, X509_REQ **x);
  X509_REQ *d2i_X509_REQ_fp(FILE *fp, X509_REQ **x);
 
- int i2d_X509_REQ_bio(X509_REQ *x, BIO *bp);
- int i2d_X509_REQ_fp(X509_REQ *x, FILE *fp);
+ int i2d_X509_REQ_bio(BIO *bp, X509_REQ *x);
+ int i2d_X509_REQ_fp(FILE *fp, X509_REQ *x);
 
 =head1 DESCRIPTION
 
