diff -Naur /tmp/OpenSSL.roots/OpenSSL/openssl/crypto/Makefile ./crypto/Makefile
--- /tmp/OpenSSL.roots/OpenSSL/openssl/crypto/Makefile	2005-11-03 14:50:14.000000000 -0800
+++ ./crypto/Makefile	2005-11-03 15:09:54.000000000 -0800
@@ -83,6 +83,7 @@
 	@$(PERL) $(TOP)/util/mklink.pl ../include/openssl $(EXHEADER)
 	@$(PERL) $(TOP)/util/mklink.pl ../test $(TEST)
 	@$(PERL) $(TOP)/util/mklink.pl ../apps $(APPS)
+	@(cd idea && $(PERL) $(TOP)/../util/mklink.pl ../../include/openssl idea.h)
 	@for i in $(SDIRS); do \
 	(cd $$i && echo "making links in crypto/$$i..." && \
 	$(MAKE) CC='$(CC)' INCLUDES='${INCLUDES}' CFLAG='${CFLAG}' INSTALLTOP='${INSTALLTOP}' PEX_LIBS='${PEX_LIBS}' EX_LIBS='${EX_LIBS}' BN_ASM='${BN_ASM}' DES_ENC='${DES_ENC}' SHA1_ASM_OBJ='${SHA1_ASM_OBJ}' MD5_ASM_OBJ='${MD5_ASM_OBJ}' RMD160_ASM_OBJ='${RMD160_ASM_OBJ}' BF_ENC='${BF_ENC}' CAST_ENC='${CAST_ENC}' RC4_ENC='${RC4_ENC}' RC5_ENC='${RC5_ENC}' AR='${AR}' PERL='${PERL}' links ); \
