diff -rup orig/snack/unix/Makefile.in snack/unix/Makefile.in
--- unix/Makefile.in	Thu Jun 24 23:40:34 2004
+++ unix/Makefile.in	Sat Jul 24 02:23:33 2004
@@ -31,7 +31,7 @@ CC        = @CC@
 
 INCLUDES  = ${XINCLUDES} @TCL_INCLUDE_SPEC@ -I${TCL_INCPATH} -I${TK_INCPATH} @AINC@ @NISTINC@
 
-CFLAGS    = -O @CFLAGS@ @AFLAG@ ${INCLUDES} @TCLAPI@ -I${GENERIC_DIR} @DEFS@
+override CFLAGS := -O @CFLAGS@ $(CFLAGS) @AFLAG@ ${INCLUDES} @TCLAPI@ -I${GENERIC_DIR} @DEFS@
 
 LIBSO     = -lc @ALIB@ @TCL_LIB_SPEC@
 
diff -rup orig/snack/unix/configure snack/unix/configure
--- unix/configure	Thu Jun 24 23:40:34 2004
+++ unix/configure	Sat Jul 24 02:25:08 2004
@@ -1567,7 +1567,7 @@ case $system in
 	AFLAG="-DOS_X_CORE_AUDIO -DMA${HACK}C_OSX_TCL -DMA${HACK}C_OSX_TK"
 	ALIB="-framework CoreFoundation -framework CoreAudio"
 	AOBJ="jkAudIO_osx.o"
-	CFLAGS="-Os -fno-common"
+	CFLAGS="-Os ${CFLAGS}"
         ;;
     *)
 	AFLAG="-DOTHER"
diff -rup orig/snack/unix/configure.in snack/unix/configure.in
--- unix/configure.in	Thu Jun 24 23:40:34 2004
+++ unix/configure.in	Sat Jul 24 02:24:47 2004
@@ -127,7 +127,7 @@ case $system in
 	AFLAG="-DOS_X_CORE_AUDIO -DMA${HACK}C_OSX_TCL -DMA${HACK}C_OSX_TK"
 	ALIB="-framework CoreFoundation -framework CoreAudio"
 	AOBJ="jkAudIO_osx.o"
-	CFLAGS="-Os -fno-common"
+	CFLAGS="-Os ${CFLAGS}"
         ;;
     *)
 	AFLAG="-DOTHER"
--- unix/Makefile.in.orig	2004-08-11 16:22:10.000000000 -0700
+++ unix/Makefile.in	2004-08-11 16:22:17.000000000 -0700
@@ -167,6 +167,7 @@
 
 stublib: ${OBJSNST}
 	ar cr libsnackstub${VERSION}.a ${OBJSNST}
+	ranlib libsnackstub${VERSION}.a
 
 editversion: ${OBJSN}
 	if test "$(TCL_VERSION)" != "8.4"; then\
