--- bt_overflow.c.orig	2009-11-06 12:39:34.000000000 -0800
+++ bt_overflow.c	2009-11-06 12:40:06.000000000 -0800
@@ -88,7 +88,7 @@ __ovfl_get(BTREE *t, void *p, size_t *ss
 
 #ifdef DEBUG
 	if (pg == P_INVALID || sz == 0)
-		abort();
+		LIBC_ABORT("%s", pg == P_INVALID ? "pg == P_INVALID" : "sz == 0");
 #endif
 	/* Make the buffer bigger as necessary. */
 	if (*bufsz < sz) {
@@ -192,7 +192,7 @@ __ovfl_delete(BTREE *t, void *p)
 
 #ifdef DEBUG
 	if (pg == P_INVALID || sz == 0)
-		abort();
+		LIBC_ABORT("%s", pg == P_INVALID ? "pg == P_INVALID" : "sz == 0");
 #endif
 	if ((h = mpool_get(t->bt_mp, pg, 0)) == NULL)
 		return (RET_ERROR);
