--- lib/readline/display.c~	2008-01-11 14:08:32.000000000 -0800
+++ lib/readline/display.c	2008-01-11 14:08:17.000000000 -0800
@@ -443,11 +443,11 @@ init_line_structures (minsize)
     {
       /* should be enough. */
       inv_lbsize = vis_lbsize = 256;
-      inv_lbreaks = (int *)xmalloc (inv_lbsize * sizeof (int));
-      vis_lbreaks = (int *)xmalloc (vis_lbsize * sizeof (int));
 #if defined (HANDLE_MULTIBYTE)
       _rl_wrapped_line = (int *)xmalloc (vis_lbsize * sizeof (int));
 #endif
+      inv_lbreaks = (int *)xmalloc (inv_lbsize * sizeof (int));
+      vis_lbreaks = (int *)xmalloc (vis_lbsize * sizeof (int));
       inv_lbreaks[0] = vis_lbreaks[0] = 0;
     }
 }
