--- shell.c~	2007-08-13 11:21:32.000000000 -0700
+++ shell.c	2007-08-13 11:55:29.000000000 -0700
@@ -476,6 +476,8 @@ main (argc, argv, env)
   if (dump_translatable_strings)
     read_but_dont_execute = 1;
 #ifdef __APPLE__
+  if (privileged_mode == 0 && act_like_sh)
+    privileged_mode = (getenv("_BASH_IMPLICIT_DASH_PEE") != NULL);
   if (running_setuid && privileged_mode == 0 && (tiger_mode ? (act_like_sh == 0) : 1))
     disable_priv_mode ();
   else if (running_setuid && privileged_mode != 0) {
