--- /SourceCache/bash/bash-71/bash/lib/readline/text.c	2006-07-28 08:55:27.000000000 -0700
+++ ./lib/readline/text.c	2007-03-14 15:25:11.000000000 -0700
@@ -296,6 +296,10 @@ rl_forward_char (count, key)
 
   if (count > 0)
     {
+      if (rl_point == rl_end) {
+	rl_ding();
+	return 0;
+      }
       point = _rl_find_next_mbchar (rl_line_buffer, rl_point, count, MB_FIND_NONZERO);
 
 #if defined (VI_MODE)
