--- generic/threadPoolCmd.c.orig	2003-11-27 11:54:19.000000000 -0800
+++ generic/threadPoolCmd.c	2005-07-20 23:10:12.000000000 -0700
@@ -1710,18 +1710,21 @@
  *
  *----------------------------------------------------------------------
  */
+#ifdef __WIN32__
+#include <sys/timeb.h>
+#else
+#include <sys/time.h>
+#endif
 static void
 GetTime(timePtr)
     Tcl_Time *timePtr;
 {
 #ifdef __WIN32__
-#include <sys/timeb.h>
     struct timeb tb;
     (void)ftime(&tb);
     timePtr->sec  = tb.time;
     timePtr->usec = tb.millitm * 1000;
 #else
-#include <sys/time.h>
     struct timeval tv;
     (void)gettimeofday(&tv, NULL);
     timePtr->sec  = tv.tv_sec;
