diff -Naur net-snmp-5.6.2.1.orig/agent/mibgroup/ucd-snmp/logmatch.c net-snmp-5.6.2.1.working/agent/mibgroup/ucd-snmp/logmatch.c
--- net-snmp-5.6.2.1.orig/agent/mibgroup/ucd-snmp/logmatch.c	2015-10-06 16:42:16.000000000 -0700
+++ net-snmp-5.6.2.1.working/agent/mibgroup/ucd-snmp/logmatch.c	2017-10-04 17:05:41.000000000 -0700
@@ -520,6 +520,8 @@
 
 
     iindex = name[*length - 1] - 1;
+    if ((iindex < 0) || (iindex >= MAXLOGMATCH))
+        return (NULL);
     logmatch = &logmatchTable[iindex];
 
     if (logmatch->myRegexError == 0)
