diff -I '\$Id: ' -u -r -b -w -p -d --new-file --exclude-from=/Users/rstory/.rcfiles/diff-ignore SVN/include/net-snmp/library/container.h APPLE/include/net-snmp/library/container.h
--- SVN/include/net-snmp/library/container.h
+++ APPLE/include/net-snmp/library/container.h
@@ -276,6 +276,9 @@ extern "C" {
     int netsnmp_compare_mem(const char * lhs, size_t lhs_len,
                             const char * rhs, size_t rhs_len);
 
+    /** no structure, just 'char *' pointers */
+    int netsnmp_compare_direct_cstring(const void * lhs, const void * rhs);
+
     /** for_each callback to call free on data item */
     void  netsnmp_container_simple_free(void *data, void *context);
 
