diff -Naur net-snmp-5.6.orig/include/net-snmp/system/darwin13.h net-snmp-5.6.working/include/net-snmp/system/darwin13.h
--- net-snmp-5.6.orig/include/net-snmp/system/darwin13.h	1969-12-31 16:00:00.000000000 -0800
+++ net-snmp-5.6.working/include/net-snmp/system/darwin13.h	2013-04-25 17:56:45.000000000 -0700
@@ -0,0 +1,15 @@
+/* This is a place for the changes needed to get net_snmp runnning on darwin13 
+   Currently this just entails including darwin12.h
+*/
+
+/* for now copy the constant into our sources, see <rdar://problem/12801140>  */
+#ifndef PR_SLOWHZ
+#define PR_SLOWHZ       2               /* 2 slow timeouts per second */
+#endif
+
+#include "darwin12.h"
+
+#ifndef darwin13
+#   define darwin13 darwin
+#endif
+
diff -Naur net-snmp-5.4.1.orig/include/net-snmp/system/darwin12.h net-snmp-5.4.1/include/net-snmp/system/darwin12.h
--- net-snmp-5.4.1.orig/include/net-snmp/system/darwin12.h      1969-12-31 16:00:00.000000000 -0800
+++ net-snmp-5.4.1/include/net-snmp/system/darwin12.h   2009-10-30 15:13:42.000000000 -0800
@@ -0,0 +1,9 @@
+/* This is a place for the changes needed to get net_snmp runnning on darwin12 
+   Currently this just entails including darwin11.h
+*/
+
+#include "darwin11.h"
+
+#ifndef darwin12
+#   define darwin12 darwin
+#endif
diff -Naur net-snmp-5.4.1.orig/include/net-snmp/system/darwin11.h net-snmp-5.4.1/include/net-snmp/system/darwin11.h
--- net-snmp-5.4.1.orig/include/net-snmp/system/darwin11.h      1969-12-31 16:00:00.000000000 -0800
+++ net-snmp-5.4.1/include/net-snmp/system/darwin11.h   2009-10-30 15:13:42.000000000 -0800
@@ -0,0 +1,10 @@
+/* This is a place for the changes needed to get net_snmp runnning on darwin11 including
+   support for the new UNAME_RELEASE version
+   Currently this just enatils including darwin10.h
+*/
+
+#include "darwin10.h"
+
+#ifndef darwin11
+#   define darwin11 darwin
+#endif
diff -Naur net-snmp-5.4.1.orig/Makefile.in net-snmp-5.4.1/Makefile.in
--- net-snmp-5.4.1.orig/Makefile.in	2008-01-24 15:08:15.000000000 -0800
+++ net-snmp-5.4.1/Makefile.in	2009-10-30 15:06:35.000000000 -0800
@@ -16,7 +16,7 @@
 INSTALLHEADERS=version.h
 INCLUDESUBDIR=system
 INCLUDESUBDIRHEADERS= aix.h bsd.h bsdi3.h bsdi4.h bsdi.h cygwin.h \
-	darwin.h darwin7.h darwin8.h darwin9.h darwin10.h dragonfly.h dynix.h \
+       darwin.h darwin7.h darwin8.h darwin9.h darwin10.h darwin11.h darwin12.h darwin13.h dragonfly.h dynix.h \
 	freebsd2.h freebsd3.h freebsd4.h freebsd5.h freebsd6.h freebsd.h \
	generic.h \
 	hpux.h irix.h linux.h mingw32.h mips.h netbsd.h openbsd.h osf5.h \
 	solaris2.3.h solaris2.4.h solaris2.5.h solaris2.6.h \

