--- Makefile.in.orig	2017-05-19 16:08:20.000000000 -0700
+++ Makefile.in	2017-05-19 16:12:26.000000000 -0700
@@ -2958,7 +2958,7 @@
 	$(MAKE) $(AM_MAKEFLAGS) check-TESTS
 check: $(BUILT_SOURCES)
 	$(MAKE) $(AM_MAKEFLAGS) check-am
-all-am: Makefile $(LTLIBRARIES) $(PROGRAMS) $(SCRIPTS) $(MANS) $(DATA) \
+all-am: Makefile $(LTLIBRARIES) $(SCRIPTS) $(MANS) $(DATA) \
 		$(HEADERS) config.h
 install-binPROGRAMS: install-libLTLIBRARIES
 
@@ -3038,7 +3038,7 @@
 
 install-dvi-am:
 
-install-exec-am: install-binPROGRAMS install-binSCRIPTS \
+install-exec-am: install-binSCRIPTS \
 	install-libLTLIBRARIES
 
 install-html: install-html-am
