--- /tmp/jabberd-2.2.13/subst/syslog.c	2011-02-23 08:24:35.000000000 -0800
+++ ./jabberd2/subst/syslog.c	2011-02-24 16:25:21.000000000 -0800
@@ -68,6 +68,7 @@ static struct dsn_c_pvt_sfnt {
         { LOG_LOCAL5,           "local5"},
         { LOG_LOCAL6,           "local6"},
         { LOG_LOCAL7,           "local7"},
+        { LOG_USER,             "log_user"},
         { 0,                    NULL }
 };
 
