--- /Volumes/XDisk/tmp/Libc/stdio/FreeBSD/findfp.c.orig	2004-05-22 08:19:41.000000000 -0700
+++ /Volumes/XDisk/tmp/Libc/stdio/FreeBSD/findfp.c	2004-10-24 17:08:30.000000000 -0700
@@ -47,6 +47,7 @@
 #include <stdlib.h>
 #include <string.h>
 
+#include <pthread.h>
 #include <spinlock.h>
 
 #include "libc_private.h"
@@ -62,12 +63,13 @@
 	 {0}, __sFX + file}
   /*	 p r w flags file _bf z  cookie      close    read    seek    write */
   /*     _ub _extra */
+#define	__sFXInit	{0, PTHREAD_MUTEX_INITIALIZER}
 				/* the usual - (stdin + stdout + stderr) */
 static FILE usual[FOPEN_MAX - 3];
 static struct __sFILEX usual_extra[FOPEN_MAX - 3];
 static struct glue uglue = { NULL, FOPEN_MAX - 3, usual };
 
-static struct __sFILEX __sFX[3];
+static struct __sFILEX __sFX[3] = {__sFXInit, __sFXInit, __sFXInit};
 
 /*
  * We can't make this 'static' until 6.0-current due to binary
@@ -113,7 +115,7 @@
 {
 	struct glue *g;
 	static FILE empty;
-	static struct __sFILEX emptyx;
+	static struct __sFILEX emptyx = __sFXInit;
 	FILE *p;
 	struct __sFILEX *fx;
 
@@ -179,6 +181,7 @@
 	fp->_lb._base = NULL;	/* no line buffer */
 	fp->_lb._size = 0;
 /*	fp->_lock = NULL; */	/* once set always set (reused) */
+	fp->_extra->fl_mutex = (pthread_mutex_t)PTHREAD_MUTEX_INITIALIZER;
 	fp->_extra->orientation = 0;
 	memset(&fp->_extra->mbstate, 0, sizeof(mbstate_t));
 	return (fp);
