--- fprintf.c.bsdnew	2009-11-11 13:33:07.000000000 -0800
+++ fprintf.c	2009-11-11 13:33:07.000000000 -0800
@@ -36,6 +36,8 @@ static char sccsid[] = "@(#)fprintf.c	8.
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/fprintf.c,v 1.11 2007/01/09 00:28:06 imp Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdio.h>
 #include <stdarg.h>
 
@@ -46,7 +48,20 @@ fprintf(FILE * __restrict fp, const char
 	va_list ap;
 
 	va_start(ap, fmt);
-	ret = vfprintf(fp, fmt, ap);
+	ret = vfprintf_l(fp, __current_locale(), fmt, ap);
+	va_end(ap);
+	return (ret);
+}
+
+int
+fprintf_l(FILE * __restrict fp, locale_t loc, const char * __restrict fmt, ...)
+{
+	int ret;
+	va_list ap;
+
+	/* no need to call NORMALIZE_LOCALE(loc), because vfprintf_l will */
+	va_start(ap, fmt);
+	ret = vfprintf_l(fp, loc, fmt, ap);
 	va_end(ap);
 	return (ret);
 }
