--- fscanf.c.bsdnew	2009-11-11 13:33:09.000000000 -0800
+++ fscanf.c	2009-11-11 13:33:09.000000000 -0800
@@ -36,6 +36,8 @@ static char sccsid[] = "@(#)fscanf.c	8.1
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/fscanf.c,v 1.13 2007/01/09 00:28:06 imp Exp $");
 
+#include "xlocale_private.h"
+
 #include "namespace.h"
 #include <stdio.h>
 #include <stdarg.h>
@@ -51,7 +53,22 @@ fscanf(FILE * __restrict fp, char const 
 
 	va_start(ap, fmt);
 	FLOCKFILE(fp);
-	ret = __svfscanf(fp, fmt, ap);
+	ret = __svfscanf_l(fp, __current_locale(), fmt, ap);
+	va_end(ap);
+	FUNLOCKFILE(fp);
+	return (ret);
+}
+
+int
+fscanf_l(FILE * __restrict fp, locale_t loc, char const * __restrict fmt, ...)
+{
+	int ret;
+	va_list ap;
+
+	NORMALIZE_LOCALE(loc);
+	va_start(ap, fmt);
+	FLOCKFILE(fp);
+	ret = __svfscanf_l(fp, loc, fmt, ap);
 	va_end(ap);
 	FUNLOCKFILE(fp);
 	return (ret);
