--- printf.c.bsdnew	2009-11-11 13:33:13.000000000 -0800
+++ printf.c	2009-11-11 13:33:13.000000000 -0800
@@ -36,6 +36,8 @@ static char sccsid[] = "@(#)printf.c	8.1
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/printf.c,v 1.11 2007/01/09 00:28:07 imp Exp $");
 
+#include "xlocale_private.h"
+
 #include <stdio.h>
 #include <stdarg.h>
 
@@ -46,7 +48,20 @@ printf(char const * __restrict fmt, ...)
 	va_list ap;
 
 	va_start(ap, fmt);
-	ret = vfprintf(stdout, fmt, ap);
+	ret = vfprintf_l(stdout, __current_locale(), fmt, ap);
+	va_end(ap);
+	return (ret);
+}
+
+int
+printf_l(locale_t loc, char const * __restrict fmt, ...)
+{
+	int ret;
+	va_list ap;
+
+	/* no need to call NORMALIZE_LOCALE(loc) because vfprintf_l will */
+	va_start(ap, fmt);
+	ret = vfprintf_l(stdout, loc, fmt, ap);
 	va_end(ap);
 	return (ret);
 }
