--- snprintf.c.orig	2009-11-30 16:15:30.000000000 -0800
+++ snprintf.c	2009-12-02 16:48:44.000000000 -0800
@@ -36,6 +36,8 @@ static char sccsid[] = "@(#)snprintf.c	8
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: src/lib/libc/stdio/snprintf.c,v 1.22 2008/04/17 22:17:54 jhb Exp $");
 
+#include "xlocale_private.h"
+
 #include <limits.h>
 #include <stdio.h>
 #include <stdarg.h>
@@ -45,26 +47,24 @@ __FBSDID("$FreeBSD: src/lib/libc/stdio/s
 int
 snprintf(char * __restrict str, size_t n, char const * __restrict fmt, ...)
 {
-	size_t on;
 	int ret;
 	va_list ap;
-	FILE f;
 
-	on = n;
-	if (n != 0)
-		n--;
-	if (n > INT_MAX)
-		n = INT_MAX;
 	va_start(ap, fmt);
-	f._file = -1;
-	f._flags = __SWR | __SSTR;
-	f._bf._base = f._p = (unsigned char *)str;
-	f._bf._size = f._w = n;
-	f._orientation = 0;
-	memset(&f._mbstate, 0, sizeof(mbstate_t));
-	ret = __vfprintf(&f, fmt, ap);
-	if (on > 0)
-		*f._p = '\0';
+	ret = vsnprintf_l(str, n, __current_locale(), fmt, ap);
+	va_end(ap);
+	return (ret);
+}
+
+int
+snprintf_l(char * __restrict str, size_t n, locale_t loc,
+    char const * __restrict fmt, ...)
+{
+	int ret;
+	va_list ap;
+
+	va_start(ap, fmt);
+	ret = vsnprintf_l(str, n, loc, fmt, ap);
 	va_end(ap);
 	return (ret);
 }
