--- lib-src/etags.c.orig	2007-02-26 19:19:57.000000000 -0800
+++ lib-src/etags.c	2012-10-31 11:59:02.000000000 -0700
@@ -6804,13 +6804,13 @@ absolute_filename (file, dir)
 	      else if (cp[0] != '/')
 		cp = slashp;
 #endif
-	      strcpy (cp, slashp + 3);
+	      memmove (cp, slashp + 3, strlen(slashp + 3)  + 1);
 	      slashp = cp;
 	      continue;
 	    }
 	  else if (slashp[2] == '/' || slashp[2] == '\0')
 	    {
-	      strcpy (slashp, slashp + 2);
+	      memmove (slashp, slashp + 2, strlen(slashp + 2) + 1);
 	      continue;
 	    }
 	}
