--- /SourceCache/emacs/emacs-90/emacs/src/regex.c	2007-01-20 20:18:15.000000000 -0800
+++ src/regex.c	2013-01-03 21:20:48.000000000 -0800
@@ -38,7 +38,7 @@
 # include <config.h>
 #endif
 
-#if defined STDC_HEADERS && !defined emacs
+#if defined STDC_HEADERS
 # include <stddef.h>
 #else
 /* We need this for `regex.h', and perhaps for the Emacs include files.  */
@@ -1863,7 +1863,7 @@ static int analyse_first _RE_ARGS ((re_c
     /* If the buffer moved, move all the pointers into it.  */		\
     if (old_buffer != bufp->buffer)					\
       {									\
-	int incr = bufp->buffer - old_buffer;				\
+	ptrdiff_t incr = bufp->buffer - old_buffer;			\
 	MOVE_BUFFER_POINTER (b);					\
 	MOVE_BUFFER_POINTER (begalt);					\
 	if (fixup_alt_jump)						\
