Index: Makefile.in
===================================================================
RCS file: /cvsroot/tcltrf/trf/Makefile.in,v
retrieving revision 1.7
diff -u -p -r1.7 Makefile.in
--- Makefile.in	9 Jan 2003 21:26:58 -0000	1.7
+++ Makefile.in	1 Jun 2004 14:16:46 -0000
@@ -208,6 +208,7 @@ MAKE_STATIC_LIB	= @MAKE_STATIC_LIB@
 MAKE_STUB_LIB	= @MAKE_STUB_LIB@
 OBJEXT		= @OBJEXT@
 RANLIB		= @RANLIB@
+RANLIB_STUB	= @RANLIB_STUB@
 SHLIB_CFLAGS	= @SHLIB_CFLAGS@
 SHLIB_SUFFIX	= @SHLIB_SUFFIX@
 SHLIB_LD	= @SHLIB_LD@
@@ -232,6 +233,7 @@ TCL_LIBS	= @TCL_LIBS@
 EXTRA_PATH	= $(top_builddir):$(TCL_BIN_DIR)
 TCLSH_ENV	= TCL_LIBRARY=`@CYGPATH@ $(TCL_SRC_DIR)/library` \
 		  LD_LIBRARY_PATH="$(EXTRA_PATH):$(LD_LIBRARY_PATH)" \
+		  DYLD_LIBRARY_PATH="$(EXTRA_PATH):$(DYLD_LIBRARY_PATH)" \
 		  LIBPATH="$(EXTRA_PATH):${LIBPATH}" \
 		  SHLIB_PATH="$(EXTRA_PATH):${SHLIB_PATH}" \
 		  PATH="$(EXTRA_PATH):$(PATH)" \
@@ -368,7 +370,7 @@ $($(PACKAGE)_LIB_FILE): $($(PACKAGE)_OBJ
 $($(PACKAGE)stub_LIB_FILE): $($(PACKAGE)stub_OBJECTS)
 	-rm -f $($(PACKAGE)stub_LIB_FILE)
 	${MAKE_STUB_LIB}
-	$(RANLIB) $($(PACKAGE)stub_LIB_FILE)
+	$(RANLIB_STUB) $($(PACKAGE)stub_LIB_FILE)
 
 #========================================================================
 # MD5 targets - Not anymore, integrated with main library
@@ -511,8 +513,14 @@ install-lib-binaries:
 	  if test -f $$p; then \
 	    echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(INSTALL_PROGRAM) $$p $(DESTDIR)$(pkglibdir)/$$p; \
+	    stub=`echo $$p|sed -e "s/.*\(stub\).*/\1/"`; \
+	    if test "x$$stub" = "xstub"; then \
+	    echo " $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p"; \
+	    $(RANLIB_STUB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    else \
 	    echo " $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p"; \
 	    $(RANLIB) $(DESTDIR)$(pkglibdir)/$$p; \
+	    fi; \
 	    ext=`echo $$p|sed -e "s/.*\.//"`; \
 	    if test "x$$ext" = "xdll"; then \
 		lib=`basename $$p|sed -e 's/.[^.]*$$//'`.lib; \
Index: configure
===================================================================
RCS file: /cvsroot/tcltrf/trf/configure,v
retrieving revision 1.14
diff -u -p -r1.14 configure
--- configure	4 Apr 2003 21:27:30 -0000	1.14
+++ configure	1 Jun 2004 14:16:46 -0000
@@ -5119,7 +5119,7 @@
 	    TCL_SHLIB_LD_EXTRAS="-compatibility_version ${TCL_MAJOR_VERSION} -current_version \${VERSION} -install_name \${LIB_RUNTIME_DIR}/\${TCL_LIB_FILE} -prebind -seg1addr 0xa000000"
 	    SHLIB_LD_LIBS='${LIBS}'
 	    SHLIB_SUFFIX=".dylib"
-	    DL_OBJS="tclLoadDyld.o"
+	    DL_OBJS="tclLoadDl.o"
 	    DL_LIBS=""
 	    LDFLAGS="-prebind"
 	    LD_SEARCH_FLAGS=""
@@ -5746,6 +5746,7 @@ EOF
     # substituted.
     #--------------------------------------------------------------------
 
+    RANLIB_STUB="${RANLIB}"
     if test "${TEA_PLATFORM}" = "windows" ; then
 	if test "${SHARED_BUILD}" = "1" ; then
 	    # We force the unresolved linking of symbols that are really in
@@ -5796,9 +5798,9 @@ EOF
 
 if test $HAVE_zlibtcl_PACKAGE -gt 0 ; then
     if test "${TEA_PLATFORM}" = "windows" -a "$GCC" != "yes" ; then
-	SHLIB_LD_LIBS="${SHLIB_LD_LIBS} \"`${CYGPATH} ${zlibtcl_STUB_LIB_PATH}`\""
+	SHLIB_LD_LIBS="${SHLIB_LD_LIBS} \"`${CYGPATH} ${zlibtcl_BUILD_STUB_LIB_PATH}`\""
     else
-	SHLIB_LD_LIBS="${SHLIB_LD_LIBS} ${zlibtcl_STUB_LIB_SPEC}"
+	SHLIB_LD_LIBS="${SHLIB_LD_LIBS} ${zlibtcl_BUILD_STUB_LIB_SPEC}"
     fi
 fi
 
@@ -6039,7 +6041,7 @@ else
             break
         fi
 
-        for libsuff in .so ".so.*" .sl .a; do
+        for libsuff in .so ".so.*" .sl .a .dylib; do
 	    if test -n "$trf_cv_lib_ZLIB_LIB_DIR"; then
                     break
             fi
@@ -6161,7 +6163,7 @@ else
             break
         fi
 
-        for libsuff in .so ".so.*" .sl .a; do
+        for libsuff in .so ".so.*" .sl .a .dylib; do
 	    if test -n "$trf_cv_lib_SSL_LIB_DIR"; then
                     break
             fi
@@ -6210,7 +6212,7 @@ else
             break
         fi
 
-        for libsuff in .so ".so.*" .sl .a; do
+        for libsuff in .so ".so.*" .sl .a .dylib; do
 	    if test -n "$trf_cv_lib_BZ2_LIB_DIR"; then
                     break
             fi
@@ -6598,6 +6600,7 @@ s%@MAKE_LIB@%$MAKE_LIB%g
 s%@MAKE_SHARED_LIB@%$MAKE_SHARED_LIB%g
 s%@MAKE_STATIC_LIB@%$MAKE_STATIC_LIB%g
 s%@MAKE_STUB_LIB@%$MAKE_STUB_LIB%g
+s%@RANLIB_STUB@%$RANLIB_STUB%g
 s%@SHLIB_SUFFIX@%$SHLIB_SUFFIX%g
 s%@TCLSH_PROG@%$TCLSH_PROG%g
 s%@DL_OBJS@%$DL_OBJS%g
Index: doc/rs_ecc.man
===================================================================
RCS file: /cvsroot/tcltrf/trf/doc/rs_ecc.man,v
retrieving revision 1.1
diff -u -p -r1.1 rs_ecc.man
--- doc/rs_ecc.man	13 Mar 2003 00:39:46 -0000	1.1
+++ doc/rs_ecc.man	1 Jun 2004 14:16:47 -0000
@@ -11,7 +11,7 @@ buffering 247 bytes.
 [para]
 [list_begin definitions]
 
-[call [cmd [vset encoding]] [opt [arg options...]] [opt [arg data]]]
+[call [cmd rs_ecc] [opt [arg options...]] [opt [arg data]]]
 
 [list_begin definitions]
 
Index: tclconfig/tcl.m4
===================================================================
RCS file: /cvsroot/tcltrf/trf/tclconfig/tcl.m4,v
retrieving revision 1.6
diff -u -p -r1.6 tcl.m4
--- tclconfig/tcl.m4	4 Apr 2003 21:27:36 -0000	1.6
+++ tclconfig/tcl.m4	1 Jun 2004 14:16:47 -0000
@@ -2591,6 +2591,7 @@ AC_DEFUN(TEA_MAKE_LIB, [
     # substituted.
     #--------------------------------------------------------------------
 
+    RANLIB_STUB="${RANLIB}"
     if test "${TEA_PLATFORM}" = "windows" ; then
 	if test "${SHARED_BUILD}" = "1" ; then
 	    # We force the unresolved linking of symbols that are really in
@@ -2632,6 +2633,7 @@ AC_DEFUN(TEA_MAKE_LIB, [
     AC_SUBST(MAKE_SHARED_LIB)
     AC_SUBST(MAKE_STATIC_LIB)
     AC_SUBST(MAKE_STUB_LIB)
+    AC_SUBST(RANLIB_STUB)
 ])
 
 #------------------------------------------------------------------------
Index: tools/mpexpand.tcl
===================================================================
RCS file: tools/mpexpand.tcl
diff -N tools/mpexpand.tcl
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ tools/mpexpand.tcl	1 Jun 2004 14:16:47 -0000
@@ -0,0 +1,165 @@
+#!/bin/sh
+# -*- tcl -*- \
+exec tclsh "$0" ${1+"$@"}
+
+lappend auto_path [file dirname [file dirname [info script]]]
+
+if 0 {
+    puts auto_path=\n\t[join $auto_path \n\t]
+    catch {puts tcl_pkgPath=\n\t[join $tcl_pkgPath \n\t]}
+    catch {puts tcl_libPath=\n\t[join $tcl_libPath \n\t]}
+
+    puts [package require doctools]
+    exit
+}
+
+package require doctools
+
+
+
+# ---------------------------------------------------------------------
+#  1. Handle command line options, input and output
+#  2. Initialize a doctools object.
+#  3. Run the input through the object.
+#  4. Write output.
+# ---------------------------------------------------------------------
+
+proc usage {{exitstate 1}} {
+    global argv0
+    puts "Usage: $argv0\
+	    ?-h|--help|-help|-??\
+	    ?-help-fmt|--help-fmt?\
+	    ?-module module?\
+	    ?-deprecated?\
+	    ?-copyright text?\
+	    format in|- ?out|-?"
+    exit $exitstate
+}
+
+# ---------------------------------------------------------------------
+
+proc fmthelp {} {
+    # Tcllib FR #527029: short reference of formatting commands.
+
+    global argv0
+    puts "$argv0 [doctools::help]"
+    exit 0
+}
+
+# ---------------------------------------------------------------------
+# 1. Handle command line options, input and output
+
+proc cmdline {} {
+    global argv0 argv format in out extmodule deprecated copyright
+
+    set copyright ""
+    set extmodule ""
+    set deprecated 0
+
+    while {[string match -* [set opt [lindex $argv 0]]]} {
+	switch -exact -- $opt {
+	    -module {
+		set extmodule [lindex $argv 1]
+		set argv [lrange $argv 2 end]
+		continue
+	    }
+	    -copyright {
+		set copyright [lindex $argv 1]
+		set argv [lrange $argv 2 end]
+		continue
+	    }
+	    -deprecated {
+		set deprecated 1
+		set argv [lrange $argv 1 end]
+	    }
+	    -help - -h - --help - -? {
+		# Tcllib FR #527029
+		usage 0
+	    }
+	    -help-fmt - --help-fmt {
+		# Tcllib FR #527029
+		fmthelp
+	    }
+	    default {
+		# Unknown option
+		usage
+	    }
+	}
+    }
+
+    if {[llength $argv] < 3} {
+	usage
+    }
+    foreach {format in out} $argv break
+
+    if {$format == {} || $in == {}} {
+	usage
+    }
+    if {$out == {}} {set out -}
+    return $format
+}
+
+# ---------------------------------------------------------------------
+#  3. Read input. Also providing the namespace with file information.
+
+proc get_input {} {
+    global in
+    if {[string equal $in -]} {
+	return [read stdin]
+    } else {
+	set if [open $in r]
+	set text [read $if]
+	close $if
+	return $text
+    }
+}
+
+# ---------------------------------------------------------------------
+# 4. Write output.
+
+proc write_out {text} {
+    global out
+    if {[string equal $out -]} {
+	puts -nonewline stdout $text
+    } else {
+	set of [open $out w]
+	puts -nonewline $of $text
+	close $of
+    }
+}
+
+
+# ---------------------------------------------------------------------
+# Get it all together
+
+proc main {} {
+    global format deprecated extmodule in copyright
+
+    #if {[catch {}
+	cmdline
+
+	::doctools::new dt -format $format -deprecated $deprecated -file $in
+	if {$extmodule != {}} {
+	    dt configure -module $extmodule
+	}
+	if {$copyright != {}} {
+	    dt configure -copyright $copyright
+	}
+
+	write_out [dt format [get_input]]
+
+	set warnings [dt warnings]
+	if {[llength $warnings] > 0} {
+	    puts stderr [join $warnings \n]
+	}
+
+	#{} msg]} {}
+	#puts stderr "Execution error: $msg"
+    #{}
+    return
+}
+
+
+# ---------------------------------------------------------------------
+main
+exit
