--- Makefile.bak	2006-10-18 10:40:26.000000000 -0700
+++ Makefile	2006-10-18 10:40:26.000000000 -0700
@@ -117,7 +117,7 @@
 INSTALL_STRIP_PROGRAM = ${SHELL} $(install_sh) -c -s
 LDFLAGS = -L/tmp/ChatServer2.roots/ChatServer2~obj/jabber_od_auth.build/lib -L/usr/lib/mysql -L/tmp/ChatServer2.roots/ChatServer2~obj/libidn.build/lib 
 LIBOBJS =  lstat$U.o
-LIBS = -lsqlite3 -lssl -lcrypto -lsasl2 -lidn 
+LIBS = -framework Security -framework DirectoryService -lxmppodauth -lmysqlclient -lssl -lcrypto -lsasl2 -lidn 
 LIBTOOL = $(SHELL) $(top_builddir)/libtool
 LN_S = ln -s
 LTLIBOBJS =  lstat$U.lo
