--- src/Makefile.in.ORIG	2006-11-13 15:59:26.000000000 -0800
+++ src/Makefile.in	2006-11-13 16:00:01.000000000 -0800
@@ -11,7 +11,7 @@
 mandir	= @mandir@
 transform = @program_transform_name@
 
-mandircat1 = ${mandir}/cat1
+mandirman1 = ${mandir}/man1
 
 CC	= @CC@
 CFLAGS	= @CFLAGS@
@@ -40,8 +40,8 @@
 install: all
 	-mkdir -p ${bindir}
 	${INSTALL} -m 555 ${PROG} ${bindir}/`echo ${PROG}|sed '$(transform)'`
-	-mkdir -p ${mandircat1}
-	${INSTALL} -m 444 ${srcdir}/${PROG}.cat1 ${mandircat1}/`echo ${PROG}|sed '$(transform)'`.1
+	-mkdir -p ${mandirman1}
+	${INSTALL} -m 444 ${srcdir}/${PROG}.1 ${mandirman1}/`echo ${PROG}|sed '$(transform)'`.1
 
 clean:
 	rm -f core ${PROG} ${OBJS}
