--- src/progressbar.c.orig	2006-11-30 10:33:12.000000000 -0800
+++ src/progressbar.c	2006-11-30 11:04:13.000000000 -0800
@@ -217,6 +217,8 @@
 	abbrevsize = cursize;
 	for (i = 0; abbrevsize >= 100000 && i < sizeof(prefixes); i++)
 		abbrevsize >>= 10;
+	if (i == sizeof(prefixes))
+		i--;	
 	len += snprintf(buf + len, BUFLEFT, " " LLFP("5") " %c%c ",
 	    (LLT)abbrevsize,
 	    prefixes[i],
@@ -325,6 +327,8 @@
 
 	for (i = 1; bytespersec >= 1024000 && i < sizeof(prefixes); i++)
 		bytespersec >>= 10;
+	if (i == sizeof(prefixes))
+	    i--;
 	len += snprintf(buf + len, BUFLEFT, "(" LLF ".%02d %cB/s)",
 	    (LLT)(bytespersec / 1024),
 	    (int)((bytespersec % 1024) * 100 / 1024),
--- src/util.c.orig	2006-11-30 11:06:24.000000000 -0800
+++ src/util.c	2006-11-30 11:10:20.000000000 -0800
@@ -466,7 +466,7 @@
 		memset(pass, 0, strlen(pass));
 	FREEPTR(pass);
 	if (acct != NULL)
-		memset(acct, 0, strlen(pass));
+		memset(acct, 0, strlen(acct));
 	FREEPTR(acct);
 	return (rval);
 }
@@ -878,10 +878,12 @@
 				putc('\n', ttyout);
 				break;
 			}
-			w = strlen(p);
-			while (w < width) {
+			if (p) {
+			    w = strlen(p);
+			    while (w < width) {
 				w = (w + 8) &~ 7;
 				(void)putc('\t', ttyout);
+			    }
 			}
 		}
 	}
