--- samba/source/configure.orig	Thu Sep  4 23:13:00 2003
+++ samba/source/configure	Thu Sep  4 23:13:08 2003
@@ -3024,7 +3024,7 @@ DYNEXP=
 
 default_static_modules="pdb_smbpasswd pdb_tdbsam rpc_lsa rpc_samr rpc_reg rpc_lsa_ds rpc_wks rpc_net rpc_dfs rpc_srv rpc_spoolss auth_rhosts auth_sam auth_unix auth_winbind auth_server auth_domain auth_builtin"
 
-default_shared_modules="vfs_recycle vfs_audit vfs_extd_audit vfs_netatalk vfs_fake_perms vfs_default_quota vfs_readonly vfs_cap charset_CP850 charset_CP437"
+default_shared_modules="vfs_recycle vfs_audit vfs_extd_audit vfs_netatalk vfs_fake_perms vfs_default_quota vfs_readonly vfs_cap"
 
 if test "x$developer" = xyes; then
    default_static_modules="$default_static_modules rpc_echo"
--- samba/source/include/smb.h.orig	Fri Aug 29 14:17:22 2003
+++ samba/source/include/smb.h	Fri Aug 29 14:40:33 2003
@@ -171,7 +171,7 @@ typedef uint16 smb_ucs2_t;
 typedef smb_ucs2_t wpstring[PSTRING_LEN];
 typedef smb_ucs2_t wfstring[FSTRING_LEN];
 
-#ifdef WORDS_BIGENDIAN
+#ifdef BIG_ENDIAN /* apple: configure test is bad for our platform ->  WORDS_BIGENDIAN */
 #define UCS2_SHIFT 8
 #else
 #define UCS2_SHIFT 0
--- samba/source/script/installman.sh.orig	Fri Aug 29 14:17:41 2003
+++ samba/source/script/installman.sh	Fri Aug 29 14:40:33 2003
@@ -25,7 +25,7 @@ for lang in $langs; do
     langdir=$MANDIR/$lang
     for d in $MANDIR $langdir $langdir/man1 $langdir/man5 $langdir/man7 $langdir/man8; do
 	if [ ! -d $d ]; then
-	    mkdir $d
+	    mkdir -p $d
 	    if [ ! -d $d ]; then
 		echo Failed to make directory $d, does $USER have privileges?
 		exit 1
--- samba/source/tdb/spinlock.c.orig	Fri Aug 29 14:17:45 2003
+++ samba/source/tdb/spinlock.c	Fri Aug 29 14:40:33 2003
@@ -75,7 +75,7 @@ static inline int __spin_is_locked(spinl
 	return (*lock != 0);
 }
 
-#elif defined(POWERPC_SPINLOCKS) 
+#elif defined(__ppc__) // defined(POWERPC_SPINLOCKS) 
 
 static inline int __spin_trylock(spinlock_t *lock)
 {
@@ -113,7 +113,7 @@ static inline int __spin_is_locked(spinl
 	return (*lock != 0);
 }
 
-#elif defined(INTEL_SPINLOCKS) 
+#elif defined(__i386__) // defined(INTEL_SPINLOCKS) 
 
 static inline int __spin_trylock(spinlock_t *lock)
 {
