Index: samba/source/script/installman.sh
===================================================================
--- samba/source/script/installman.sh.orig
+++ samba/source/script/installman.sh
@@ -33,7 +33,7 @@ for lang in $langs; do
     langdir=$MANDIR/$lang
     for d in $MANDIR $langdir $langdir/man1 $langdir/man5 $langdir/man7 $langdir/man8; do
 	if [ ! -d $d ]; then
-	    mkdir $d
+	    mkdir -p $d
 	    if [ ! -d $d ]; then
 		echo Failed to make directory $d, does $USER have privileges?
 		exit 1
Index: samba/source/configure.in
===================================================================
--- samba/source/configure.in.orig
+++ samba/source/configure.in
@@ -809,6 +809,8 @@ main() {
 # is so horribly broken....
 #
     *darwin*)
+	AC_DEFINE(DARWINOS,1,[Whether the host os is Darwin/MacOSX])
+	AC_DEFINE(STAT_ST_BLOCKSIZE,512)
 	AC_DEFINE(BROKEN_UNICODE_COMPOSE_CHARACTERS, 1, [Does this system use unicode compose characters])
 
 # Add a system specific charset module.
