--- samba/source/param/loadparm.c.orig	Thu Dec 18 15:12:46 2003
+++ samba/source/param/loadparm.c	Fri Dec 19 08:31:14 2003
@@ -287,6 +287,9 @@ typedef struct
 	BOOL bUnixExtensions;
 	BOOL bDisableNetbios;
 	BOOL bKernelChangeNotify;
+#ifdef WITH_OPENDIRECTORY
+	BOOL bOpenDirectory;
+#endif
 	int restrict_anonymous;
 	int name_cache_timeout;
 	int client_signing;
@@ -1523,6 +1526,10 @@ static void init_globals(void)
 	Globals.server_signing = False;
 
 	string_set(&Globals.smb_ports, SMB_PORTS);
+
+#ifdef WITH_OPENDIRECTORY
+	Globals.bOpenDirectory = True;
+#endif
 }
 
 static TALLOC_CTX *lp_talloc;
@@ -1899,6 +1906,9 @@ FN_GLOBAL_INTEGER(lp_algorithmic_rid_bas
 FN_GLOBAL_INTEGER(lp_name_cache_timeout, &Globals.name_cache_timeout)
 FN_GLOBAL_INTEGER(lp_client_signing, &Globals.client_signing)
 FN_GLOBAL_INTEGER(lp_server_signing, &Globals.server_signing)
+#ifdef WITH_OPENDIRECTORY
+FN_GLOBAL_BOOL(lp_opendirectory, &Globals.bOpenDirectory)
+#endif
 
 /* local prototypes */
 
@@ -4087,6 +4097,7 @@ static void set_default_server_announce_
 	default_server_announce |= SV_TYPE_WORKSTATION;
 	default_server_announce |= SV_TYPE_SERVER;
 	default_server_announce |= SV_TYPE_SERVER_UNIX;
+	default_server_announce |= SV_TYPE_PRINTQ_SERVER;
 
 	switch (lp_announce_as()) {
 		case ANNOUNCE_AS_NT_SERVER:
