--- samba/source/libsmb/clientgen.c.orig	Mon Aug 18 22:18:24 2003
+++ samba/source/libsmb/clientgen.c	Fri Aug 29 14:40:42 2003
@@ -335,6 +335,7 @@ close the session
 
 void cli_nt_session_close(struct cli_state *cli)
 {
+	if (cli != NULL) {
 	if (cli->ntlmssp_pipe_state) {
 		ntlmssp_client_end(&cli->ntlmssp_pipe_state);
 	}
@@ -344,6 +345,7 @@ void cli_nt_session_close(struct cli_sta
 
 	cli->nt_pipe_fnum = 0;
 	cli->pipe_idx = -1;
+	}
 }
 
 /****************************************************************************
@@ -394,11 +396,14 @@ void cli_close_connection(struct cli_sta
 
 void cli_shutdown(struct cli_state *cli)
 {
-	BOOL allocated = cli->allocated;
+	BOOL allocated = False;
+	if (cli != NULL) {
+		allocated = cli->allocated;
 	cli_close_connection(cli);
 	ZERO_STRUCTP(cli);
 	if (allocated)
 		free(cli);
+	}
 }
 
 /****************************************************************************
