--- samba/source/tdb/tdbtool.c.orig	Mon Aug 18 23:49:01 2003
+++ samba/source/tdb/tdbtool.c	Fri Dec 19 08:31:20 2003
@@ -208,9 +208,9 @@ static void insert_tdb(void)
 	}
 
 	key.dptr = k;
-	key.dsize = strlen(k)+1;
+	key.dsize = strlen(k);
 	dbuf.dptr = d;
-	dbuf.dsize = strlen(d)+1;
+	dbuf.dsize = strlen(d);
 
 	if (tdb_store(tdb, key, dbuf, TDB_INSERT) == -1) {
 		terror("insert failed");
@@ -229,9 +229,9 @@ static void store_tdb(void)
 	}
 
 	key.dptr = k;
-	key.dsize = strlen(k)+1;
+	key.dsize = strlen(k);
 	dbuf.dptr = d;
-	dbuf.dsize = strlen(d)+1;
+	dbuf.dsize = strlen(d);
 
 	printf("Storing key:\n");
 	print_rec(tdb, key, dbuf, NULL);
@@ -252,7 +252,7 @@ static void show_tdb(void)
 	}
 
 	key.dptr = k;
-	key.dsize = strlen(k)+1;
+	key.dsize = strlen(k);
 
 	dbuf = tdb_fetch(tdb, key);
 	if (!dbuf.dptr) {
@@ -285,7 +285,7 @@ static void delete_tdb(void)
 	}
 
 	key.dptr = k;
-	key.dsize = strlen(k)+1;
+	key.dsize = strlen(k);
 
 	if (tdb_delete(tdb, key) != 0) {
 		terror("delete failed");
@@ -310,7 +310,7 @@ static void move_rec(void)
 	}
 
 	key.dptr = k;
-	key.dsize = strlen(k)+1;
+	key.dsize = strlen(k);
 
 	dbuf = tdb_fetch(tdb, key);
 	if (!dbuf.dptr) {
