Index: samba/source/nmbd/nmbd_responserecordsdb.c
===================================================================
--- samba/source/nmbd/nmbd_responserecordsdb.c.orig
+++ samba/source/nmbd/nmbd_responserecordsdb.c
@@ -54,6 +54,13 @@ void remove_response_record(struct subne
 
 	struct response_record *p = NULL;
 
+	DEBUG(4,("remove_response_record: rrec (%x) subrec (%x) \n",
+		    rrec, subrec));
+
+	if (!rrec || !subrec) {
+		return;
+	}
+
 	for (p = subrec->responselist; p; p = p->next) {
 		if (p == rrec) {
 			break;
