Index: samba/source/include/includes.h
===================================================================
RCS file: /cvs/root/samba/samba/source/include/includes.h,v
retrieving revision 1.14
diff -u -d -b -r1.14 includes.h
--- samba/source/include/includes.h.orig	2004/02/11 18:51:53	1.14
+++ samba/source/include/includes.h	2004/06/04 23:18:40
@@ -1072,9 +1072,9 @@
 /* default socket options. Dave Miller thinks we should default to TCP_NODELAY
    given the socket IO pattern that Samba uses */
 #ifdef TCP_NODELAY
-#define DEFAULT_SOCKET_OPTIONS "TCP_NODELAY"
+#define DEFAULT_SOCKET_OPTIONS "TCP_NODELAY SO_RCVBUF=64240"
 #else
-#define DEFAULT_SOCKET_OPTIONS ""
+#define DEFAULT_SOCKET_OPTIONS "SO_RCVBUF=64240"
 #endif
 
 /* Load header file for dynamic linking stuff */
