--- samba/source/nmbd/nmbd.c.orig	Thu Jun  3 17:30:07 2004
+++ samba/source/nmbd/nmbd.c	Thu Jun  3 17:32:31 2004
@@ -43,7 +43,7 @@
 
 /* what server type are we currently */
 
-time_t StartupTime = 0;
+time_t NMBDStartupTime = 0;
 
 /**************************************************************************** **
  Handle a SIGTERM in band.
@@ -619,7 +619,7 @@
 
 	global_in_nmbd = True;
 	
-	StartupTime = time(NULL);
+	NMBDStartupTime = time(NULL);
 	
 	sys_srandom(time(NULL) ^ sys_getpid());
 	
--- samba/source/nmbd/nmbd_elections.c.orig	Thu Jun  3 17:37:56 2004
+++ samba/source/nmbd/nmbd_elections.c	Thu Jun  3 17:39:26 2004
@@ -24,7 +24,7 @@
 #include "includes.h"
 
 /* Election parameters. */
-extern time_t StartupTime;
+extern time_t NMBDStartupTime;
 
 /****************************************************************************
   Send an election datagram packet.
@@ -193,7 +193,7 @@
 				}
 
 				send_election_dgram(subrec, work->work_group, work->ElectionCriterion,
-						t - StartupTime, global_myname());
+						t - NMBDStartupTime, global_myname());
 	      
 				if (work->ElectionCount++ >= 4) {
 					/* Won election (4 packets were sent out uncontested. */
@@ -217,7 +217,7 @@
 static BOOL win_election(struct work_record *work, int version,
                          uint32 criterion, int timeup, const char *server_name)
 {  
-	int mytimeup = time(NULL) - StartupTime;
+	int mytimeup = time(NULL) - NMBDStartupTime;
 	uint32 mycriterion = work->ElectionCriterion;
 
 	/* If local master is false then never win in election broadcasts. */
