Index: samba-trunk/samba/source/nmbd/nmbd.c
===================================================================
--- samba-trunk.orig/samba/source/nmbd/nmbd.c
+++ samba-trunk/samba/source/nmbd/nmbd.c
@@ -47,7 +47,7 @@ BOOL found_lm_clients = False;
 
 /* what server type are we currently */
 
-time_t StartupTime = 0;
+time_t NMBDStartupTime = 0;
 
 /**************************************************************************** **
  Handle a SIGTERM in band.
@@ -649,7 +649,7 @@ static BOOL open_sockets(BOOL isdaemon, 
 
 	global_in_nmbd = True;
 	
-	StartupTime = time(NULL);
+	NMBDStartupTime = time(NULL);
 	
 	sys_srandom(time(NULL) ^ sys_getpid());
 	
Index: samba-trunk/samba/source/nmbd/nmbd_elections.c
===================================================================
--- samba-trunk.orig/samba/source/nmbd/nmbd_elections.c
+++ samba-trunk/samba/source/nmbd/nmbd_elections.c
@@ -24,7 +24,7 @@
 #include "includes.h"
 
 /* Election parameters. */
-extern time_t StartupTime;
+extern time_t NMBDStartupTime;
 
 /****************************************************************************
   Send an election datagram packet.
@@ -196,7 +196,7 @@ yet registered on subnet %s\n", nmb_name
 				}
 
 				send_election_dgram(subrec, work->work_group, work->ElectionCriterion,
-						t - StartupTime, global_myname());
+						t - NMBDStartupTime, global_myname());
 	      
 				if (work->ElectionCount++ >= 4) {
 					/* Won election (4 packets were sent out uncontested. */
@@ -220,7 +220,7 @@ yet registered on subnet %s\n", nmb_name
 static BOOL win_election(struct work_record *work, int version,
                          uint32 criterion, int timeup, const char *server_name)
 {  
-	int mytimeup = time(NULL) - StartupTime;
+	int mytimeup = time(NULL) - NMBDStartupTime;
 	uint32 mycriterion = work->ElectionCriterion;
 
 	/* If local master is false then never win in election broadcasts. */
