Index: samba/source/passdb/pdb_get_set.c
===================================================================
--- samba/source/passdb/pdb_get_set.c.orig
+++ samba/source/passdb/pdb_get_set.c
@@ -992,6 +992,8 @@ BOOL pdb_set_plaintext_passwd(struct sam
 	if (!plaintext)
 		return False;
 
+if (!lp_opendirectory()) {
+
 	/* Calculate the MD4 hash (NT compatible) of the password */
 	E_md4hash(plaintext, new_nt_p16);
 
@@ -1011,6 +1013,8 @@ BOOL pdb_set_plaintext_passwd(struct sam
 			return False;
 	}
 
+} /* !lp_opendirectory() */
+
 	if (!pdb_set_plaintext_pw_only (sampass, plaintext, PDB_CHANGED)) 
 		return False;
 
