--- /Volumes/XDisk/tmp/Libc/gen/FreeBSD/opendir.c.orig	2004-08-14 10:46:10.000000000 -0700
+++ /Volumes/XDisk/tmp/Libc/gen/FreeBSD/opendir.c	2004-10-24 17:08:28.000000000 -0700
@@ -48,6 +48,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <pthread.h>
 #include "un-namespace.h"
 
 #include "telldir.h"
@@ -269,7 +270,7 @@
 	dirp->dd_loc = 0;
 	dirp->dd_fd = fd;
 	dirp->dd_flags = flags;
-	dirp->dd_lock = NULL;
+	dirp->dd_lock = (pthread_mutex_t)PTHREAD_MUTEX_INITIALIZER;
 
 	/*
 	 * Set up seek point for rewinddir.
