--- wait.c.orig	2009-11-07 14:51:38.000000000 -0800
+++ wait.c	2009-11-07 14:51:40.000000000 -0800
@@ -40,10 +40,20 @@ __FBSDID("$FreeBSD: src/lib/libc/gen/wai
 #include <sys/resource.h>
 #include "un-namespace.h"
 
+#ifdef VARIANT_CANCELABLE
+int __wait4(pid_t, int *, int , struct rusage *);
+#else /* !VARIANT_CANCELABLE */
+int __wait4_nocancel(pid_t, int *, int , struct rusage *);
+#endif /* VARIANT_CANCELABLE */
+
 pid_t
 __wait(int *istat)
 {
-	return (_wait4(WAIT_ANY, istat, 0, (struct rusage *)0));
+#ifdef VARIANT_CANCELABLE
+	return (__wait4(WAIT_ANY, istat, 0, (struct rusage *)0));
+#else /* !VARIANT_CANCELABLE */
+	return (__wait4_nocancel(WAIT_ANY, istat, 0, (struct rusage *)0));
+#endif /* VARIANT_CANCELABLE */
 }
 
 __weak_reference(__wait, wait);
