diff --git a/cpan/Compress-Raw-Bzip2/Makefile.PL b/cpan/Compress-Raw-Bzip2/Makefile.PL
index efcdf3f..08db9cd 100644
--- a/cpan/Compress-Raw-Bzip2/Makefile.PL
+++ b/cpan/Compress-Raw-Bzip2/Makefile.PL
@@ -11,9 +11,9 @@ $WALL = ' -Wall -Wno-comment ' if $Config{'cc'} =~ /gcc/ ;
 my $USE_PPPORT_H = ($ENV{PERL_CORE}) ? '' : '-DUSE_PPPORT_H';
 
 
-my $BUILD_BZIP2 = defined($ENV{BUILD_BZIP2}) ? $ENV{BUILD_BZIP2} : 1;
-my $BZIP2_LIB = defined($ENV{BZIP2_LIB}) ? $ENV{BZIP2_LIB} : 'bzip2-src';
-my $BZIP2_INCLUDE = defined($ENV{BZIP2_INCLUDE}) ? $ENV{BZIP2_INCLUDE} : '.';
+my $BUILD_BZIP2 = defined($ENV{BUILD_BZIP2}) ? $ENV{BUILD_BZIP2} : 0;
+my $BZIP2_LIB = defined($ENV{BZIP2_LIB}) ? $ENV{BZIP2_LIB} : '@SDKROOT@/usr/lib';
+my $BZIP2_INCLUDE = defined($ENV{BZIP2_INCLUDE}) ? $ENV{BZIP2_INCLUDE} : '@SDKROOT@/usr/include';
 
 #ParseCONFIG() ;
 
