Index: gcc/config/rs6000/t-darwin
===================================================================
--- gcc/config/rs6000/t-darwin	(revision 121445)
+++ gcc/config/rs6000/t-darwin	(working copy)
@@ -25,8 +25,6 @@
 
 SHLIB_VERPFX = $(srcdir)/config/rs6000/darwin-libgcc
 
-LIB2ADDEH += $(srcdir)/config/rs6000/darwin-fallback.c
-
 darwin-fpsave.o:	$(srcdir)/config/rs6000/darwin-asm.h
 darwin-tramp.o:		$(srcdir)/config/rs6000/darwin-asm.h
 
Index: gcc/config/rs6000/darwin.h
===================================================================
--- gcc/config/rs6000/darwin.h	(revision 121445)
+++ gcc/config/rs6000/darwin.h	(working copy)
@@ -442,14 +442,6 @@
     } \
   while (0)
 
-#ifdef IN_LIBGCC2
-#include <stdbool.h>
-#endif
-
-#define MD_UNWIND_SUPPORT "config/rs6000/darwin-unwind.h"
-
-#define HAS_MD_FALLBACK_FRAME_STATE_FOR 1
-
 /* True, iff we're generating fast turn around debugging code.  When
    true, we arrange for function prologues to start with 5 nops so
    that gdb may insert code to redirect them, and for data to be
