--- base5067/innobase/btr/btr0cur.c	2008-08-04 05:19:12.000000000 -0700
+++ m5067/innobase/btr/btr0cur.c	2008-09-08 06:53:12.000000000 -0700
@@ -312,8 +312,8 @@
 
 #ifdef UNIV_SEARCH_PERF_STAT
 	info->n_searches++;
-#endif	
-	if (btr_search_latch.writer == RW_LOCK_NOT_LOCKED
+#endif
+	if (rw_lock_get_writer(&btr_search_latch) == RW_LOCK_NOT_LOCKED
 		&& latch_mode <= BTR_MODIFY_LEAF && info->last_hash_succ
 		&& !estimate
 #ifdef PAGE_CUR_LE_OR_EXTENDS
