--- pyobjc-framework-Cocoa/Lib/AppKit/PyObjC.bridgesupport.orig	2010-06-29 14:02:53.000000000 -0700
+++ pyobjc-framework-Cocoa/Lib/AppKit/PyObjC.bridgesupport	2010-07-28 13:47:32.000000000 -0700
@@ -2290,7 +2290,7 @@
     <arg type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}' type64='{CGRect={CGPoint=dd}{CGSize=dd}}' />
     <arg type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}' type64='{CGRect={CGPoint=dd}{CGSize=dd}}' />
     <arg c_array_length_in_arg='4' type='^i' type64='^q' type_modifier='n' />
-    <arg c_array_length_in_arg='4' type='^f' type64='^f' type_modifier='n' />
+    <arg c_array_length_in_arg='4' type='^f' type64='^d' type_modifier='n' />
     <arg type='i' type64='q' />
   </function>
   <function name='NSDrawWhiteBezel'>
@@ -2448,7 +2448,7 @@
   <function name='NSRectFillListWithGrays'>
     <retval type='v' type64='v' />
     <arg c_array_length_in_arg='2' type='^{_NSRect={_NSPoint=ff}{_NSSize=ff}}' type64='^{CGRect={CGPoint=dd}{CGSize=dd}}' type_modifier='n' />
-    <arg c_array_length_in_arg='2' type='^f' type64='^f' type_modifier='n' />
+    <arg c_array_length_in_arg='2' type='^f' type64='^d' type_modifier='n' />
     <arg type='i' type64='q' />
   </function>
   <function name='NSRectFillUsingOperation'>
@@ -2987,7 +2987,7 @@
     <method selector='tokenField:writeRepresentedObjects:toPasteboard:' type='B@:@@@' />
   </informal_protocol>
   <informal_protocol name='NSToolTipOwner'>
-    <method selector='view:stringForToolTip:point:userData:' type='@@:@i{_NSPoint=ff}^v' />
+    <method selector='view:stringForToolTip:point:userData:' type='@@:@i{_NSPoint=ff}^v' type64='@@:@i{CGPoint=dd}^v' />
   </informal_protocol>
   <informal_protocol name='NSToolbarDelegate'>
     <method selector='toolbar:itemForItemIdentifier:willBeInsertedIntoToolbar:' type='@@:@@B' />
--- pyobjc-framework-CoreText/Lib/CoreText/PyObjC.bridgesupport.orig	2010-07-11 23:00:38.000000000 -0700
+++ pyobjc-framework-CoreText/Lib/CoreText/PyObjC.bridgesupport	2010-07-28 13:55:37.000000000 -0700
@@ -867,9 +867,9 @@
     <retval type='d' />
     <arg type='^{__CTRun=}' />
     <arg type='{_CFRange=ll}' type64='{_CFRange=ll}' />
-    <arg type='^f' type64='^f' type_modifier='o' />
-    <arg type='^f' type64='^f' type_modifier='o' />
-    <arg type='^f' type64='^f' type_modifier='o' />
+    <arg type='^f' type64='^d' type_modifier='o' />
+    <arg type='^f' type64='^d' type_modifier='o' />
+    <arg type='^f' type64='^d' type_modifier='o' />
   </function>
   <function name='CTTextTabCreate'>
     <retval already_cfretained='true' type='^{__CTTextTab=}' />
--- pyobjc-framework-InterfaceBuilderKit/Exceptions/InterfaceBuilderKit.bridgesupport.orig	2010-06-22 13:52:18.000000000 -0700
+++ pyobjc-framework-InterfaceBuilderKit/Exceptions/InterfaceBuilderKit.bridgesupport	2010-07-28 13:36:49.000000000 -0700
@@ -199,7 +199,7 @@
       <retval type='@' />
     </method>
     <method selector='ibLayoutInset'>
-      <retval type='{IBInsetTag=ffff}' />
+      <retval type='{IBInsetTag=ffff}' type64='{IBInsetTag=dddd}' />
     </method>
     <method selector='ibMaximumSize'>
       <retval type='{_NSSize=ff}' type64='{_NSSize=dd}' />
--- pyobjc-framework-InterfaceBuilderKit/Lib/InterfaceBuilderKit/PyObjC.bridgesupport.orig	2010-07-01 00:18:58.000000000 -0700
+++ pyobjc-framework-InterfaceBuilderKit/Lib/InterfaceBuilderKit/PyObjC.bridgesupport	2010-07-28 13:36:49.000000000 -0700
@@ -2,7 +2,7 @@
 <!DOCTYPE signatures SYSTEM "file://localhost/System/Library/DTDs/BridgeSupport.dtd">
 <signatures version='1.0'>
   <depends_on path='/System/Library/Frameworks/Cocoa.framework/Cocoa' />
-  <struct name='IBInset' type='{IBInsetTag="left"f"top"f"right"f"bottom"f}' type64='{IBInsetTag="left"f"top"f"right"f"bottom"f}' />
+  <struct name='IBInset' type='{IBInsetTag="left"f"top"f"right"f"bottom"f}' type64='{IBInsetTag="left"d"top"d"right"d"bottom"d}' />
   <opaque name='IBDocumentStorage*' type='^{IBDocumentStorage=}' type64='^{IBDocumentStorage=}' />
   <constant name='IBAdditionalLocalizableKeyPaths' type='@' />
   <constant name='IBAttributeKeyPaths' type='@' />
@@ -234,7 +234,7 @@
       <retval type='@' />
     </method>
     <method selector='ibLayoutInset'>
-      <retval type='{IBInsetTag=ffff}' />
+      <retval type='{IBInsetTag=ffff}' type64='{IBInsetTag=dddd}' />
     </method>
     <method selector='ibMaximumSize'>
       <retval type='{_NSSize=ff}' type64='{CGSize=dd}' />
--- pyobjc-framework-Quartz/Exceptions/QuartzCore.bridgesupport.orig	2010-06-22 13:52:18.000000000 -0700
+++ pyobjc-framework-Quartz/Exceptions/QuartzCore.bridgesupport	2010-07-28 13:57:26.000000000 -0700
@@ -182,7 +182,7 @@
   </class>
   <class name='CIKernel'>
     <method selector='setROISelector:'>
-      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' />
+      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' sel_of_type64='{CGRect={CGPoint=dd}{CGSize=dd}}@:i{CGRect={CGPoint=dd}{CGSize=dd}}@' />
     </method>
   </class>
   <class name='CISampler'>
@@ -349,7 +349,7 @@
   </class>
   <class name='CIKernel'>
     <method selector='setROISelector:'>
-      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' />
+      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' sel_of_type64='{CGRect={CGPoint=dd}{CGSize=dd}}@:i{CGRect={CGPoint=dd}{CGSize=dd}}@' />
     </method>
   </class>
   <class name='CISampler'>
@@ -516,7 +516,7 @@
   </class>
   <class name='CIKernel'>
     <method selector='setROISelector:'>
-      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' />
+      <arg index='0' sel_of_type='{_NSRect={_NSPoint=ff}{_NSSize=ff}}@:i{_NSRect={_NSPoint=ff}{_NSSize=ff}}@' sel_of_type64='{CGRect={CGPoint=dd}{CGSize=dd}}@:i{CGRect={CGPoint=dd}{CGSize=dd}}@' />
     </method>
   </class>
   <class name='CISampler'>
--- pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport.orig	2010-07-01 03:56:20.000000000 -0700
+++ pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport	2010-07-28 14:54:41.000000000 -0700
@@ -761,10 +761,10 @@
   </function>
   <function name='CGColorSpaceCreateCalibratedRGB'>
     <retval already_cfretained='true' type='^{CGColorSpace=}' />
-    <arg type='[3f]' type64='[3f]' />
-    <arg type='[3f]' type64='[3f]' />
-    <arg type='[3f]' type64='[3f]' />
-    <arg type='[9f]' type64='[9f]' />
+    <arg type='[3f]' type64='[3d]' />
+    <arg type='[3f]' type64='[3d]' />
+    <arg type='[3f]' type64='[3d]' />
+    <arg type='[9f]' type64='[9d]' />
   </function>
   <function name='CGColorSpaceCreateDeviceCMYK'>
     <retval already_cfretained='true' type='^{CGColorSpace=}' />
@@ -790,9 +790,9 @@
   </function>
   <function name='CGColorSpaceCreateLab'>
     <retval already_cfretained='true' type='^{CGColorSpace=}' />
-    <arg type='[3f]' type64='[3f]' />
-    <arg type='[3f]' type64='[3f]' />
-    <arg type='[4f]' type64='[4f]' />
+    <arg type='[3f]' type64='[3d]' />
+    <arg type='[3f]' type64='[3d]' />
+    <arg type='[4f]' type64='[4d]' />
   </function>
   <function name='CGColorSpaceCreatePattern'>
     <retval already_cfretained='true' type='^{CGColorSpace=}' />
@@ -3013,7 +3013,7 @@
   <function name='CGPathContainsPoint'>
     <retval type='Z' />
     <arg type='^{CGPath=}' />
-    <arg type='^{CGAffineTransform=ffffff}' type_modifier='n' />
+    <arg type='^{CGAffineTransform=ffffff}' type64='^{CGAffineTransform=dddddd}' type_modifier='n' />
     <arg type='{CGPoint=ff}' type64='{CGPoint=dd}' />
     <arg type='Z' />
   </function>
