--- pyobjc-core/setup.py.orig	2010-07-24 07:41:14.000000000 -0700
+++ pyobjc-core/setup.py	2010-07-28 11:01:51.000000000 -0700
@@ -433,7 +433,7 @@
     platforms = [ 'MacOS X' ],
     ext_modules = ExtensionList,
     packages = [ 'objc', 'PyObjCTools', ], 
-    namespace_packages = ['PyObjCTools'],
+    #namespace_packages = ['PyObjCTools'],
     package_dir = { '': 'Lib', 'PyObjCTest': 'PyObjCTest' },
     extra_path = "PyObjC",
     cmdclass = {'build_ext': pyobjc_build_ext, 'install_lib': pyobjc_install_lib, 'build_py': oc_build_py, 'test': oc_test, 'egg_info':egg_info },
