--- pyobjc-core/Lib/objc/_pythonify.py.orig	2010-01-02 04:39:16.000000000 -0800
+++ pyobjc-core/Lib/objc/_pythonify.py	2010-07-28 11:43:07.000000000 -0700
@@ -107,8 +107,10 @@
             return OC_PythonLong(obj, obj.longLongValue())
         elif tp in b'QL':
             return OC_PythonLong(obj, obj.unsignedLongLongValue())
-        else:
+        elif tp == 'd':
             return OC_PythonFloat(obj, obj.doubleValue())
+        else:
+            return OC_PythonFloat(obj, obj.floatValue())
     elif sys.version_info[0] == 2:
         return OC_PythonInt(obj, obj.longValue())
     else:
