diff -rud ./pyobjc-framework-Quartz/Exceptions/CoreGraphics.bridgesupport ../pyobjc-2.3.2a0.mod/pyobjc-framework-Quartz/Exceptions/CoreGraphics.bridgesupport
--- ./pyobjc-framework-Quartz/Exceptions/CoreGraphics.bridgesupport	2010-06-22 13:52:18.000000000 -0700
+++ ../pyobjc-2.3.2a0.mod/pyobjc-framework-Quartz/Exceptions/CoreGraphics.bridgesupport	2016-04-15 16:52:23.000000000 -0700
@@ -93,6 +93,9 @@
   <function name='CGColorSpaceCopyICCProfile'>
     <retval already_cfretained='true' type='^{__CFData=}' type64='^{__CFData=}' />
   </function>
+  <function name='CGColorSpaceCopyICCData'>
+    <retval already_cfretained='true' type='^{__CFData=}' type64='^{__CFData=}' />
+  </function>
   <function name='CGColorSpaceCreateCalibratedGray'>
     <retval already_cfretained='true' />
   </function>
diff -rud ./pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport ../pyobjc-2.3.2a0.mod/pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport
--- ./pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport	2010-08-24 12:52:36.000000000 -0700
+++ ../pyobjc-2.3.2a0.mod/pyobjc-framework-Quartz/Lib/Quartz/CoreGraphics/PyObjC.bridgesupport	2016-04-15 16:51:51.000000000 -0700
@@ -753,6 +753,10 @@
     <retval already_cfretained='true' type='^{__CFData=}' type64='^{__CFData=}' />
     <arg type='^{CGColorSpace=}' />
   </function>
+  <function name='CGColorSpaceCopyICCData'>
+    <retval already_cfretained='true' type='^{__CFData=}' type64='^{__CFData=}' />
+    <arg type='^{CGColorSpace=}' />
+  </function>
   <function name='CGColorSpaceCreateCalibratedGray'>
     <retval already_cfretained='true' type='^{CGColorSpace=}' />
     <arg type='[3f]' type64='[3d]' />
