diff -r -u -N --exclude='*.orig' libiconv-1.11.orig/src/iconv.c libiconv-1.11/src/iconv.c
--- libiconv-1.11.orig/src/iconv.c	2006-07-04 06:30:23.000000000 -0700
+++ libiconv-1.11/src/iconv.c	2007-02-11 19:45:31.000000000 -0800
@@ -896,6 +896,10 @@
       const char *option = argv[i] + 1;
       if (*option == '\0')
         usage(1);
+      if (!strcmp(option,"-")) { /* handle -- option delimiter */
+	i++;
+	break;
+      }
       for (; *option; option++)
         switch (*option) {
           case 'c': discard_unconvertible = 1; break;
