--- lib/transfer.c.orig	2009-03-02 15:18:02.000000000 -0800
+++ lib/transfer.c	2009-03-02 15:18:16.000000000 -0800
@@ -910,6 +910,7 @@
         k->header = FALSE; /* no more header to parse! */
 
         if((k->size == -1) && !k->chunk && !conn->bits.close &&
+           (data->set.httpreq != HTTPREQ_HEAD) &&
            (conn->httpversion >= 11) ) {
           /* On HTTP 1.1, when connection is not to get closed, but no
              Content-Length nor Content-Encoding chunked have been
