--- /tmp/jabberd-2.1.24.1/mio/Makefile.am	2008-04-27 02:57:34.000000000 -0700
+++ ./jabberd2/mio/Makefile.am	2009-01-25 18:35:06.000000000 -0800
@@ -1,6 +1,6 @@
 noinst_LTLIBRARIES = libmio.la
 
-noinst_HEADERS = mio.h mio_impl.h mio_epoll.h mio_poll.h mio_select.h
+noinst_HEADERS = mio.h mio_impl.h mio_epoll.h mio_kqueue.h mio_poll.h mio_select.h
 
-libmio_la_SOURCES = mio.c mio_epoll.c mio_poll.c mio_select.c
+libmio_la_SOURCES = mio.c mio_epoll.c mio_kqueue.c mio_poll.c mio_select.c
 libmio_la_LIBADD = @LDFLAGS@
