diff -Naurp ncurses/tinfo/MKcaptab.sh ncurses/tinfo/MKcaptab.sh
--- ncurses/tinfo/MKcaptab.sh	2007-08-12 06:13:51.000000000 -0700
+++ ncurses/tinfo/MKcaptab.sh	2010-08-28 19:21:38.000000000 -0700
@@ -45,8 +45,8 @@ cat <<'EOF'
 
 EOF
 
-./make_hash 1 info $OPT1 <$DATA
-./make_hash 3 cap  $OPT1 <$DATA
+$BUILT_PRODUCTS_DIR/make_hash 1 info $OPT1 <$DATA
+$BUILT_PRODUCTS_DIR/make_hash 3 cap  $OPT1 <$DATA
 
 $AWK -f $OPT2 bigstrings=$OPT1 tablename=capalias <$DATA 
 
