--- base5067/innobase/ib_config.h	2008-08-04 05:24:29.000000000 -0700
+++ m5067/innobase/ib_config.h	2008-09-08 06:53:12.000000000 -0700
@@ -4,6 +4,9 @@
 /* Define to 1 if you have the <aio.h> header file. */
 #define HAVE_AIO_H 1
 
+/* Define to 1 if compiler provides atomic builtins. */
+#define HAVE_ATOMIC_BUILTINS 1
+
 /* Define to 1 if you have the <dlfcn.h> header file. */
 #define HAVE_DLFCN_H 1
 
