Index: autogen.sh
===================================================================
--- autogen.sh	(revision 659)
+++ autogen.sh	(working copy)
@@ -59,6 +59,16 @@
 autoconf${AUTOCONF_SUFFIX}
 cd $ORIGDIR
 
+## Fix the makefile
+#sed -e 's/^\(all:.*\)/\1\
+#	echo "s|\\(ECHO=\\(.*\\)\\)|echo=\\2\\\\" > echoFix.sed \
+#	echo " \\1|" >> echoFix.sed \
+#	sed -f echoFix.sed -i .bak libtool/' -i .bak Makefile.in | exit 1
+
+if [ "$1" == "--skip-configure" ]; then
+	exit
+fi
+
 $srcdir/configure --enable-maintainer-mode "$@" || exit
 
 echo 
