diff -r -u -N --exclude='*.orig' fetchmail-6.3.11.orig/configure fetchmail-6.3.11/configure
--- fetchmail-6.3.11.orig/configure	2009-08-05 16:33:55.000000000 -0700
+++ fetchmail-6.3.11/configure	2009-09-25 13:37:02.000000000 -0700
@@ -12485,7 +12485,6 @@
 $as_echo_n "checking if your getaddrinfo is async-signal-safe... " >&6; }
     gai_ts=no
         case $host_os in
-	darwin9*)				gai_ts=yes ;;
 	linux*)					gai_ts=yes ;;
 	freebsd5.5|freebsd6*|freebsd7*)		gai_ts=yes ;;
 	solaris2.8|solaris2.9|solaris2.10)	gai_ts=yes ;;
