--- mysql/sql/mysqld.cc	2011-01-25 03:27:51.000000000 -0800
+++ mysqld.cc	2011-03-22 17:05:41.000000000 -0700
@@ -2843,8 +2843,6 @@
     return 1;
   }
 
-  if (init_thread_environment())
-    return 1;
   mysql_init_variables();
 
 #ifdef OS2
@@ -2896,6 +2894,8 @@
   load_defaults(conf_file_name, groups, &argc, &argv);
   defaults_argv=argv;
   get_options(argc,argv);
+  if (init_thread_environment())
+    return 1;
   set_server_version();
 
   DBUG_PRINT("info",("%s  Ver %s for %s on %s\n",my_progname,
