Doing diffs in .:
--- ./libstdc++-v3/config/abi/pre/gnu.ver.~1~	2008-05-06 19:23:21.000000000 -0700
+++ ./libstdc++-v3/config/abi/pre/gnu.ver	2008-05-07 10:48:22.000000000 -0700
@@ -415,6 +415,33 @@ GLIBCXX_3.4 {
     _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEED2Ev;
     # APPLE LOCAL end libstdc++ abi compatibility fix 5897367
 
+    # APPLE LOCAL start libstdc++ abi compatibility fix 5768558
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEE2fdEv;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEE4fileEv;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC1EP7__sFILESt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC1EiSt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC1Ev;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC2EP7__sFILESt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC2EiSt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEC2Ev;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEED0Ev;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEED1Ev;
+    _ZN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEED2Ev;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEE2fdEv;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEE4fileEv;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC1EP7__sFILESt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC1EiSt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC1Ev;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC2EP7__sFILESt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC2EiSt13_Ios_Openmodem;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEC2Ev;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEED0Ev;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEED1Ev;
+    _ZN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEED2Ev;
+    _ZTVN9__gnu_cxx13stdio_filebufIcSt11char_traitsIcEEE;
+    _ZTVN9__gnu_cxx13stdio_filebufIwSt11char_traitsIwEEE;
+    # APPLE LOCAL end libstdc++ abi compatibility fix 5768558
+
     # __gnu_cxx::__atomic_add
     # __gnu_cxx::__exchange_and_add
     _ZN9__gnu_cxx12__atomic_addEPV[il]i;
--------------
