Doing diffs in .:
--- ./libstdc++-v3/config/abi/pre/gnu.ver.~1~	2008-05-06 19:06:51.000000000 -0700
+++ ./libstdc++-v3/config/abi/pre/gnu.ver	2008-05-06 19:14:04.000000000 -0700
@@ -100,7 +100,7 @@ GLIBCXX_3.4 {
       std::out_of_range*;
       std::overflow_error*;
 # APPLE LOCAL libstdc++ abi compatibility fix 5724986
-      std::search*
+      std::search*;
       std::set_new_handler*;
       std::set_terminate*;
       std::set_unexpected*;
@@ -386,6 +386,34 @@ GLIBCXX_3.4 {
 
     # __gnu_cxx::stdio_sync_filebuf
     _ZTVN9__gnu_cxx18stdio_sync_filebufI[cw]St11char_traitsI[cw]EEE;
+    # APPLE LOCAL begin libstdc++ abi compatibility fix 5897367
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE4syncEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE5uflowEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE6xsgetnEPci;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE6xsputnEPKci;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE7seekoffExSt12_Ios_SeekdirSt13_Ios_Openmode;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE8overflowEi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE9pbackfailEi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEE9underflowEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEEC1EP7__sFILE;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEEC2EP7__sFILE;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEED0Ev;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEED1Ev;
+    _ZN9__gnu_cxx18stdio_sync_filebufIcSt11char_traitsIcEED2Ev;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE4syncEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE5uflowEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE6xsgetnEPwi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE6xsputnEPKwi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE7seekoffExSt12_Ios_SeekdirSt13_Ios_Openmode;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE8overflowEi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE9pbackfailEi;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEE9underflowEv;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEEC1EP7__sFILE;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEEC2EP7__sFILE;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEED0Ev;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEED1Ev;
+    _ZN9__gnu_cxx18stdio_sync_filebufIwSt11char_traitsIwEED2Ev;
+    # APPLE LOCAL end libstdc++ abi compatibility fix 5897367
 
     # __gnu_cxx::__atomic_add
     # __gnu_cxx::__exchange_and_add
--------------
