Index: libstdc++-v3/include/std/std_fstream.h
===================================================================
RCS file: /cvs/gcc/gcc/libstdc++-v3/include/std/std_fstream.h,v
retrieving revision 1.66.8.3
retrieving revision 1.66.8.2
diff -u -p -r1.66.8.3 -r1.66.8.2
--- libstdc++-v3/include/std/std_fstream.h	12 Jul 2005 00:55:10 -0000	1.66.8.3
+++ libstdc++-v3/include/std/std_fstream.h	9 Apr 2005 06:13:26 -0000	1.66.8.2
@@ -436,7 +436,8 @@ _GLIBCXX_BEGIN_NAMESPACE(std)
       typedef basic_istream<char_type, traits_type>	__istream_type;
 
     private:
-      __filebuf_type	_M_filebuf;
+      /* APPLE LOCAL work with -malign-natural.  */
+      __filebuf_type	_M_filebuf  __attribute__((aligned(8)));
 
     public:
       // Constructors/Destructors:
@@ -564,7 +565,8 @@ _GLIBCXX_BEGIN_NAMESPACE(std)
       typedef basic_ostream<char_type, traits_type>	__ostream_type;
 
     private:
-      __filebuf_type	_M_filebuf;
+      /* APPLE LOCAL work with -malign-natural.  */
+      __filebuf_type	_M_filebuf __attribute__((aligned(8)));
 
     public:
       // Constructors:
