<rdar://problem/6339271>
diff -upr libstdc++-v3/testsuite/27_io/ios_base/storage/2.cc libstdc++-v3/testsuite/27_io/ios_base/storage/2.cc
--- libstdc++-v3/testsuite/27_io/ios_base/storage/2.cc	2008-11-04 11:20:23.000000000 -0500
+++ libstdc++-v3/testsuite/27_io/ios_base/storage/2.cc	2008-11-06 12:11:07.000000000 -0500
@@ -30,6 +30,10 @@
 // NULL even if an allocation fails (filed as Radar 3884894).
 // { dg-do run { xfail *-*-darwin8.[0-4].* } }
 
+// Allocates too much memory on 64-bit systems
+// <rdar://problem/6339271>
+// { dg-require-effective-target ilp32 }
+
 #include <sstream>
 #include <iostream>
 #include <testsuite_hooks.h>
