--- gcc-4.2/gcc/libstdc++-v3/testsuite/util/testsuite_hooks.cc	2006-11-14 20:23:34.000000000 -0500
+++ libstdc++-v3/testsuite/util/testsuite_hooks.cc	2006-12-06 15:16:04.000000000 -0500
@@ -232,7 +232,7 @@ namespace __gnu_test
   unsigned int destructor::_M_count = 0;
   int copy_tracker::next_id_ = 0;
 
-#ifdef _GLIBCXX_SYSV_SEM
+#if defined(_GLIBCXX_SYSV_SEM) && ! defined(LIBSTDCXX_FOR_JAGUAR)
   // This union is not declared in system headers.  Instead, it must
   // be defined by user programs.
   union semun 
