Doing diffs in .:
--- ./ltcf-c.sh.~1~	2007-02-20 18:00:51.000000000 -0800
+++ ./ltcf-c.sh	2008-01-23 18:42:44.000000000 -0800
@@ -402,7 +402,7 @@ else
     esac
     # Disable shared library build on OS-X older than 10.3.
     case $host_os in
-	darwin[1-5]*)
+	darwin[3-5]*)
 	    can_build_shared=no
 	    ;;
 	darwin[67]*)
--- ./ltcf-cxx.sh.~1~	2007-02-20 18:00:51.000000000 -0800
+++ ./ltcf-cxx.sh	2008-01-23 18:42:39.000000000 -0800
@@ -247,7 +247,7 @@ case $host_os in
     esac
     # Disable shared library build on OS-X older than 10.3.
     case $host_os in
-	darwin[1-5]*)
+	darwin[3-5]*)
 	    can_build_shared=no
 	    ;;
 	darwin[67]*)
--- ./ltcf-gcj.sh.~1~	2006-11-14 18:45:14.000000000 -0800
+++ ./ltcf-gcj.sh	2008-01-23 18:42:34.000000000 -0800
@@ -401,7 +401,7 @@ else
     esac
     # Disable shared library build on OS-X older than 10.3.
     case $host_os in
-	darwin[1-6]*)
+	darwin[3-6]*)
 	    can_build_shared=no
 	    ;;
 	darwin7*)
--- ./ltconfig.~1~	2007-02-20 18:00:51.000000000 -0800
+++ ./ltconfig	2008-01-23 18:42:27.000000000 -0800
@@ -1113,7 +1113,7 @@ darwin* | rhapsody*)
   need_version=no
   # Disable shared library build on OS-X older than 10.3.
   case $host_os in
-      darwin[1-5]*)
+      darwin[3-5]*)
 	  can_build_shared=no
 	  ;;
       darwin[67]*)
--------------
