diff -upr libstdc++-v3/include/ext/throw_allocator.h libstdc++-v3/include/ext/throw_allocator.h
--- libstdc++-v3/include/ext/throw_allocator.h	2007-06-26 11:35:42.000000000 -0400
+++ libstdc++-v3/include/ext/throw_allocator.h	2008-10-28 16:51:29.000000000 -0400
@@ -423,11 +423,11 @@ _GLIBCXX_BEGIN_NAMESPACE(__gnu_cxx)
     s += buf;
     s += tab;
     s += "label: ";
-    sprintf(buf, "%u", ref.second.first);
+    sprintf(buf, "%lu", ref.second.first);
     s += buf;
     s += tab;
     s += "size: ";
-    sprintf(buf, "%u", ref.second.second);
+    sprintf(buf, "%lu", ref.second.second);
     s += buf;
     s += '\n';
   }

diff -upr libstdc++-v3/testsuite/23_containers/map/operators/1_neg.cc libstdc++-v3/testsuite/23_containers/map/operators/1_neg.cc
--- libstdc++-v3/testsuite/23_containers/map/operators/1_neg.cc	2006-01-03 08:19:23.000000000 -0500
+++ libstdc++-v3/testsuite/23_containers/map/operators/1_neg.cc	2008-10-28 17:00:22.000000000 -0400
@@ -42,5 +42,5 @@ void test01()
   test &= itr == mapByName.end(); // { dg-error "no" } 
 }
  
-// { dg-error "candidates are" "" { target *-*-* } 211 }
-// { dg-error "candidates are" "" { target *-*-* } 215 }
+// { dg-error "candidates are" "" { target *-*-* } 213 }
+// { dg-error "candidates are" "" { target *-*-* } 217 }

diff -upr libstdc++-v3/testsuite/23_containers/set/operators/1_neg.cc libstdc++-v3/testsuite/23_containers/set/operators/1_neg.cc
--- libstdc++-v3/testsuite/23_containers/set/operators/1_neg.cc	2006-01-03 08:19:23.000000000 -0500
+++ libstdc++-v3/testsuite/23_containers/set/operators/1_neg.cc	2008-10-28 17:01:49.000000000 -0400
@@ -40,5 +40,5 @@ void test01()
   test &= itr == setByName.end(); // { dg-error "no" } 
 }
 
-// { dg-error "candidates are" "" { target *-*-* } 286 }
-// { dg-error "candidates are" "" { target *-*-* } 290 }
+// { dg-error "candidates are" "" { target *-*-* } 288 }
+// { dg-error "candidates are" "" { target *-*-* } 292 }
