--- /Network/Servers/hills/Volumes/capanna/gkeating/co/gcc-4_0-branch/gcc/libstdc++-v3/testsuite/lib/libstdc++.exp	2005-02-16 13:23:56.000000000 -0800
+++ libstdc++-v3/testsuite/lib/libstdc++.exp	2005-05-16 16:38:23.000000000 -0700
@@ -193,7 +193,14 @@ proc libstdc++_init { testfile } {
         } else {
             set cxx [transform "g++"]
             set cxxldflags ""
-            set cxxflags "-g -O2 -D_GLIBCXX_ASSERT -fmessage-length=0" 
+            # APPLE LOCAL begin libstdc++ testing
+            # For standalone libstdc++ testing, we need to provide all
+            # the necessary compiler flags, since we do not have a 
+            # testsuite_flags file to fall back on.
+            set cxxflags [concat "-g -O2 -D_GLIBCXX_ASSERT" \
+            " -ffunction-sections" \
+            " -fdata-sections -fmessage-length=0 -DAPPLE_KEYMGR"]
+            # APPLE LOCAL end libstdc++ testing
             set includes "-I${srcdir}/util"
         }
     }
