Index: configure.host
===================================================================
--- gcc-4.0.0/libstdc++-v3/configure.host 2006-10-12 14:30:38.000000000 -0400
+++ libstdc++-v3/configure.host    2006-10-12 16:03:08.000000000 -0400
@@ -202,6 +202,11 @@ case "${host_os}" in
     # On Darwin, performance is improved if libstdc++ is single-module,
     # and on 8+ compatibility is better if not -flat_namespace.
     OPT_LDFLAGS="${OPT_LDFLAGS} -Wl,-single_module"
+    case "${host_cpu}" in
+      i[34567]86 | x86_64)
+        OPTIMIZE_CXXFLAGS="${OPTIMIZE_CXXFLAGS} -fvisibility-inlines-hidden"
+        ;;
+    esac
     os_include_dir="os/bsd/darwin"
     ;;
   *djgpp*)      # leading * picks up "msdosdjgpp"
Index: acinclude.m4
===================================================================
--- gcc-4.0.0/libstdc++-v3/acinclude.m4      2006-10-11 12:57:13.000000000 -0400
+++ libstdc++-v3/acinclude.m4   2006-10-11 12:57:26.000000000 -0400
@@ -654,7 +654,6 @@ AC_DEFUN([GLIBCXX_EXPORT_FLAGS], [
   # Optimization flags that are probably a good idea for thrill-seekers. Just
   # uncomment the lines below and make, everything else is ready to go...
   # OPTIMIZE_CXXFLAGS = -O3 -fstrict-aliasing -fvtable-gc
-  OPTIMIZE_CXXFLAGS=
   AC_SUBST(OPTIMIZE_CXXFLAGS)
 
   WARN_FLAGS='-Wall -Wextra -Wwrite-strings -Wcast-qual'
Index: configure
===================================================================
--- libstdc++-v3/configure	2006-12-04 14:15:39.000000000 -0500
+++ libstdc++-v3/configure	2006-12-08 11:41:11.000000000 -0500
@@ -110246,7 +110246,6 @@ echo "${ECHO_T}$gxx_include_dir" >&6
   # Optimization flags that are probably a good idea for thrill-seekers. Just
   # uncomment the lines below and make, everything else is ready to go...
   # OPTIMIZE_CXXFLAGS = -O3 -fstrict-aliasing -fvtable-gc
-  OPTIMIZE_CXXFLAGS=
 
 
   WARN_FLAGS='-Wall -Wextra -Wwrite-strings -Wcast-qual'
