--- Makefile.orig	2008-11-24 16:18:37.000000000 -0800
+++ Makefile	2008-11-24 16:19:21.000000000 -0800
@@ -206,14 +206,14 @@
 	  case '$(am__configure_deps)' in \
 	    *$$dep*) \
 	      echo ' cd $(srcdir) && $(AUTOMAKE) --gnu '; \
-	      cd $(srcdir) && $(AUTOMAKE) --gnu  \
+         cd $(srcdir) && echo ' SKIPPING AUTOMAKE --gnu'  \
 		&& exit 0; \
 	      exit 1;; \
 	  esac; \
 	done; \
 	echo ' cd $(top_srcdir) && $(AUTOMAKE) --gnu  Makefile'; \
 	cd $(top_srcdir) && \
-	  $(AUTOMAKE) --gnu  Makefile
+	  echo ' SKIPPING AUTOMAKE --gnu  Makefile'
 .PRECIOUS: Makefile
 Makefile: $(srcdir)/Makefile.in $(top_builddir)/config.status
 	@case '$?' in \
@@ -229,7 +229,7 @@
 	$(SHELL) ./config.status --recheck
 
 $(top_srcdir)/configure:  $(am__configure_deps)
-	cd $(srcdir) && $(AUTOCONF)
+	cd $(srcdir) && echo ' SKIPPING AUTOCONF'
 $(ACLOCAL_M4):  $(am__aclocal_m4_deps)
 	cd $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
 
@@ -243,10 +243,10 @@
 	@rm -f stamp-h1
 	cd $(top_builddir) && $(SHELL) ./config.status config.h
 $(srcdir)/config.h.in:  $(am__configure_deps) 
-	cd $(top_srcdir) && $(AUTOHEADER)
-	rm -f stamp-h1
-	touch $@
-
+	cd $(top_srcdir) && echo ' SKIPPING AUTOHEADER'
+	echo 'SKIPPING rm -f stamp-h1'
+	echo 'SKIPPING touch $@'
+ 
 distclean-hdr:
 	-rm -f config.h stamp-h1
 Doxyfile: $(top_builddir)/config.status $(srcdir)/Doxyfile.in
