--- src/cmp.c	2002-04-05 12:37:31.000000000 -0800
+++ ../diffutils-2.8.1.pwc/src/cmp.c	2004-11-16 11:14:00.000000000 -0800
@@ -39,6 +39,12 @@
 # define hard_locale_LC_MESSAGES 0
 #endif
 
+#ifdef __APPLE__
+#include "get_compat.h"
+#else
+#define COMPAT_MODE(func, mode) 1
+#endif
+
 static char const authorship_msgid[] =
   N_("Written by Torbjorn Granlund and David MacKenzie.");
 
@@ -297,6 +303,7 @@
   /* If the files are links to the same inode and have the same file position,
      they are identical.  */
 
+	if(!COMPAT_MODE("bin/cmp", "unix2003")) //conformance tests expect cmp to access the file
   if (0 < same_file (&stat_buf[0], &stat_buf[1])
       && same_file_attributes (&stat_buf[0], &stat_buf[1])
       && file_position (0) == file_position (1))
