diff -r -u -N --exclude='*.orig' vim72.orig/src/Makefile vim72/src/Makefile
--- vim72.orig/src/Makefile	2008-08-09 07:50:04.000000000 -0700
+++ vim72/src/Makefile	2008-08-19 01:06:04.000000000 -0700
@@ -894,7 +894,7 @@
 
 ### DESTDIR	root of the installation tree.  This is prepended to the other
 #		directories.  This directory must exist.
-#DESTDIR  = ~/pkg/vim
+DESTDIR  = $(DSTROOT)
 
 ### Directory of the man pages
 MAN1DIR = /man1
diff -r -u -N --exclude='*.orig' vim72.orig/src/osdef.sh vim72/src/osdef.sh
--- vim72.orig/src/osdef.sh	2004-06-07 07:32:26.000000000 -0700
+++ vim72/src/osdef.sh	2008-08-19 01:06:04.000000000 -0700
@@ -11,6 +11,7 @@
 #
 # 31.10.95 jw.
 
+CC=cc
 if test -z "$CC"; then
   CC=cc
 fi
