--- /SourceCache/vim/vim-32/vim/src/edit.c	2006-05-07 04:48:51.000000000 -0700
+++ src/edit.c	2007-01-06 13:58:54.000000000 -0800
@@ -317,7 +317,7 @@ edit(cmdchar, startln, count)
     int		inserted_space = FALSE;     /* just inserted a space */
     int		replaceState = REPLACE;
     int		nomove = FALSE;		    /* don't move cursor on return */
-
+    int		orig_count = count;	    /* for autoindent calculation */
     /* Remember whether editing was restarted after CTRL-O. */
     did_restart_edit = restart_edit;
 
@@ -406,7 +406,7 @@ edit(cmdchar, startln, count)
 	    AppendCharToRedobuff(cmdchar);
 	    if (cmdchar == 'g')		    /* "gI" command */
 		AppendCharToRedobuff('I');
-	    else if (cmdchar == 'r')	    /* "r<CR>" command */
+	    else if (cmdchar == 'r' && !Unix2003_compat)	    /* "r<CR>" command */
 		count = 1;		    /* insert only one <CR> */
 	}
     }
@@ -1248,6 +1248,8 @@ doESCkey:
 		goto doESCkey;
 	    }
 #endif
+	    if (Unix2003_compat && (cmdchar == 'r') && (orig_count > 1))
+		did_ai = TRUE;
 	    if (ins_eol(c) && !p_im)
 		goto doESCkey;	    /* out of memory */
 	    auto_format(FALSE, FALSE);
