diff --git a/src/ex_cmds.c b/src/ex_cmds.c
index 677c859..4f983fc 100644
--- a/src/ex_cmds.c
+++ b/src/ex_cmds.c
@@ -2878,7 +2878,7 @@ check_overwrite(eap, buf, fname, ffname, other)
 	    swapname = makeswapname(fname, ffname, curbuf, dir);
 	    vim_free(dir);
 	    r = vim_fexists(swapname);
-	    if (r)
+	    if (r && !eap->forceit)
 	    {
 #if defined(FEAT_GUI_DIALOG) || defined(FEAT_CON_DIALOG)
 		if (p_confirm || cmdmod.confirm)
@@ -5511,6 +5511,7 @@ ex_global(eap)
 #ifdef FEAT_CLIPBOARD
 	start_global_changes();
 #endif
+	global_last_cmd = 1;
 	global_exe(cmd);
 #ifdef FEAT_CLIPBOARD
 	end_global_changes();
