diff -r -u -N --exclude='*.orig' vim72.orig/src/fileio.c vim72/src/fileio.c
--- vim72.orig/src/fileio.c	2008-08-06 04:01:03.000000000 -0700
+++ vim72/src/fileio.c	2008-08-19 01:06:05.000000000 -0700
@@ -2039,6 +2039,17 @@
 	    }
 	}
 
+	if (p_bf) {
+		/* beautify throws away all non-printable characters on input
+		   except tab, newline, and form-feed	*/
+		int i=0;
+		for (p = ptr; p < ptr + size; ++p) {
+		    if (isprint(*p) || *p == '\t' || *p == '\n' || *p == '\f') 
+			ptr[i++] = *p;
+		}
+		size=i;
+	}
+
 	/*
 	 * This loop is executed once for every character read.
 	 * Keep it fast!
