Index: src/globals.h
===================================================================
--- src/globals.h	(revision 57762)
+++ src/globals.h	(working copy)
@@ -11,6 +11,14 @@
  */
 
 /*
+ * Control over standard conformance
+ */
+
+EXTERN int Unix2003_compat;
+/* Copy of -w value for unix2003 compatible use */
+EXTERN int p_window_unix2003;
+
+/*
  * Number of Rows and Columns in the screen.
  * Must be long to be able to use them as options in option.c.
  * Note: Use screen_Rows and screen_Columns to access items in ScreenLines[].
@@ -1030,6 +1038,7 @@
 					    /* set after swap write error msg */
 EXTERN int	undo_off INIT(= FALSE);	    /* undo switched off for now */
 EXTERN int	global_busy INIT(= 0);	    /* set when :global is executing */
+EXTERN int	global_last_cmd INIT(= 0);  /* set when :global was last cmd */
 EXTERN int	listcmd_busy INIT(= FALSE); /* set when :argdo, :windo or
 					       :bufdo is executing */
 EXTERN int	need_start_insertmode INIT(= FALSE);
