diff --git a/src/ops.c b/src/ops.c
index 62d88a7..845b4cb 100644
--- a/src/ops.c
+++ b/src/ops.c
@@ -57,6 +57,7 @@ static struct yankreg
     char_u	**y_array;	/* pointer to array of line pointers */
     linenr_T	y_size;		/* number of lines in y_array */
     char_u	y_type;		/* MLINE, MCHAR or MBLOCK */
+    char_u	y_op_change;
     colnr_T	y_width;	/* only set if y_type == MBLOCK */
 } y_regs[NUM_REGISTERS];
 
@@ -1172,6 +1173,7 @@ stuff_yank(regname, p)
 	y_current->y_array[0] = p;
 	y_current->y_size = 1;
 	y_current->y_type = MCHAR;  /* used to be MLINE, why? */
+	y_current->y_op_change = 0;
     }
     return OK;
 }
@@ -3000,6 +3002,7 @@ op_yank(oap, deleting, mess)
 
     y_current->y_size = yanklines;
     y_current->y_type = yanktype;   /* set the yank register type */
+    y_current->y_op_change = (oap->op_type == OP_CHANGE) && inindent(0);
     y_current->y_width = 0;
     y_current->y_array = (char_u **)lalloc_clear((long_u)(sizeof(char_u *) *
 							    yanklines), TRUE);
@@ -3452,6 +3455,9 @@ do_put(regname, dir, count, flags)
 	y_type = y_current->y_type;
 	y_width = y_current->y_width;
 	y_size = y_current->y_size;
+	if (Unix2003_compat && y_size > 1 && y_type == MCHAR && curwin->w_cursor.col == 0 && y_current->y_op_change) {
+	    y_type = MLINE;
+	}
 	y_array = y_current->y_array;
     }
 
