diff -r -u -N --exclude='*.orig' vim72.orig/src/option.h vim72/src/option.h
--- vim72.orig/src/option.h	2008-06-22 05:55:09.000000000 -0700
+++ vim72/src/option.h	2008-08-19 01:06:07.000000000 -0700
@@ -312,6 +312,7 @@
 #if defined(FEAT_GUI) && defined(MACOS_X)
 EXTERN int	*p_antialias;	/* 'antialias' */
 #endif
+EXTERN int	p_ap;		/* 'autoprint' */
 EXTERN int	p_ar;		/* 'autoread' */
 EXTERN int	p_aw;		/* 'autowrite' */
 EXTERN int	p_awa;		/* 'autowriteall' */
@@ -340,6 +341,7 @@
 EXTERN char_u	*p_bexpr;
 # endif
 #endif
+EXTERN int	p_bf;		/* 'beautify' */
 #ifdef FEAT_BROWSE
 EXTERN char_u	*p_bsdir;	/* 'browsedir' */
 #endif
