--- src/os_unix.c.orig	2005-03-27 04:04:06.000000000 -0800
+++ src/os_unix.c	2005-05-12 11:33:45.000000000 -0700
@@ -291,6 +291,32 @@
     {-1,	    "Unknown!", FALSE}
 };
 
+/*
+ *  mch_am_i_owner(name): do I own the file (or am I root)
+ *  (used to check whether .exrc should be used or not)
+ */
+
+    int
+mch_am_i_owner(name)
+    char_u	*name;
+{
+    struct stat statb;
+    uid_t current_uid=getuid();
+
+    if (current_uid == 0)	/* Root always looks like owner */
+	return OK;
+    /* Keep the #ifdef outside of stat(), it may be a macro. */
+#ifdef VMS
+    if (stat((char *)vms_fixfilename(name), &statb))
+#else
+    if (stat((char *)name, &statb))
+#endif
+	return -1;
+    if (getuid() == statb.st_uid)
+	return OK;
+    return 0;
+}
+
     void
 mch_write(s, len)
     char_u	*s;
@@ -721,11 +747,6 @@
     if (signal_stack != NULL)
     {
 # ifdef HAVE_SIGALTSTACK
-#  ifdef __APPLE__
-	/* missing prototype.  Adding it to osdef?.h.in doesn't work, because
-	 * "struct sigaltstack" needs to be declared. */
-	extern int sigaltstack __ARGS((const struct sigaltstack *ss, struct sigaltstack *oss));
-#  endif
 
 #  ifdef HAVE_SS_BASE
 	sigstk.ss_base = signal_stack;
@@ -3361,6 +3382,10 @@
     if (columns <= 0 || rows <= 0)
 	return FAIL;
 
+    if (Unix2003_compat) {
+	/* Use the -w value specified on command line */
+	if (p_window_unix2003) rows = p_window_unix2003;
+    }
     Rows = rows;
     Columns = columns;
     return OK;
