diff --git a/src/if_python.c b/src/if_python.c
index 2ef0dfc..6cf0760 100644
--- a/src/if_python.c
+++ b/src/if_python.c
@@ -55,7 +55,7 @@
 
 #define PY_SSIZE_T_CLEAN
 
-#include <Python.h>
+#include <Python/Python.h>
 
 #if !defined(PY_VERSION_HEX) || PY_VERSION_HEX < 0x02050000
 # undef PY_SSIZE_T_CLEAN
diff --git a/src/if_ruby.c b/src/if_ruby.c
index 3767bf3..a58f525 100644
--- a/src/if_ruby.c
+++ b/src/if_ruby.c
@@ -106,7 +106,7 @@
 # define rb_check_type rb_check_type_stub
 #endif
 
-#include <ruby.h>
+#include <Ruby/ruby.h>
 #ifdef RUBY19_OR_LATER
 # include <ruby/encoding.h>
 #endif
