diff --git a/src/auto/config.h b/src/auto/config.h
index eb7d820..e574d9c 100644
--- a/src/auto/config.h
+++ b/src/auto/config.h
@@ -460,3 +460,9 @@
 
 /* Define if Xutf8SetWMProperties() is in an X library. */
 /* #undef HAVE_XUTF8SETWMPROPERTIES */
+
+/* Define if there is a copyfile() */
+#define HAVE_COPYFILE 1
+
+/* Define if there is a copyfile.h */
+#define HAVE_COPYFILE_H 1
diff --git a/src/fileio.c b/src/fileio.c
index 20c0294..4772ae1 100644
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -21,6 +21,10 @@
 # include <utime.h>		/* for struct utimbuf */
 #endif
 
+#if defined(HAVE_COPYFILE_H)
+#include <copyfile.h>
+#endif
+
 #define BUFSIZE		8192	/* size of normal write buffer */
 #define SMBUFSIZE	256	/* size of emergency write buffer */
 
@@ -3199,6 +3203,9 @@ buf_write(buf, fname, sfname, start, end, eap, append, forceit,
 #ifdef FEAT_CRYPT
     write_info.bw_buffer = buf;
 #endif
+#ifdef HAVE_COPYFILE
+    copyfile_state_t	copyfile_state = NULL;
+#endif
 
     /* After writing a file changedtick changes but we don't want to display
      * the line. */
@@ -3624,6 +3631,13 @@ buf_write(buf, fname, sfname, start, end, eap, append, forceit,
     if (!newfile)
 	acl = mch_get_acl(fname);
 #endif
+#ifdef HAVE_COPYFILE
+    if (!newfile && copyfile((char*)fname, NULL, 0, COPYFILE_XATTR | COPYFILE_CHECK))
+    {
+	copyfile_state = copyfile_state_alloc();
+	copyfile((char*)fname, NULL, copyfile_state, 0);
+    }
+#endif
 
     /*
      * If 'backupskip' is not empty, don't make a backup for some files.
@@ -4006,6 +4020,10 @@ buf_write(buf, fname, sfname, start, end, eap, append, forceit,
 #ifdef HAVE_ACL
 			mch_set_acl(backup, acl);
 #endif
+ #ifdef HAVE_COPYFILE
+			if (copyfile_state)
+			copyfile(NULL, (char*)backup, copyfile_state, COPYFILE_XATTR);
+#endif
 #if defined(HAVE_SELINUX) || defined(HAVE_SMACK)
 			mch_copy_sec(fname, backup);
 #endif
@@ -4686,6 +4704,10 @@ restore_backup:
 	/* Set the inode when creating a new file. */
 	buf_setino(buf);
 #endif
+#ifdef HAVE_COPYFILE
+    if (!backup_copy && copyfile_state)
+	copyfile(NULL, (char*)wfname, copyfile_state, COPYFILE_XATTR);
+#endif
 
     if (close(fd) != 0)
     {
@@ -5009,6 +5031,10 @@ nofail:
 #ifdef HAVE_ACL
     mch_free_acl(acl);
 #endif
+#ifdef HAVE_COPYFILE
+    if (copyfile_state)
+	copyfile_state_free(copyfile_state);
+#endif
 
     if (errmsg != NULL)
     {
