diff -r -u -N --exclude='*.orig' vim72.orig/src/normal.c vim72/src/normal.c
--- vim72.orig/src/normal.c	2008-07-31 13:03:08.000000000 -0700
+++ vim72/src/normal.c	2008-08-19 01:14:15.000000000 -0700
@@ -6412,18 +6412,23 @@
 	 * Imitate strange Vi behaviour: When using "]]" with an operator
 	 * we also stop at '}'.
 	 */
-	if (!findpar(&cap->oap->inclusive, cap->arg, cap->count1, flag,
-	      (cap->oap->op_type != OP_NOP
-				      && cap->arg == FORWARD && flag == '{')))
-	    clearopbeep(cap->oap);
-	else
-	{
-	    if (cap->oap->op_type == OP_NOP)
-		beginline(BL_WHITE | BL_FIX);
+	if (curbuf->b_p_lisp) {
+	    findpar(cap->oap, cap->arg, cap->count1, 
+	      '(', FALSE);
+	} else {
+	    if (!findpar(&cap->oap->inclusive, cap->arg, cap->count1, flag,
+		  (cap->oap->op_type != OP_NOP
+					  && cap->arg == FORWARD && flag == '{')))
+		clearopbeep(cap->oap);
+	    else
+	    {
+		if (cap->oap->op_type == OP_NOP)
+		    beginline(BL_WHITE | BL_FIX);
 #ifdef FEAT_FOLDING
-	    if ((fdo_flags & FDO_BLOCK) && KeyTyped && cap->oap->op_type == OP_NOP)
-		foldOpenCursor();
+		if ((fdo_flags & FDO_BLOCK) && KeyTyped && cap->oap->op_type == OP_NOP)
+		    foldOpenCursor();
 #endif
+	    }
 	}
     }
 
@@ -6593,6 +6598,11 @@
 nv_brace(cap)
     cmdarg_T	*cap;
 {
+    if (curbuf->b_p_lisp) {
+	cap->nchar = '(';
+	nv_csearch(cap);
+	return;
+    }
     cap->oap->motion_type = MCHAR;
     cap->oap->use_reg_one = TRUE;
     /* The motion used to be inclusive for "(", but that is not what Vi does. */
diff -r -u -N --exclude='*.orig' vim72.orig/src/search.c vim72/src/search.c
--- vim72.orig/src/search.c	2008-07-13 11:18:27.000000000 -0700
+++ vim72/src/search.c	2008-08-19 01:14:15.000000000 -0700
@@ -2772,6 +2772,10 @@
     char_u	*s;
 
     s = ml_get(lnum);
+    if (para == '(')
+	return *s == '(';
+    if (para == ')')
+	return *s == ')';
     if (*s == para || *s == '\f' || (both && *s == '}'))
 	return TRUE;
     if (*s == '.' && (inmacro(p_sections, s + 1) ||
