Index: src/option.c
===================================================================
--- src/option.c	(revision 57762)
+++ src/option.c	(working copy)
@@ -2059,7 +2059,7 @@
 #endif
 			    SCRIPTID_INIT},
     {"readonly",    "ro",   P_BOOL|P_VI_DEF|P_RSTAT|P_NOGLOB,
-			    (char_u *)&p_ro, PV_RO,
+			    (char_u *)&readonlymode, PV_RO,
 			    {(char_u *)FALSE, (char_u *)0L} SCRIPTID_INIT},
     {"redraw",	    NULL,   P_BOOL|P_VI_DEF,
 			    (char_u *)NULL, PV_NONE,
