diff -r -u -N --exclude='*.orig' vim72.orig/src/undo.c vim72/src/undo.c
--- vim72.orig/src/undo.c	2008-02-13 06:17:54.000000000 -0800
+++ vim72/src/undo.c	2008-08-19 01:17:51.000000000 -0700
@@ -1839,6 +1839,10 @@
     t = curbuf->b_u_line_colnr;
     if (curwin->w_cursor.lnum == curbuf->b_u_line_lnum)
 	curbuf->b_u_line_colnr = curwin->w_cursor.col;
+    if (Unix2003_compat) {
+        /* vi_05 test 276: "U" sets column to start of line */
+        t = 0;
+    } 
     curwin->w_cursor.col = t;
     curwin->w_cursor.lnum = curbuf->b_u_line_lnum;
     check_cursor_col();
