diff -r -u -N --exclude='*.orig' vim72.orig/src/xxd/xxd.c vim72/src/xxd/xxd.c
--- vim72.orig/src/xxd/xxd.c	2007-11-29 12:05:16.000000000 -0800
+++ vim72/src/xxd/xxd.c	2008-08-19 01:20:30.000000000 -0700
@@ -502,7 +502,7 @@
 	}
       else if (!STRNCMP(pp, "-g", 2))
 	{
-	  if (pp[2] && STRNCMP("group", pp + 2, 5))
+	  if (pp[2] && STRNCMP("roupsize", pp + 2, 8))
 	    octspergrp = (int)strtol(pp + 2, NULL, 0);
 	  else
 	    {
