--- /tmp/jabberd-2.2.14/storage/Makefile.am	2011-05-31 15:11:37.000000000 -0700
+++ ./jabberd2/storage/Makefile.am	2011-06-22 20:13:00.000000000 -0700
@@ -107,3 +107,8 @@ storage_sqlite_la_LDFLAGS = $(MODULE_LDF
 storage_sqlite_la_LIBADD  = $(MODULE_LIBADD) $(SQLITE_LIBS)
 endif
 
+if STORAGE_APPLE
+pkglib_LTLIBRARIES += authreg_apple_od.la
+authreg_apple_od_la_SOURCES = authreg_apple_od.c
+authreg_apple_od_la_LDFLAGS = $(APPLE_OD_LIBS)
+endif
\ No newline at end of file
