diff -Naur ../lukemftpd.roots/lukemftpd/tnftpd/src/ftpd.c ./src/ftpd.c
--- ../lukemftpd.roots/lukemftpd/tnftpd/src/ftpd.c	2005-01-26 15:27:23.000000000 -0800
+++ ./src/ftpd.c	2005-01-26 15:36:18.000000000 -0800
@@ -3379,6 +3379,9 @@
 
 	reply = calloc(num_msg, sizeof(struct pam_response));
 	if( reply == NULL )
+		return PAM_BUF_ERR;
+
+	if( mystuff == NULL )
 		return PAM_CONV_ERR;
 
 	reply[0].resp = xstrdup(mystuff);
