--- tnftpd-20091122/src/ftpd.c	2009-11-06 19:26:21.000000000 -0800
+++ tnftpd/src/ftpd.c	2010-01-28 17:03:18.000000000 -0800
@@ -1263,7 +1263,7 @@
 					    "Local resource failure: realloc");
 				groups = ng;
 			} while (getgrouplist(pw->pw_name, pw->pw_gid,
-						groups, &gsize) == -1);
+						(int*)groups, &gsize) == -1);
 			found = 0;
 			for (i = 0; i < gsize; i++) {
 				struct group *g;
