--- /tmp/jabberd-2.2.13/sx/plugins.h	2011-02-23 08:24:34.000000000 -0800
+++ ./jabberd2/sx/plugins.h	2011-02-24 16:25:21.000000000 -0800
@@ -63,10 +63,10 @@ extern "C" {
 JABBERD2_API int                         sx_ssl_init(sx_env_t env, sx_plugin_t p, va_list args);
 
 /** add cert function */
-JABBERD2_API int                         sx_ssl_server_addcert(sx_plugin_t p, char *name, char *pemfile, char *cachain, int mode);
+JABBERD2_API int                         sx_ssl_server_addcert(sx_plugin_t p, char *name, char *pemfile, char *cachain, int mode, char *private_key_password);
 
 /** trigger for client starttls */
-JABBERD2_API int                         sx_ssl_client_starttls(sx_plugin_t p, sx_t s, char *pemfile);
+JABBERD2_API int                         sx_ssl_client_starttls(sx_plugin_t p, sx_t s, char *pemfile, char *private_key_password);
 
 /* previous states */
 #define SX_SSL_STATE_NONE       (0)
@@ -89,6 +89,8 @@ typedef struct _sx_ssl_conn_st {
     int         last_state;
 
     char        *pemfile;
+
+    char        *private_key_password;
 } *_sx_ssl_conn_t;
 
 #endif /* HAVE_SSL */
