--- /tmp/jabberd-2.2.13/sx/sasl.h	2011-02-23 08:24:34.000000000 -0800
+++ ./jabberd2/sx/sasl.h	2011-02-24 16:25:21.000000000 -0800
@@ -30,5 +30,6 @@
 #define _sasl_err_MECH_TOO_WEAK         "mechanism-too-weak"
 #define _sasl_err_NOT_AUTHORIZED        "not-authorized"
 #define _sasl_err_TEMPORARY_FAILURE     "temporary-auth-failure"
+#define _sasl_err_INTERNAL_SERVER_ERROR  "internal-server-error"
 
 #endif
