--- /tmp/jabberd-2.2.13/sx/sx.c	2011-02-23 08:24:34.000000000 -0800
+++ ./jabberd2/sx/sx.c	2011-03-01 14:21:15.000000000 -0800
@@ -155,6 +155,7 @@ void _sx_reset(sx_t s) {
     temp.cb_arg = s->cb_arg;
 
     temp.ip = s->ip;
+    temp.port = s->port;
     temp.flags = s->flags;
     temp.reentry = s->reentry;
     temp.ssf = s->ssf;
@@ -183,6 +184,7 @@ void _sx_reset(sx_t s) {
 
     s->env = temp.env;
     s->ip = temp.ip;
+    s->port = temp.port;
     s->flags = temp.flags;
     s->reentry = temp.reentry;
     s->ssf = temp.ssf;
