--- framework/src/objc/RBRuntime.m.orig	2008-04-01 13:07:26.000000000 -0700
+++ framework/src/objc/RBRuntime.m	2008-04-01 13:07:59.000000000 -0700
@@ -906,12 +906,17 @@
 
 static void RBCocoaInstallRubyThreadSchedulerHooks()
 {
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_6
+  /* The threading support is not implemented yet in 10.6. */
+  return;
+#else
   if (getenv("RUBYCOCOA_THREAD_HOOK_DISABLE") != NULL) {
     if (rb_cocoa_thread_debug) {
       NSLog(@"RBCocoaInstallRubyThreadSchedulerHooks: warning: disabled hooks due to RUBYCOCOA_THREAD_HOOK_DISABLE environment variable");
     }
     return;
   }
+#endif
   
   rb_cocoa_thread_debug = getenv("RUBYCOCOA_THREAD_DEBUG") != NULL;
   
