Index: ext/rubycocoa/extconf.rb.in
===================================================================
--- ext/rubycocoa/extconf.rb.in	(revision 2293)
+++ ext/rubycocoa/extconf.rb.in	(working copy)
@@ -12,7 +12,7 @@
   $stderr.puts "execute '#{cmd}' done"
 end
 
-$CFLAGS = '-F../../framework/%%%build_dir%%%'
+$CFLAGS = "-F../../framework/%%%build_dir%%% #{ENV['RC_CFLAGS']}"
 $LDFLAGS  = '-F../../framework/%%%build_dir%%% -framework RubyCocoa'
 
 $CFLAGS << ' %%%other_cflags%%% %%%arch_flags%%%'
