--- ext/rubycocoa/extconf.rb.in.old    2006-07-26 18:39:16.000000000 -0700
+++ ext/rubycocoa/extconf.rb.in    2006-07-26 18:38:52.000000000 -0700
@@ -15,7 +15,7 @@
 $CFLAGS = '-F../../framework/%%%build_dir%%%'
 $LDFLAGS  = '-F../../framework/%%%build_dir%%% -framework %%%framework_name%%%'
 
-$CFLAGS << ' %%%other_cflags%%%'
+$CFLAGS << " %%%other_cflags%%% #{ENV['RC_CFLAGS']}"
 $LDFLAGS << ' %%%other_ldflags%%%'
 
 create_makefile('rubycocoa')
