--- config.h.in-orig    2005-02-15 07:19:29.000000000 -0800
+++ config.h.in 2005-02-15 07:19:47.000000000 -0800
@@ -1019,7 +1019,7 @@
 
 /* Define to 1 if your processor stores words with the most significant byte
    first (like Motorola and SPARC, unlike Intel and VAX). */
-#undef WORDS_BIGENDIAN
+#define WORDS_BIGENDIAN 1
 
 /* Number of bits in a file offset, on hosts where this is settable. */
 #undef _FILE_OFFSET_BITS

