*** configure.in.orig	Thu Feb 19 18:44:04 2004
--- configure.in	Thu Feb 19 18:40:14 2004
***************
*** 1023,1028 ****
--- 1023,1037 ----
        MAX_C_OPTIMIZE="-O"
      fi
      ;;
+   *darwin[7-8]*)
+     if test "$ac_cv_prog_gcc" = "yes"
+     then
+       FLAGS="-DHAVE_DARWIN_THREADS -D_P1003_1B_VISIBLE -DSIGNAL_WITH_VIO_CLOSE -DSIGNALS_DONT_BREAK_READ -DFN_NO_CASE_SENCE"
+       CFLAGS="$CFLAGS $FLAGS"
+       CXXFLAGS="$CXXFLAGS $FLAGS"
+       MAX_C_OPTIMIZE="-O"
+     fi
+     ;;
    *freebsd*)
      echo "Adding fix for interrupted reads"
      OSVERSION=`sysctl -a | grep osreldate | awk '{ print $2 }'`
