*** configure.orig	Wed Feb 18 15:17:23 2004
--- configure	Wed Feb 18 13:40:32 2004
***************
*** 23703,23708 ****
--- 23703,23717 ----
        MAX_C_OPTIMIZE="-O"
      fi
      ;;
+   *darwin[7-8]*)
+     if test "$ac_cv_c_compiler_gnu" = "yes"
+     then
+       FLAGS="-DHAVE_DARWIN_THREADS -D_P1003_1B_VISIBLE -DSIGNAL_WITH_VIO_CLOSE -DSIGNALS_DONT_BREAK_READ -DFN_NO_CASE_SENCE"
+       CFLAGS="$CFLAGS $FLAGS"
+       CXXFLAGS="$CXXFLAGS $FLAGS"
+       MAX_C_OPTIMIZE="-O"
+     fi
+     ;;
    *freebsd*)
      echo "Adding fix for interrupted reads"
      OSVERSION=`sysctl -a | grep osreldate | awk '{ print $2 }'`
