--- compress.c.orig	2003-07-06 22:37:56.000000000 -0700
+++ compress.c	2006-08-31 10:53:18.000000000 -0700
@@ -31,8 +31,8 @@ int ZEXPORT compress2 (dest, destLen, so
 
     stream.next_in = (Bytef*)source;
     stream.avail_in = (uInt)sourceLen;
-#ifdef MAXSEG_64K
-    /* Check for source > 64K on 16-bit machine: */
+#if defined(__LP64__) || defined(MAXSEG_64K)
+    /* Check for source > 4G under LP64 or > 64K on 16-bit machine: */
     if ((uLong)stream.avail_in != sourceLen) return Z_BUF_ERROR;
 #endif
     stream.next_out = dest;
