--- win32/Makefile.msc.orig	2003-08-26 21:42:00.000000000 -0700
+++ win32/Makefile.msc	2008-11-07 12:42:57.000000000 -0800
@@ -13,23 +13,32 @@
 #   nmake -f win32/Makefile.msc LOC=-DASMV OBJA=match.obj  (use ASM code)
 
 
-# optional build flags
-LOC =
 
+CC = cl
+AS = ml
+LD = link
+AR = lib
+RC = rc
 
+# optional build flags
+!IFDEF DEBUG
+# variables
+STATICLIB = zlib_debug.lib
+SHAREDLIB = zlib1_debug.dll
+IMPLIB    = zdll_debug.lib
+CFLAGS  = -nologo /MDd /Od /Zi /GS /D "DEBUG" /D "_DEBUG" $(LOC)
+LDFLAGS = -nologo -debug /SAFESEH /FIXED:NO /dynamicbase
+!ELSE
 # variables
 STATICLIB = zlib.lib
 SHAREDLIB = zlib1.dll
 IMPLIB    = zdll.lib
+CFLAGS  = -nologo /MD /O2 /GS $(LOC)
+LDFLAGS = -nologo -release /debug /SAFESEH /FIXED:NO /dynamicbase
+!ENDIF
+
 
-CC = cl
-AS = ml
-LD = link
-AR = lib
-RC = rc
-CFLAGS  = -nologo -MD -O2 $(LOC)
 ASFLAGS = -coff
-LDFLAGS = -nologo -release
 ARFLAGS = -nologo
 RCFLAGS = /dWIN32 /r
 
