--- Makefile.in.orig	Tue Dec 28 22:15:58 2004
+++ Makefile.in	Tue Dec 28 22:16:44 2004
@@ -77,8 +77,13 @@
 	mv _match.o match.o
 	rm -f _match.s
 
+SECTORDER_FILE=/usr/local/lib/OrderFiles/libz.1.order
+ifeq "$(shell test -f $(SECTORDER_FILE) && echo YES)" "YES"
+	SECTORDER_FLAGS=-sectorder __TEXT __text $(SECTORDER_FILE)
+endif
+
 $(SHAREDLIBV): $(OBJS)
-	$(LDSHARED) -o $@ $(OBJS)
+	$(LDSHARED) -exported_symbols_list $(SRCROOT)/libz.exp $(SECTORDER_FLAGS) -o $@ $(OBJS)
 	rm -f $(SHAREDLIB) $(SHAREDLIBM)
 	ln -s $@ $(SHAREDLIB)
 	ln -s $@ $(SHAREDLIBM)
