--- zconf.in.h.orig	Tue Dec 28 22:13:16 2004
+++ zconf.in.h	Tue Dec 28 22:13:42 2004
@@ -284,16 +284,17 @@
 #  ifdef VMS
 #    include <unixio.h>   /* for off_t */
 #  endif
-#  define z_off_t off_t
 #endif
 #ifndef SEEK_SET
 #  define SEEK_SET        0       /* Seek from beginning of file.  */
 #  define SEEK_CUR        1       /* Seek from current position.  */
 #  define SEEK_END        2       /* Set file pointer to EOF plus "offset" */
 #endif
-#ifndef z_off_t
-#  define z_off_t long
-#endif
+
+/* Hardcode this value since zlib doesn't actually support it being other
+ * than a long.
+ */
+#define  z_off_t long
 
 #if defined(__OS400__)
 #  define NO_vsnprintf
