--- /tmp/jabberd-2.2.14/tools/db-setup.sqlite	2011-05-31 15:11:37.000000000 -0700
+++ ./jabberd2/tools/db-setup.sqlite	2011-06-22 20:13:00.000000000 -0700
@@ -52,6 +52,11 @@ CREATE TABLE "roster-items" (
 
 CREATE INDEX i_rosteri_owner ON "roster-items"("collection-owner");
 
+---
+--- Apple index to speed up jabber_autobuddy operations (still necessary in jabberd 2.2, with i_rosteri_owner added?)
+---
+CREATE INDEX "roster-items-owner-jid-index" on "roster-items" ("collection-owner", "jid");
+
 --
 -- Roster groups
 -- Used by: mod_roster
@@ -149,6 +154,8 @@ CREATE TABLE "vcard" (
     
     "rev" TEXT );
 
+CREATE INDEX i_vcard_owner ON "vcard"("collection-owner");
+
 --
 -- Offline message queue
 -- Used by: mod_offline
