--- /tmp/jabberd-2.2.13/tools/Makefile.am	2011-02-23 08:24:34.000000000 -0800
+++ ./jabberd2/tools/Makefile.am	2011-02-24 16:25:21.000000000 -0800
@@ -9,7 +9,8 @@ EXTRA_DIST = db-setup.mysql db-update.my
 edit = sed \
 	-e 's,@sysconfdir\@,$(sysconfdir),g' \
 	-e 's,@VERSION\@,$(VERSION),g' \
-	-e 's,@bindir\@,$(bindir),g'
+	-e 's,@bindir\@,$(bindir),g' \
+	-e 's,@libexecdir\@,$(libexecdir),g'
 
 $(bin_SCRIPTS): $(EXTRA_DIST)
 	@echo "generating $@ from $@.in"; \
