--- /tmp/jabberd-2.2.13/tools/Makefile.in	2011-02-23 08:24:52.000000000 -0800
+++ ./jabberd2/tools/Makefile.in	2011-02-24 16:25:21.000000000 -0800
@@ -207,7 +207,8 @@ EXTRA_DIST = db-setup.mysql db-update.my
 edit = sed \
 	-e 's,@sysconfdir\@,$(sysconfdir),g' \
 	-e 's,@VERSION\@,$(VERSION),g' \
-	-e 's,@bindir\@,$(bindir),g'
+	-e 's,@bindir\@,$(bindir),g' \
+	-e 's,@libexecdir\@,$(libexecdir),g'
 
 all: all-am
 
