--- /tmp/jabberd-2.2.14/tools/Makefile.in	2011-05-31 15:11:49.000000000 -0700
+++ ./jabberd2/tools/Makefile.in	2011-06-22 20:13:00.000000000 -0700
@@ -207,7 +207,8 @@ EXTRA_DIST = db-setup.mysql db-update.my
 edit = sed \
 	-e 's,@sysconfdir\@,$(sysconfdir),g' \
 	-e 's,@VERSION\@,$(VERSION),g' \
-	-e 's,@bindir\@,$(bindir),g'
+	-e 's,@bindir\@,$(bindir),g' \
+	-e 's,@libexecdir\@,$(libexecdir),g'
 
 all: all-am
 
