diff -uNr ../openssl-0.9.7i.orig/doc/crypto/bn.pod ./doc/crypto/bn.pod
--- ../openssl-0.9.7i.orig/doc/crypto/bn.pod	2001-09-03 05:58:16.000000000 -0700
+++ ./doc/crypto/bn.pod	2006-09-29 11:49:05.000000000 -0700
@@ -121,9 +121,9 @@
 
 =head1 DESCRIPTION
 
-This library performs arithmetic operations on integers of arbitrary
-size. It was written for use in public key cryptography, such as RSA
-and Diffie-Hellman.
+The Big Number library is part of libcrypto.  It performs arithmetic operations
+on integers of arbitrary size. It was written for use in public key cryptography,
+such as RSA and Diffie-Hellman.
 
 It uses dynamic memory allocation for storing its data structures.
 That means that there is no limit on the size of the numbers
