--- macosx/Makefile.orig	2004-03-01 09:50:20.000000000 -0800
+++ macosx/Makefile	2005-06-22 15:12:18.000000000 -0700
@@ -84,6 +84,9 @@
 
 PROJECT			:= tcl
 PRODUCT_NAME		:= Tcl
+_nothing		:=
+# need a leading space before the '1'
+ALT_VERSION		:= ${_nothing} 1
 
 UNIX_DIR		:= ${CURDIR}/../unix
 GENERIC_DIR		:= ${CURDIR}/../generic
@@ -211,7 +214,7 @@
 		<key>CFBundleExecutable</key>\n\
 		<string>Tcl</string>\n\
 		<key>CFBundleGetInfoString</key>\n\
-		<string>Tcl Library ${PRODUCT_VERSION}, Copyright © ${YEAR} Tcl Core Team.\n\
+		<string>Tcl Library ${PRODUCT_VERSION}${ALT_VERSION}, Copyright © ${YEAR} Tcl Core Team.\n\
 	MacOS X Port by Jim Ingham &lt;jingham@apple.com&gt; &amp; Ian Reid, Copyright\
 	© 2001-2002, Apple Computer, Inc.</string>\n\
 		<key>CFBundleIdentifier</key>\n\
@@ -219,15 +222,15 @@
 		<key>CFBundleInfoDictionaryVersion</key>\n\
 		<string>6.0</string>\n\
 		<key>CFBundleName</key>\n\
-		<string>Tcl Library ${PRODUCT_VERSION}</string>\n\
+		<string>Tcl Library ${PRODUCT_VERSION}${ALT_VERSION}</string>\n\
 		<key>CFBundlePackageType</key>\n\
 		<string>FMWK</string>\n\
 		<key>CFBundleShortVersionString</key>\n\
-		<string>${PRODUCT_LONGVERSION}</string>\n\
+		<string>${PRODUCT_LONGVERSION}${ALT_VERSION}</string>\n\
 		<key>CFBundleSignature</key>\n\
 		<string>Tcl </string>\n\
 		<key>CFBundleVersion</key>\n\
-		<string>${PRODUCT_LONGVERSION}</string>\n\
+		<string>${PRODUCT_LONGVERSION}${ALT_VERSION}</string>\n\
 	</dict>\n\
 	</plist>\n'
 
