--- perl.original.c	2019-04-29 13:35:18.000000000 -0700
+++ perl.c	2019-05-01 17:34:24.000000000 -0700
@@ -4242,9 +4242,23 @@
 void
 Perl_init_debugger(pTHX)
 {
+    int fd;
     HV * const ostash = PL_curstash;
     MAGIC *mg;
 
+    fd = PerlIO_fileno(PerlIO_stderr());
+    if (PerlLIO_isatty(fd)) {
+        PerlIO *out = PerlIO_stderr();
+
+        PerlIO_printf(out,
+                      "\nWARNING: This version of perl is included in macOS "
+                      "for compatibility with legacy software. "
+                      "\nIn future versions of "
+                      "macOS the perl runtime will not be available by "
+                      "\ndefault, and may require you to install an "
+                      "additional package.\n\n");
+    }
+
     PL_curstash = (HV *)SvREFCNT_inc_simple(PL_debstash);
 
     Perl_init_dbargs(aTHX);
