Index: ODBC.m4
===================================================================
RCS file: /cvsroot/tclodbc/tclodbc/ODBC.m4,v
retrieving revision 1.7
diff -u -p -r1.7 ODBC.m4
--- ODBC.m4	19 Mar 2004 20:32:21 -0000	1.7
+++ ODBC.m4	2 Jun 2004 05:21:19 -0000
@@ -102,46 +102,6 @@ AC_DEFUN(TCLODBC_PATH_ODBCH, [
     fi
 
     dnl ===============================
-    dnl unixODBC or Intersolv... I have tested this
-    dnl ===============================
-    dnl Check for unixODBC headers
-    if test x"${ac_cv_c_odbch}" = x ; then
-	for i in \
-		/opt/odbc/include \
-		/opt/ODBC/include \
-		${prefix}/include/odbc \
-		${prefix}/include/ODBC \
-		/opt/unixodbc/include \
-		/opt/unixODBC/include \
-		${prefix}/include/unixodbc \
-		${prefix}/include/unixODBC \
-		/usr/local/mysql/include \
-		/usr/local/include \
-		/usr/include \
-		${prefix}/include ; do
-	    if test -f "$i/sql.h"; then
-		ac_cv_c_odbch=`(cd $i; pwd)`
-		ODBCTYPE=unixODBC
-	    fi
-	done
-    fi
-
-    dnl see if one is installed
-    if test x"${ac_cv_c_odbch}" = x ; then
-	AC_CHECK_HEADER(sql.h, ac_cv_c_odbch=installed, ac_cv_c_odbch="")
-	if test x"${ac_cv_c_odbch}" != x ; then
-	    ODBCTYPE=unixODBC
-	fi
-    fi
-
-    dnl Figure out whether this is unixODBC or Intersolv
-    if test x"${ODBCTYPE}" = xunixODBC ; then
-	if test -f "${ac_cv_c_odbch}/odbcver.h" ; then
-	    ODBCTYPE=intersolv
-	fi
-    fi
-
-    dnl ===============================
     dnl IODBC... I have not tested this a whole lot
     dnl ===============================
     dnl Check in the source tree
@@ -159,6 +119,8 @@ AC_DEFUN(TCLODBC_PATH_ODBCH, [
 		/usr/local/include/iodbc \
 		/usr/local/share/iodbc \
 		${prefix}/include/iodbc \
+		/usr/local/include \
+		/usr/include \
 		${prefix}/include ; do
 	    if test -f $i/isql.h; then
 		ac_cv_c_odbch=`(cd $i; pwd)`
@@ -186,6 +148,46 @@ AC_DEFUN(TCLODBC_PATH_ODBCH, [
 	fi
     fi
 
+    dnl ===============================
+    dnl unixODBC or Intersolv... I have tested this
+    dnl ===============================
+    dnl Check for unixODBC headers
+    if test x"${ac_cv_c_odbch}" = x ; then
+	for i in \
+		/opt/odbc/include \
+		/opt/ODBC/include \
+		${prefix}/include/odbc \
+		${prefix}/include/ODBC \
+		/opt/unixodbc/include \
+		/opt/unixODBC/include \
+		${prefix}/include/unixodbc \
+		${prefix}/include/unixODBC \
+		/usr/local/mysql/include \
+		/usr/local/include \
+		/usr/include \
+		${prefix}/include ; do
+	    if test -f "$i/sql.h"; then
+		ac_cv_c_odbch=`(cd $i; pwd)`
+		ODBCTYPE=unixODBC
+	    fi
+	done
+    fi
+
+    dnl see if one is installed
+    if test x"${ac_cv_c_odbch}" = x ; then
+	AC_CHECK_HEADER(sql.h, ac_cv_c_odbch=installed, ac_cv_c_odbch="")
+	if test x"${ac_cv_c_odbch}" != x ; then
+	    ODBCTYPE=unixODBC
+	fi
+    fi
+
+    dnl Figure out whether this is unixODBC or Intersolv
+    if test x"${ODBCTYPE}" = xunixODBC ; then
+	if test -f "${ac_cv_c_odbch}/odbcver.h" ; then
+	    ODBCTYPE=intersolv
+	fi
+    fi
+
     ODBC_INCLUDE_DIR=""
     if test x"${ac_cv_c_odbch}" = x ; then
 	AC_MSG_ERROR([Cannot find any ODBC headers])
@@ -250,7 +252,7 @@ AC_DEFUN(TCLODBC_PATH_ODBCLIB, [
     if test x"${ODBCTYPE}" = xiodbc ; then
 	liblist="libpsqlodbc libmysqlodbc librbodbc libiodbc iodbc"
 	libpathlist="/usr/local/lib/iodbc ${prefix}/iodbc/lib \
-	    ${prefix}/lib/iodbc /usr/local/mysql/lib /usr/local/lib ${prefix}/lib"
+	    ${prefix}/lib/iodbc /usr/local/mysql/lib /usr/local/lib /usr/lib ${prefix}/lib"
     elif test x"${ODBCTYPE}" = xunixODBC ; then 
 	liblist="libodbc"
 	libpathlist="/opt/unixodbc/lib /opt/unixODBC/lib /usr/include/odbc \
@@ -326,6 +328,7 @@ AC_DEFUN(TCLODBC_PATH_ODBCLIB, [
     if test x"${ODBCTYPE}" = xiodbc ; then
 	AC_MSG_RESULT(Using IODBC driver)
 	AC_DEFINE(HAVE_IODBC)
+	ODBC_LIB="$ODBC_LIB -liodbcinst"
     fi
 
     if test x"${ODBCTYPE}" = xunixODBC ; then
Index: configure
===================================================================
RCS file: /cvsroot/tclodbc/tclodbc/configure,v
retrieving revision 1.4
diff -u -p -r1.4 configure
--- configure	20 Mar 2004 22:14:18 -0000	1.4
+++ configure	2 Jun 2004 05:21:19 -0000
@@ -6653,40 +6653,51 @@ echo "$as_me: error: ${with_odbcinclude}
 	fi
     fi
 
-                    if test x"${ac_cv_c_odbch}" = x ; then
+                    for i in $dirlist; do
+	if test -f "${srcdir}/$i/postodbc/isql.h" ; then
+	    ac_cv_c_odbch=`(cd ${srcdir}/$i/postodbc/; pwd)`
+	    ODBCTYPE=iodbc
+	fi
+    done
+
+        if test x"${ac_cv_c_odbch}" = x ; then
 	for i in \
-		/opt/odbc/include \
-		/opt/ODBC/include \
-		${prefix}/include/odbc \
-		${prefix}/include/ODBC \
-		/opt/unixodbc/include \
-		/opt/unixODBC/include \
-		${prefix}/include/unixodbc \
-		${prefix}/include/unixODBC \
 		/usr/local/mysql/include \
+		/usr/local/include/iodbc \
+		/usr/local/share/iodbc \
+		${prefix}/include/iodbc \
 		/usr/local/include \
 		/usr/include \
 		${prefix}/include ; do
-	    if test -f "$i/sql.h"; then
+	    if test -f $i/isql.h; then
 		ac_cv_c_odbch=`(cd $i; pwd)`
-		ODBCTYPE=unixODBC
+		ODBCTYPE=iodbc
 	    fi
 	done
     fi
 
         if test x"${ac_cv_c_odbch}" = x ; then
-	if test "${ac_cv_header_sql_h+set}" = set; then
-  echo "$as_me:$LINENO: checking for sql.h" >&5
-echo $ECHO_N "checking for sql.h... $ECHO_C" >&6
-if test "${ac_cv_header_sql_h+set}" = set; then
+		ccprog=`echo ${CC} | sed -e 's: .*::'`
+	ccpath=`which ${ccprog} | sed -e 's:/bin/.*::'`/include/iodbc
+	if test -f $ccpath/isql.h; then
+	    ac_cv_c_odbch=$ccpath
+	    ODBCTYPE=iodbc
+	fi
+    fi
+
+        if test x"${ac_cv_c_odbch}" = x ; then
+	if test "${ac_cv_header_isql_h+set}" = set; then
+  echo "$as_me:$LINENO: checking for isql.h" >&5
+echo $ECHO_N "checking for isql.h... $ECHO_C" >&6
+if test "${ac_cv_header_isql_h+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 fi
-echo "$as_me:$LINENO: result: $ac_cv_header_sql_h" >&5
-echo "${ECHO_T}$ac_cv_header_sql_h" >&6
+echo "$as_me:$LINENO: result: $ac_cv_header_isql_h" >&5
+echo "${ECHO_T}$ac_cv_header_isql_h" >&6
 else
   # Is the header compilable?
-echo "$as_me:$LINENO: checking sql.h usability" >&5
-echo $ECHO_N "checking sql.h usability... $ECHO_C" >&6
+echo "$as_me:$LINENO: checking isql.h usability" >&5
+echo $ECHO_N "checking isql.h usability... $ECHO_C" >&6
 cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -6695,7 +6706,7 @@ cat confdefs.h >>conftest.$ac_ext
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
 $ac_includes_default
-#include <sql.h>
+#include <isql.h>
 _ACEOF
 rm -f conftest.$ac_objext
 if { (eval echo "$as_me:$LINENO: \"$ac_compile\"") >&5
@@ -6721,8 +6732,8 @@ echo "$as_me:$LINENO: result: $ac_header
 echo "${ECHO_T}$ac_header_compiler" >&6
 
 # Is the header present?
-echo "$as_me:$LINENO: checking sql.h presence" >&5
-echo $ECHO_N "checking sql.h presence... $ECHO_C" >&6
+echo "$as_me:$LINENO: checking isql.h presence" >&5
+echo $ECHO_N "checking isql.h presence... $ECHO_C" >&6
 cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -6730,7 +6741,7 @@ _ACEOF
 cat confdefs.h >>conftest.$ac_ext
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
-#include <sql.h>
+#include <isql.h>
 _ACEOF
 if { (eval echo "$as_me:$LINENO: \"$ac_cpp conftest.$ac_ext\"") >&5
   (eval $ac_cpp conftest.$ac_ext) 2>conftest.er1
@@ -6763,10 +6774,10 @@ echo "${ECHO_T}$ac_header_preproc" >&6
 # So?  What about this header?
 case $ac_header_compiler:$ac_header_preproc in
   yes:no )
-    { echo "$as_me:$LINENO: WARNING: sql.h: accepted by the compiler, rejected by the preprocessor!" >&5
-echo "$as_me: WARNING: sql.h: accepted by the compiler, rejected by the preprocessor!" >&2;}
-    { echo "$as_me:$LINENO: WARNING: sql.h: proceeding with the preprocessor's result" >&5
-echo "$as_me: WARNING: sql.h: proceeding with the preprocessor's result" >&2;}
+    { echo "$as_me:$LINENO: WARNING: isql.h: accepted by the compiler, rejected by the preprocessor!" >&5
+echo "$as_me: WARNING: isql.h: accepted by the compiler, rejected by the preprocessor!" >&2;}
+    { echo "$as_me:$LINENO: WARNING: isql.h: proceeding with the preprocessor's result" >&5
+echo "$as_me: WARNING: isql.h: proceeding with the preprocessor's result" >&2;}
     (
       cat <<\_ASBOX
 ## ------------------------------------ ##
@@ -6777,12 +6788,12 @@ _ASBOX
       sed "s/^/$as_me: WARNING:     /" >&2
     ;;
   no:yes )
-    { echo "$as_me:$LINENO: WARNING: sql.h: present but cannot be compiled" >&5
-echo "$as_me: WARNING: sql.h: present but cannot be compiled" >&2;}
-    { echo "$as_me:$LINENO: WARNING: sql.h: check for missing prerequisite headers?" >&5
-echo "$as_me: WARNING: sql.h: check for missing prerequisite headers?" >&2;}
-    { echo "$as_me:$LINENO: WARNING: sql.h: proceeding with the preprocessor's result" >&5
-echo "$as_me: WARNING: sql.h: proceeding with the preprocessor's result" >&2;}
+    { echo "$as_me:$LINENO: WARNING: isql.h: present but cannot be compiled" >&5
+echo "$as_me: WARNING: isql.h: present but cannot be compiled" >&2;}
+    { echo "$as_me:$LINENO: WARNING: isql.h: check for missing prerequisite headers?" >&5
+echo "$as_me: WARNING: isql.h: check for missing prerequisite headers?" >&2;}
+    { echo "$as_me:$LINENO: WARNING: isql.h: proceeding with the preprocessor's result" >&5
+echo "$as_me: WARNING: isql.h: proceeding with the preprocessor's result" >&2;}
     (
       cat <<\_ASBOX
 ## ------------------------------------ ##
@@ -6793,18 +6804,18 @@ _ASBOX
       sed "s/^/$as_me: WARNING:     /" >&2
     ;;
 esac
-echo "$as_me:$LINENO: checking for sql.h" >&5
-echo $ECHO_N "checking for sql.h... $ECHO_C" >&6
-if test "${ac_cv_header_sql_h+set}" = set; then
+echo "$as_me:$LINENO: checking for isql.h" >&5
+echo $ECHO_N "checking for isql.h... $ECHO_C" >&6
+if test "${ac_cv_header_isql_h+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
-  ac_cv_header_sql_h=$ac_header_preproc
+  ac_cv_header_isql_h=$ac_header_preproc
 fi
-echo "$as_me:$LINENO: result: $ac_cv_header_sql_h" >&5
-echo "${ECHO_T}$ac_cv_header_sql_h" >&6
+echo "$as_me:$LINENO: result: $ac_cv_header_isql_h" >&5
+echo "${ECHO_T}$ac_cv_header_isql_h" >&6
 
 fi
-if test $ac_cv_header_sql_h = yes; then
+if test $ac_cv_header_isql_h = yes; then
   ac_cv_c_odbch=installed
 else
   ac_cv_c_odbch=""
@@ -6812,59 +6823,44 @@ fi
 
 
 	if test x"${ac_cv_c_odbch}" != x ; then
-	    ODBCTYPE=unixODBC
-	fi
-    fi
-
-        if test x"${ODBCTYPE}" = xunixODBC ; then
-	if test -f "${ac_cv_c_odbch}/odbcver.h" ; then
-	    ODBCTYPE=intersolv
-	fi
-    fi
-
-                    for i in $dirlist; do
-	if test -f "${srcdir}/$i/postodbc/isql.h" ; then
-	    ac_cv_c_odbch=`(cd ${srcdir}/$i/postodbc/; pwd)`
 	    ODBCTYPE=iodbc
 	fi
-    done
+    fi
 
-        if test x"${ac_cv_c_odbch}" = x ; then
+                    if test x"${ac_cv_c_odbch}" = x ; then
 	for i in \
+		/opt/odbc/include \
+		/opt/ODBC/include \
+		${prefix}/include/odbc \
+		${prefix}/include/ODBC \
+		/opt/unixodbc/include \
+		/opt/unixODBC/include \
+		${prefix}/include/unixodbc \
+		${prefix}/include/unixODBC \
 		/usr/local/mysql/include \
-		/usr/local/include/iodbc \
-		/usr/local/share/iodbc \
-		${prefix}/include/iodbc \
+		/usr/local/include \
+		/usr/include \
 		${prefix}/include ; do
-	    if test -f $i/isql.h; then
+	    if test -f "$i/sql.h"; then
 		ac_cv_c_odbch=`(cd $i; pwd)`
-		ODBCTYPE=iodbc
+		ODBCTYPE=unixODBC
 	    fi
 	done
     fi
 
         if test x"${ac_cv_c_odbch}" = x ; then
-		ccprog=`echo ${CC} | sed -e 's: .*::'`
-	ccpath=`which ${ccprog} | sed -e 's:/bin/.*::'`/include/iodbc
-	if test -f $ccpath/isql.h; then
-	    ac_cv_c_odbch=$ccpath
-	    ODBCTYPE=iodbc
-	fi
-    fi
-
-        if test x"${ac_cv_c_odbch}" = x ; then
-	if test "${ac_cv_header_isql_h+set}" = set; then
-  echo "$as_me:$LINENO: checking for isql.h" >&5
-echo $ECHO_N "checking for isql.h... $ECHO_C" >&6
-if test "${ac_cv_header_isql_h+set}" = set; then
+	if test "${ac_cv_header_sql_h+set}" = set; then
+  echo "$as_me:$LINENO: checking for sql.h" >&5
+echo $ECHO_N "checking for sql.h... $ECHO_C" >&6
+if test "${ac_cv_header_sql_h+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 fi
-echo "$as_me:$LINENO: result: $ac_cv_header_isql_h" >&5
-echo "${ECHO_T}$ac_cv_header_isql_h" >&6
+echo "$as_me:$LINENO: result: $ac_cv_header_sql_h" >&5
+echo "${ECHO_T}$ac_cv_header_sql_h" >&6
 else
   # Is the header compilable?
-echo "$as_me:$LINENO: checking isql.h usability" >&5
-echo $ECHO_N "checking isql.h usability... $ECHO_C" >&6
+echo "$as_me:$LINENO: checking sql.h usability" >&5
+echo $ECHO_N "checking sql.h usability... $ECHO_C" >&6
 cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -6873,7 +6869,7 @@ cat confdefs.h >>conftest.$ac_ext
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
 $ac_includes_default
-#include <isql.h>
+#include <sql.h>
 _ACEOF
 rm -f conftest.$ac_objext
 if { (eval echo "$as_me:$LINENO: \"$ac_compile\"") >&5
@@ -6899,8 +6895,8 @@ echo "$as_me:$LINENO: result: $ac_header
 echo "${ECHO_T}$ac_header_compiler" >&6
 
 # Is the header present?
-echo "$as_me:$LINENO: checking isql.h presence" >&5
-echo $ECHO_N "checking isql.h presence... $ECHO_C" >&6
+echo "$as_me:$LINENO: checking sql.h presence" >&5
+echo $ECHO_N "checking sql.h presence... $ECHO_C" >&6
 cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -6908,7 +6904,7 @@ _ACEOF
 cat confdefs.h >>conftest.$ac_ext
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
-#include <isql.h>
+#include <sql.h>
 _ACEOF
 if { (eval echo "$as_me:$LINENO: \"$ac_cpp conftest.$ac_ext\"") >&5
   (eval $ac_cpp conftest.$ac_ext) 2>conftest.er1
@@ -6941,10 +6937,10 @@ echo "${ECHO_T}$ac_header_preproc" >&6
 # So?  What about this header?
 case $ac_header_compiler:$ac_header_preproc in
   yes:no )
-    { echo "$as_me:$LINENO: WARNING: isql.h: accepted by the compiler, rejected by the preprocessor!" >&5
-echo "$as_me: WARNING: isql.h: accepted by the compiler, rejected by the preprocessor!" >&2;}
-    { echo "$as_me:$LINENO: WARNING: isql.h: proceeding with the preprocessor's result" >&5
-echo "$as_me: WARNING: isql.h: proceeding with the preprocessor's result" >&2;}
+    { echo "$as_me:$LINENO: WARNING: sql.h: accepted by the compiler, rejected by the preprocessor!" >&5
+echo "$as_me: WARNING: sql.h: accepted by the compiler, rejected by the preprocessor!" >&2;}
+    { echo "$as_me:$LINENO: WARNING: sql.h: proceeding with the preprocessor's result" >&5
+echo "$as_me: WARNING: sql.h: proceeding with the preprocessor's result" >&2;}
     (
       cat <<\_ASBOX
 ## ------------------------------------ ##
@@ -6955,12 +6951,12 @@ _ASBOX
       sed "s/^/$as_me: WARNING:     /" >&2
     ;;
   no:yes )
-    { echo "$as_me:$LINENO: WARNING: isql.h: present but cannot be compiled" >&5
-echo "$as_me: WARNING: isql.h: present but cannot be compiled" >&2;}
-    { echo "$as_me:$LINENO: WARNING: isql.h: check for missing prerequisite headers?" >&5
-echo "$as_me: WARNING: isql.h: check for missing prerequisite headers?" >&2;}
-    { echo "$as_me:$LINENO: WARNING: isql.h: proceeding with the preprocessor's result" >&5
-echo "$as_me: WARNING: isql.h: proceeding with the preprocessor's result" >&2;}
+    { echo "$as_me:$LINENO: WARNING: sql.h: present but cannot be compiled" >&5
+echo "$as_me: WARNING: sql.h: present but cannot be compiled" >&2;}
+    { echo "$as_me:$LINENO: WARNING: sql.h: check for missing prerequisite headers?" >&5
+echo "$as_me: WARNING: sql.h: check for missing prerequisite headers?" >&2;}
+    { echo "$as_me:$LINENO: WARNING: sql.h: proceeding with the preprocessor's result" >&5
+echo "$as_me: WARNING: sql.h: proceeding with the preprocessor's result" >&2;}
     (
       cat <<\_ASBOX
 ## ------------------------------------ ##
@@ -6971,18 +6967,18 @@ _ASBOX
       sed "s/^/$as_me: WARNING:     /" >&2
     ;;
 esac
-echo "$as_me:$LINENO: checking for isql.h" >&5
-echo $ECHO_N "checking for isql.h... $ECHO_C" >&6
-if test "${ac_cv_header_isql_h+set}" = set; then
+echo "$as_me:$LINENO: checking for sql.h" >&5
+echo $ECHO_N "checking for sql.h... $ECHO_C" >&6
+if test "${ac_cv_header_sql_h+set}" = set; then
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
-  ac_cv_header_isql_h=$ac_header_preproc
+  ac_cv_header_sql_h=$ac_header_preproc
 fi
-echo "$as_me:$LINENO: result: $ac_cv_header_isql_h" >&5
-echo "${ECHO_T}$ac_cv_header_isql_h" >&6
+echo "$as_me:$LINENO: result: $ac_cv_header_sql_h" >&5
+echo "${ECHO_T}$ac_cv_header_sql_h" >&6
 
 fi
-if test $ac_cv_header_isql_h = yes; then
+if test $ac_cv_header_sql_h = yes; then
   ac_cv_c_odbch=installed
 else
   ac_cv_c_odbch=""
@@ -6990,7 +6986,13 @@ fi
 
 
 	if test x"${ac_cv_c_odbch}" != x ; then
-	    ODBCTYPE=iodbc
+	    ODBCTYPE=unixODBC
+	fi
+    fi
+
+        if test x"${ODBCTYPE}" = xunixODBC ; then
+	if test -f "${ac_cv_c_odbch}/odbcver.h" ; then
+	    ODBCTYPE=intersolv
 	fi
     fi
 
@@ -7065,7 +7067,7 @@ echo "$as_me: error: ${with_odbclibrary}
         if test x"${ODBCTYPE}" = xiodbc ; then
 	liblist="libpsqlodbc libmysqlodbc librbodbc libiodbc iodbc"
 	libpathlist="/usr/local/lib/iodbc ${prefix}/iodbc/lib \
-	    ${prefix}/lib/iodbc /usr/local/mysql/lib /usr/local/lib ${prefix}/lib"
+	    ${prefix}/lib/iodbc /usr/local/mysql/lib /usr/local/lib /usr/lib ${prefix}/lib"
     elif test x"${ODBCTYPE}" = xunixODBC ; then
 	liblist="libodbc"
 	libpathlist="/opt/unixodbc/lib /opt/unixODBC/lib /usr/include/odbc \
@@ -7207,6 +7209,7 @@ echo "${ECHO_T}Using IODBC driver" >&6
 #define HAVE_IODBC 1
 _ACEOF
 
+	ODBC_LIB="$ODBC_LIB -liodbcinst"
     fi
 
     if test x"${ODBCTYPE}" = xunixODBC ; then
Index: tclodbc.hxx
===================================================================
RCS file: /cvsroot/tclodbc/tclodbc/tclodbc.hxx,v
retrieving revision 1.4
diff -u -p -r1.4 tclodbc.hxx
--- tclodbc.hxx	20 Mar 2004 22:14:19 -0000	1.4
+++ tclodbc.hxx	2 Jun 2004 05:21:19 -0000
@@ -110,7 +110,9 @@ extern "C" {
 #ifdef HAVE_UNIXODBC
 #define INCLUDE_EXTENSIONS // Uncomment this if odbc extensions available
 #endif
-//#define INCLUDE_EXTENSIONS // Uncomment this if odbc extensions available
+#ifdef HAVE_IODBC
+#define INCLUDE_EXTENSIONS // Uncomment this if odbc extensions available
+#endif
 #endif
 
 #include "tclobj.hxx"
@@ -128,8 +132,7 @@ extern "C" {
 /* iODBC 2.12 */
 #include <isql.h>
 #include <isqlext.h>
-#include <odbc_funcs.h>
-#include <odbc_types.h>
+#include <iodbcinst.h>
 #else
 #include <sql.h>
 #include <odbcinst.h>
diff -u -rupN tclodbc.cxx tclodbc.cxx
--- tclodbc.cxx	2004-03-19 12:32:21.000000000 -0800
+++ tclodbc.cxx	2005-06-20 13:22:16.000000000 -0700
@@ -121,13 +121,15 @@ TclObj SqlErr (HENV env, HDBC dbc, HSTMT
     char SqlMessage[SQL_MAX_MESSAGE_LENGTH];
     char SqlState[6];
     SDWORD NativeError;
+    SQLINTEGER tmp;
     SWORD Available;
     RETCODE rc;
     TclObj errObj;
 
     rc = SQLError(env, dbc, stmt, 
-            (UCHAR*) SqlState, &NativeError, (UCHAR*) SqlMessage, 
+            (UCHAR*) SqlState, &tmp, (UCHAR*) SqlMessage, 
             SQL_MAX_MESSAGE_LENGTH-1, &Available);
+    NativeError = tmp;
 
     // sql error object is a triple:
     // {standard error code} {native error code} {error message}
--- statemnt.cxx	2004-03-20 14:14:19.000000000 -0800
+++ statemnt.cxx	2005-06-20 13:56:06.000000000 -0700
@@ -358,9 +358,9 @@ TclObj TclStatement::Result() {
 		result.appendElement(row);
 	    }
 	} else {
-	    long i;
-	    SQLRowCount(stmt, (SQLINTEGER *) &i);
-	    result = i;
+	    SQLLEN i;
+	    SQLRowCount(stmt, &i);
+	    result = (long)i;
 	}
 	return result;
     } else {
@@ -398,9 +398,9 @@ TclObj TclStatement::Result() {
 
 
 	    } else {
-		long i;
-		SQLRowCount(stmt, (SQLINTEGER *) &i);
-		return i;
+		SQLLEN i;
+		SQLRowCount(stmt, &i);
+		return (long)i;
 	    }
 
 	} while(more);
--- /dev/null	2007-02-14 02:20:41.000000000 -0800
+++ tclodbc.exports	2007-02-14 03:23:30.000000000 -0800
@@ -0,0 +1,257 @@
+__Z10MapSqlTypel.eh
+__Z11EncodedTypei.eh
+__ZN12TclCmdObject11AddToMyListEPS_.eh
+__ZN12TclCmdObject14RemoveFromListEv.eh
+__ZN12TclCmdObject7DestroyEPv.eh
+__ZN12TclCmdObject8DispatchEPvP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclCmdObjectD0Ev.eh
+__ZN12TclCmdObjectD1Ev.eh
+__ZN12TclCmdObjectD2Ev.eh
+__ZN6TclObj12EncodedValueEv.eh
+__ZN6TclObj6isNullEv.eh
+__ZN6TclObjC1EP7Tcl_Obj.eh
+__ZN6TclObjC1ERKS_.eh
+__ZN6TclObjC2EP7Tcl_Obj.eh
+__ZN6TclObjC2ERKS_.eh
+__ZN6TclObjD1Ev.eh
+__ZN6TclObjD2Ev.eh
+__ZNK6TclObjcvPcEv.eh
+__ZN6TclObjC2EPKci
+__ZN6TclObjC1EPKci
+__ZN6TclObjcvP7Tcl_ObjEv
+__ZN6TclObjC2El
+__ZN6TclObjC1El
+__ZN6TclObjC2ERKS_
+__ZN6TclObjC1ERKS_
+__ZN6TclObjC2EP7Tcl_Obj
+__ZN6TclObjC1EP7Tcl_Obj
+__ZN6TclObjD2Ev
+__ZN6TclObjD1Ev
+__ZN6TclObjaSERKS_
+__ZN6TclObj3setEPKci
+__ZNK6TclObjcvPcEv
+__ZN6TclObj6lenghtEv
+__ZN6TclObj6isNullEv
+__ZN6TclObj12EncodedValueEv
+__ZN6TclObj6DecodeEP13Tcl_Encoding_
+__ZN6TclObj13EncodedLenghtEv
+__ZN6TclObjC2EPKcP13Tcl_Encoding_i
+__ZN6TclObjC1EPKcP13Tcl_Encoding_i
+__ZN6TclObj9setLengthEi
+__ZN6TclObj13appendElementES_P10Tcl_Interp
+__ZN6TclObj6appendEPKci
+__ZN6TclObj6lindexEiP10Tcl_Interp
+__ZN6TclObj7llenghtEP10Tcl_Interp
+__ZN6TclObj5asIntEP10Tcl_Interp
+__ZN6TclObj6EncodeEP13Tcl_Encoding_P10Tcl_Interp
+__ZN6TclObj4evalEP10Tcl_Interp
+__Z8StrToNumPcP6NumStrS_h
+__Z8NumToStrsP6NumStr
+__Z11EncodedTypei
+__Z10MapSqlTypel
+__Z12Tclodbc_KillPv
+__Z6SqlErrPvS_S_
+_Tclodbc_Init
+__Z12tcl_databasePvP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclCmdObject7DestroyEPv
+__ZN12TclCmdObject14RemoveFromListEv
+__ZN12TclCmdObject8DispatchEPvP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclCmdObject11AddToMyListEPS_
+__ZN11TclDatabase10AutocommitEi
+__ZN11TclDatabase9SetOptionEP10Tcl_InterpPcS2_
+__ZN11TclDatabase8TransactEt
+__ZN11TclDatabase9GetOptionEPc
+__ZN11TclDatabase11DatasourcesEv
+__ZN11TclDatabase7DriversEv
+__ZN11TclDatabase9ConfigureEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclCmdObjectD2Ev
+__ZN11TclDatabaseD0Ev
+__ZN11TclDatabaseD1Ev
+__ZN11TclDatabaseD2Ev
+__ZN11TclDatabaseC1E6TclObj
+__ZN11TclDatabaseC2E6TclObj
+__ZN11TclDatabaseC1E6TclObjS0_S0_
+__ZN11TclDatabaseC2E6TclObjS0_S0_
+__ZN12TclCmdObjectD1Ev
+__ZN12TclCmdObjectD0Ev
+__ZN11TclDatabase8DispatchEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatement8FreeStmtEv
+__ZN12TclStatement7SqlWaitEi
+__ZN12TclStatement11ColumnCountEv
+__ZN12TclStatement16FreeResultBufferEv
+__ZN12TclStatement19ReserveResultBufferEv
+__ZN12TclStatement9SetOptionEPcS0_
+__ZN12TclStatement6Fetch1Ev
+__ZN19TclPrimaryKeysQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatement9GetOptionEPc
+__ZN15TclSqlStatement10SetArgDefsEP10Tcl_Interp6TclObj
+__ZN16TclTypeInfoQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatement10ColumnInfoEit
+__ZN12TclStatement7ColumnsEiPKP7Tcl_Obj
+__ZN12TclStatement12ColumnLabelsEv
+__ZN12TclStatement5FetchER6TclObj
+__ZN12TclStatement4ReadEP10Tcl_Interp6TclObjiPKP7Tcl_Obj
+__ZN12TclStatement4EvalEP10Tcl_Interp6TclObjiPKP7Tcl_Obj
+__ZN15TclSqlStatement7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN13TclTableQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN14TclColumnQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN13TclIndexQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatementC1ER11TclDatabase
+__ZN12TclStatement6ResultEv
+__ZN12TclStatement5ValueEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatement8DispatchEP10Tcl_InterpiPKP7Tcl_Obj
+__ZN12TclStatementC2ER11TclDatabase
+__ZN12TclStatementD2Ev
+__ZN15TclSqlStatementC1ER11TclDatabase6TclObjb
+__ZN15TclSqlStatementC2ER11TclDatabase6TclObjb
+__ZN12TclStatementD1Ev
+__ZN12TclStatementD0Ev
+__Z24FreeTclodbcEncodedStringP7Tcl_Obj
+__Z29DuplicateTclodbcEncodedStringP7Tcl_ObjS0_
+__Z26UpdateTclodbcEncodedStringP7Tcl_Obj
+__Z23SetTclodbcEncodedStringP10Tcl_InterpP7Tcl_Obj
+_Tcl_InitStubs
+__ZN6TclObjC2EPKci.eh
+__ZN6TclObjC1EPKci.eh
+__ZN6TclObjcvP7Tcl_ObjEv.eh
+__ZN6TclObjC2El.eh
+__ZN6TclObjC1El.eh
+__ZN6TclObjaSERKS_.eh
+__ZN6TclObj3setEPKci.eh
+__ZN6TclObj6lenghtEv.eh
+__ZN6TclObj6DecodeEP13Tcl_Encoding_.eh
+__ZN6TclObj13EncodedLenghtEv.eh
+__ZN6TclObjC2EPKcP13Tcl_Encoding_i.eh
+__ZN6TclObjC1EPKcP13Tcl_Encoding_i.eh
+__ZN6TclObj9setLengthEi.eh
+__ZN6TclObj13appendElementES_P10Tcl_Interp.eh
+__ZN6TclObj6appendEPKci.eh
+__ZN6TclObj6lindexEiP10Tcl_Interp.eh
+__ZN6TclObj7llenghtEP10Tcl_Interp.eh
+__ZN6TclObj5asIntEP10Tcl_Interp.eh
+__ZN6TclObj6EncodeEP13Tcl_Encoding_P10Tcl_Interp.eh
+__ZN6TclObj4evalEP10Tcl_Interp.eh
+__Z8StrToNumPcP6NumStrS_h.eh
+__Z8NumToStrsP6NumStr.eh
+__Z12Tclodbc_KillPv.eh
+__Z6SqlErrPvS_S_.eh
+_Tclodbc_Init.eh
+__Z12tcl_databasePvP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN11TclDatabase10AutocommitEi.eh
+__ZN11TclDatabase9SetOptionEP10Tcl_InterpPcS2_.eh
+__ZN11TclDatabase8TransactEt.eh
+__ZN11TclDatabase9GetOptionEPc.eh
+__ZN11TclDatabase11DatasourcesEv.eh
+__ZN11TclDatabase7DriversEv.eh
+__ZN11TclDatabase9ConfigureEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN11TclDatabaseD0Ev.eh
+__ZN11TclDatabaseD1Ev.eh
+__ZN11TclDatabaseD2Ev.eh
+__ZN11TclDatabaseC1E6TclObj.eh
+__ZN11TclDatabaseC2E6TclObj.eh
+__ZN11TclDatabaseC1E6TclObjS0_S0_.eh
+__ZN11TclDatabaseC2E6TclObjS0_S0_.eh
+__ZN11TclDatabase8DispatchEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatement8FreeStmtEv.eh
+__ZN12TclStatement7SqlWaitEi.eh
+__ZN12TclStatement11ColumnCountEv.eh
+__ZN12TclStatement16FreeResultBufferEv.eh
+__ZN12TclStatement19ReserveResultBufferEv.eh
+__ZN12TclStatement9SetOptionEPcS0_.eh
+__ZN12TclStatement6Fetch1Ev.eh
+__ZN19TclPrimaryKeysQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatement9GetOptionEPc.eh
+__ZN15TclSqlStatement10SetArgDefsEP10Tcl_Interp6TclObj.eh
+__ZN16TclTypeInfoQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatement10ColumnInfoEit.eh
+__ZN12TclStatement7ColumnsEiPKP7Tcl_Obj.eh
+__ZN12TclStatement12ColumnLabelsEv.eh
+__ZN12TclStatement5FetchER6TclObj.eh
+__ZN12TclStatement4ReadEP10Tcl_Interp6TclObjiPKP7Tcl_Obj.eh
+__ZN12TclStatement4EvalEP10Tcl_Interp6TclObjiPKP7Tcl_Obj.eh
+__ZN15TclSqlStatement7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN13TclTableQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN14TclColumnQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN13TclIndexQuery7ExecuteEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatementC1ER11TclDatabase.eh
+__ZN12TclStatement6ResultEv.eh
+__ZN12TclStatement5ValueEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatement8DispatchEP10Tcl_InterpiPKP7Tcl_Obj.eh
+__ZN12TclStatementC2ER11TclDatabase.eh
+__ZN12TclStatementD2Ev.eh
+__ZN15TclSqlStatementC1ER11TclDatabase6TclObjb.eh
+__ZN15TclSqlStatementC2ER11TclDatabase6TclObjb.eh
+__ZN12TclStatementD1Ev.eh
+__ZN12TclStatementD0Ev.eh
+__Z24FreeTclodbcEncodedStringP7Tcl_Obj.eh
+__Z29DuplicateTclodbcEncodedStringP7Tcl_ObjS0_.eh
+__Z26UpdateTclodbcEncodedStringP7Tcl_Obj.eh
+__Z23SetTclodbcEncodedStringP10Tcl_InterpP7Tcl_Obj.eh
+_strFunctionSequenceError
+_envRefCounter
+_env
+_statementCmds
+_databaseCmds
+_databaseOptions
+_cursorOp
+_concurrencyOp
+_booleanOp
+_connectOp
+_configOp
+_stmtOp
+_attrDef
+_sqlType
+_strOldSyntax
+_strInvalidOption
+_strAsyncEnable
+_strRowCount
+_strStatement
+_strEncoding
+_strNoscan
+_strCursortype
+_strRowsetsize
+_strMaxlength
+_strMaxlenght
+_strTimeout
+_strMaxrows
+_strConcurrency
+_strFetch
+_strMoreResults
+_strExecute
+_strDrop
+_strRun
+_strDisconnect
+_strEmpty
+_strCmdNotAvailable
+_strUsage
+_strWrongArgs
+_strRollback
+_strCommit
+_strAutocommit
+_strSet
+_strGet
+_strRead
+_strEval
+_strTypeinfo
+_strPrimarykeys
+_strIndexes
+_strColumns
+_strTables
+_strOK
+_strInvalidHandle
+_strMemoryAllocationFailed
+_strVersion
+_TclodbcEncodedString
+_tclIntPlatStubsPtr
+_tclIntStubsPtr
+_tclPlatStubsPtr
+_tclStubsPtr
+__ZTV12TclCmdObject
+__ZTV11TclDatabase
+__ZTV12TclStatement
+__ZTV15TclSqlStatement
+__ZTV13TclTableQuery
+__ZTV14TclColumnQuery
+__ZTV13TclIndexQuery
+__ZTV16TclTypeInfoQuery
+__ZTV19TclPrimaryKeysQuery
--- Makefile.in~	2004-03-19 12:32:21.000000000 -0800
+++ Makefile.in	2007-02-14 03:27:46.000000000 -0800
@@ -114,7 +114,7 @@
 RANLIB_STUB	= @RANLIB_STUB@
 SHLIB_CFLAGS	= @SHLIB_CFLAGS@
-SHLIB_LD	= @SHLIB_LD@
+SHLIB_LD	= g++
-SHLIB_LD_FLAGS	= @SHLIB_LD_FLAGS@
+SHLIB_LD_FLAGS	= @SHLIB_LD_FLAGS@ -exported_symbols_list ${srcdir}/tclodbc.exports
 SHLIB_LD_LIBS	= @SHLIB_LD_LIBS@
 STLIB_LD	= @STLIB_LD@
 TCL_DEFS	= @TCL_DEFS@
