diff -u --exclude=CVS -r rsync.h rsync.h
--- rsync.h	2004-09-21 21:10:10.000000000 -0700
+++ rsync.h	2005-01-12 11:53:00.000000000 -0800
@@ -772,3 +772,9 @@
 #define UNUSED(x) x __attribute__((__unused__))
 
 extern const char *io_write_phase, *io_read_phase;
+ /* hack for building fat on Mac OS X */
+
+#if defined(__APPLE__) && defined(WORDS_BIGENDIAN)
+#undef WORDS_BIGENDIAN
+#define WORDS_BIGENDIAN __BIG_ENDIAN__
+#endif
