--- lib/xmlparse.c.orig	2008-08-05 20:19:23.000000000 -0700
+++ lib/xmlparse.c	2008-08-05 20:20:04.000000000 -0700
@@ -2552,14 +2552,13 @@
       return XML_ERROR_NONE;
     case XML_TOK_DATA_CHARS: 
       {
-        XML_CharacterDataHandler charDataHandler = characterDataHandler;
-        if (charDataHandler) {
+        if (characterDataHandler) {
           if (MUST_CONVERT(enc, s)) {
             for (;;) {
               ICHAR *dataPtr = (ICHAR *)dataBuf;
               XmlConvert(enc, &s, next, &dataPtr, (ICHAR *)dataBufEnd);
               *eventEndPP = s;
-              charDataHandler(handlerArg, dataBuf,
+              characterDataHandler(handlerArg, dataBuf,
                               (int)(dataPtr - (ICHAR *)dataBuf));
               if (s == next)
                 break;
@@ -2567,7 +2566,7 @@
             }
           }
           else
-            charDataHandler(handlerArg,
+            characterDataHandler(handlerArg,
                             (XML_Char *)s,
                             (int)((XML_Char *)next - (XML_Char *)s));
         }
@@ -3139,14 +3138,13 @@
       break;
     case XML_TOK_DATA_CHARS:
       {
-        XML_CharacterDataHandler charDataHandler = characterDataHandler;
-        if (charDataHandler) {
+        if (characterDataHandler) {
           if (MUST_CONVERT(enc, s)) {
             for (;;) {
               ICHAR *dataPtr = (ICHAR *)dataBuf;
               XmlConvert(enc, &s, next, &dataPtr, (ICHAR *)dataBufEnd);
               *eventEndPP = next;
-              charDataHandler(handlerArg, dataBuf,
+              characterDataHandler(handlerArg, dataBuf,
                               (int)(dataPtr - (ICHAR *)dataBuf));
               if (s == next)
                 break;
@@ -3154,7 +3152,7 @@
             }
           }
           else
-            charDataHandler(handlerArg,
+            characterDataHandler(handlerArg,
                             (XML_Char *)s,
                             (int)((XML_Char *)next - (XML_Char *)s));
         }
