--- configure.orig	2006-01-24 12:56:40.000000000 -0800
+++ configure	2006-01-24 13:07:53.000000000 -0800
@@ -6111,7 +6111,7 @@
     old_postinstall_cmds="$old_postinstall_cmds~\$RANLIB -t \$oldlib"
     ;;
   *)
-    old_postinstall_cmds="$old_postinstall_cmds~\$RANLIB \$oldlib"
+    old_postinstall_cmds="$old_postinstall_cmds:\$RANLIB \$oldlib"
     ;;
   esac
   old_archive_cmds="$old_archive_cmds~\$RANLIB \$oldlib"
--- conftools/ltmain.sh.orig	2006-01-24 12:54:58.000000000 -0800
+++ conftools/ltmain.sh	2006-01-24 13:05:19.000000000 -0800
@@ -6196,7 +6196,7 @@
 
       # Do each command in the postinstall commands.
       cmds=$old_postinstall_cmds
-      save_ifs="$IFS"; IFS='~'
+      save_ifs="$IFS"; IFS=':'
       for cmd in $cmds; do
 	IFS="$save_ifs"
 	eval cmd=\"$cmd\"
