--- php-5.3.1/acconfig.h	2009-11-18 12:11:57.000000000 -0800
+++ php/acconfig.h	2009-11-19 18:09:31.000000000 -0800
@@ -20,6 +20,8 @@
 
 /* $Id: acconfig.h 272370 2008-12-31 11:15:49Z sebastian $ */
 
+#ifndef ZEND_API
+
 #if defined(__GNUC__) && __GNUC__ >= 4
 # define ZEND_API __attribute__ ((visibility("default")))
 # define ZEND_DLEXPORT __attribute__ ((visibility("default")))
@@ -120,6 +122,26 @@
 #endif
 
 /*
+ * Darwin's GCC can generate multiple architectures in a single pass so the size and
+ * byte order will only be accurate for the one architecture that happened to invoke
+ * configure.
+ */
+#ifdef __DARWIN_BYTE_ORDER
+#/* This symbol must be allowed to be cleared. */ undef WORDS_BIGENDIAN
+# if (__DARWIN_BYTE_ORDER == __DARWIN_BIG_ENDIAN)
+#  define WORDS_BIGENDIAN 1
+# endif
+#/* This symbol must be allowed to be cleared. */ undef SIZEOF_LONG
+# ifdef __LP64__
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 8
+# else
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 4
+# endif
+#endif
+
+#endif	/* ZEND_API */
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
--- php-5.3.1/main/php_config.h.in	2009-11-18 12:12:01.000000000 -0800
+++ php/main/php_config.h.in	2009-11-19 18:09:31.000000000 -0800
@@ -21,6 +21,8 @@
 
 /* $Id: acconfig.h 272370 2008-12-31 11:15:49Z sebastian $ */
 
+#ifndef ZEND_API
+
 #if defined(__GNUC__) && __GNUC__ >= 4
 # define ZEND_API __attribute__ ((visibility("default")))
 # define ZEND_DLEXPORT __attribute__ ((visibility("default")))
@@ -2918,6 +2920,26 @@
 #endif
 
 /*
+ * Darwin's GCC can generate multiple architectures in a single pass so the size and
+ * byte order will only be accurate for the one architecture that happened to invoke
+ * configure.
+ */
+#ifdef __DARWIN_BYTE_ORDER
+#/* This symbol must be allowed to be cleared. */ undef WORDS_BIGENDIAN
+# if (__DARWIN_BYTE_ORDER == __DARWIN_BIG_ENDIAN)
+#  define WORDS_BIGENDIAN 1
+# endif
+#/* This symbol must be allowed to be cleared. */ undef SIZEOF_LONG
+# ifdef __LP64__
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 8
+# else
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 4
+# endif
+#endif
+
+#endif	/* ZEND_API */
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
