diff -aur php-5.3.2/acconfig.h php/acconfig.h
--- php-5.3.2/acconfig.h	2010-03-03 08:52:37.000000000 -0800
+++ php/acconfig.h	2010-03-05 16:38:04.000000000 -0800
@@ -20,6 +20,8 @@
 
 /* $Id: acconfig.h 293155 2010-01-05 20:46:53Z sebastian $ */
 
+#ifndef ZEND_API
+
 #if defined(__GNUC__) && __GNUC__ >= 4
 # define ZEND_API __attribute__ ((visibility("default")))
 # define ZEND_DLEXPORT __attribute__ ((visibility("default")))
@@ -120,6 +122,26 @@
 #endif
 
 /*
+ * Darwin's GCC can generate multiple architectures in a single pass so the size and
+ * byte order will only be accurate for the one architecture that happened to invoke
+ * configure.
+ */
+#ifdef __DARWIN_BYTE_ORDER
+#/* This symbol must be allowed to be cleared. */ undef WORDS_BIGENDIAN
+# if (__DARWIN_BYTE_ORDER == __DARWIN_BIG_ENDIAN)
+#  define WORDS_BIGENDIAN 1
+# endif
+#/* This symbol must be allowed to be cleared. */ undef SIZEOF_LONG
+# ifdef __LP64__
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 8
+# else
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 4
+# endif
+#endif
+
+#endif	/* ZEND_API */
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
diff -aur php-5.3.2/main/php_config.h.in php/main/php_config.h.in
--- php-5.3.2/main/php_config.h.in	2010-03-03 08:52:40.000000000 -0800
+++ php/main/php_config.h.in	2010-03-05 16:38:04.000000000 -0800
@@ -21,6 +21,8 @@
 
 /* $Id: acconfig.h 293155 2010-01-05 20:46:53Z sebastian $ */
 
+#ifndef ZEND_API
+
 #if defined(__GNUC__) && __GNUC__ >= 4
 # define ZEND_API __attribute__ ((visibility("default")))
 # define ZEND_DLEXPORT __attribute__ ((visibility("default")))
@@ -2993,6 +2995,26 @@
 #endif
 
 /*
+ * Darwin's GCC can generate multiple architectures in a single pass so the size and
+ * byte order will only be accurate for the one architecture that happened to invoke
+ * configure.
+ */
+#ifdef __DARWIN_BYTE_ORDER
+#/* This symbol must be allowed to be cleared. */ undef WORDS_BIGENDIAN
+# if (__DARWIN_BYTE_ORDER == __DARWIN_BIG_ENDIAN)
+#  define WORDS_BIGENDIAN 1
+# endif
+#/* This symbol must be allowed to be cleared. */ undef SIZEOF_LONG
+# ifdef __LP64__
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 8
+# else
+#/* This symbol should not be modified by configure. */  define SIZEOF_LONG 4
+# endif
+#endif
+
+#endif	/* ZEND_API */
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
