--- php-5.3.0/acconfig.h	2009-06-10 11:23:33.000000000 -0700
+++ php/acconfig.h	2009-06-16 17:10:25.000000000 -0700
@@ -120,6 +120,15 @@
 #endif
 
 /*
+ * Don't use <mach-o/dyld.h> on Mac OS X / Darwin if dl*() functions are available.
+ * Because this header is processed multiple times during compilation (thanks to lack of
+ * #ifdef protectors), this logic had to move here.
+ */
+#if defined(HAVE_DLFCN_H) && defined(HAVE_MACH_O_DYLD_H)
+#/* This symbol must be allowed to be cleared. */ undef HAVE_MACH_O_DYLD_H
+#endif
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
--- php-5.3.0RC3/main/php_config.h.in	2009-06-10 11:23:36.000000000 -0700
+++ php/main/php_config.h.in	2009-06-16 17:10:25.000000000 -0700
@@ -2948,6 +2948,15 @@
 #endif
 
 /*
+ * Don't use <mach-o/dyld.h> on Mac OS X / Darwin if dl*() functions are available.
+ * Because this header is processed multiple times during compilation (thanks to lack of
+ * #ifdef protectors), this logic had to move here.
+ */
+#if defined(HAVE_DLFCN_H) && defined(HAVE_MACH_O_DYLD_H)
+#/* This symbol must be allowed to be cleared. */ undef HAVE_MACH_O_DYLD_H
+#endif
+
+/*
  * Local variables:
  * tab-width: 4
  * c-basic-offset: 4
