diff --git a/ext/ldap/config.m4 b/ext/ldap/config.m4
index 69651a89cd..e25b60c79c 100644
--- a/ext/ldap/config.m4
+++ b/ext/ldap/config.m4
@@ -206,7 +206,7 @@ if test "$PHP_LDAP" != "no"; then
 
   dnl Solaris 2.8 claims to be 2004 API, but doesn't have
   dnl ldap_parse_reference() nor ldap_start_tls_s()
-  AC_CHECK_FUNCS([ldap_parse_result ldap_parse_reference ldap_start_tls_s ldap_control_find ldap_parse_extended_result ldap_extended_operation ldap_extended_operation_s ldap_passwd ldap_whoami_s ldap_refresh_s])
+  AC_CHECK_FUNCS([ldap_parse_result ldap_parse_reference ldap_start_tls_s ldap_control_find ldap_parse_extended_result ldap_extended_operation ldap_extended_operation_s ldap_passwd ldap_whoami_s ldap_refresh_s ldap_create_passwordpolicy_control ldap_parse_passwordpolicy_control ldap_create_sort_control_value ldap_create_vlv_control_value])
 
   dnl
   dnl SASL check
--- php/ext/ldap/ldap.c.orig	2020-09-23 07:05:46.000000000 -0700
+++ php/ext/ldap/ldap.c	2020-09-23 07:03:23.000000000 -0700
@@ -151,8 +151,11 @@
 		int expire = 0, grace = 0, rc;
 		LDAPPasswordPolicyError pperr;
 		zval value;
-
+#ifdef HAVE_LDAP_PARSE_PASSWORDPOLICY_CONTROL
 		rc = ldap_parse_passwordpolicy_control(ld, ctrl, &expire, &grace, &pperr);
+#else
+        rc = LDAP_NOT_SUPPORTED;
+#endif
 		if ( rc == LDAP_SUCCESS ) {
 			array_init(&value);
 			add_assoc_long(&value, "expire", expire);
@@ -496,7 +499,11 @@
 			sort_keys[num_keys] = NULL;
 			/* ldap_create_sort_control_value() allocates memory for control_value.bv_val */
 			control_value_alloc = 1;
+#ifdef HAVE_LDAP_CREATE_SORT_CONTROL_VALUE
 			rc = ldap_create_sort_control_value(ld, sort_keys, &control_value);
+#else
+            rc = LDAP_NOT_SUPPORTED;
+#endif
 			if (rc != LDAP_SUCCESS) {
 				php_error_docref(NULL, E_WARNING, "Failed to create sort control value: %s (%d)", ldap_err2string(rc), rc);
 			}
@@ -562,7 +569,11 @@
 
 			/* ldap_create_vlv_control_value() allocates memory for control_value.bv_val */
 			control_value_alloc = 1;
+#ifdef HAVE_LDAP_CREATE_VLV_CONTROL_VALUE
 			rc = ldap_create_vlv_control_value(ld, &vlvInfo, &control_value);
+#else
+            rc = LDAP_NOT_SUPPORTED;
+#endif
 			if (rc != LDAP_SUCCESS) {
 				php_error_docref(NULL, E_WARNING, "Failed to create VLV control value: %s (%d)", ldap_err2string(rc), rc);
 			}
@@ -4320,10 +4331,14 @@
 
 	switch (myargcount) {
 		case 5:
+#ifdef HAVE_LDAP_CREATE_PASSWORDPOLICY_CONTROL
 			/* ldap_create_passwordpolicy_control() allocates ctrl */
 			if (ldap_create_passwordpolicy_control(ld->link, &ctrl) == LDAP_SUCCESS) {
 				requestctrls[0] = ctrl;
 			}
+#else
+			requestctrls[0] = NULL;
+#endif
 	}
 
 	/* asynchronous call to get result and controls */
