diff -ur a/acinclude.m4 b/acinclude.m4
--- a/acinclude.m4	2016-06-21 18:50:40.000000000 -0700
+++ b/acinclude.m4	2016-07-08 14:30:30.000000000 -0700
@@ -2401,7 +2401,7 @@
     ])
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
     PHP_CHECK_LIBRARY(ssl, SSL_CTX_set_ssl_version, [
       found_openssl=yes
     ],[
diff -ur a/aclocal.m4 b/aclocal.m4
--- a/aclocal.m4	2016-06-21 19:11:59.000000000 -0700
+++ b/aclocal.m4	2016-07-08 14:30:11.000000000 -0700
@@ -2401,7 +2401,7 @@
     ])
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
     PHP_CHECK_LIBRARY(ssl, SSL_CTX_set_ssl_version, [
       found_openssl=yes
     ],[
diff -ur a/configure b/configure
--- a/configure	2016-06-21 19:12:02.000000000 -0700
+++ b/configure	2016-07-08 14:29:10.000000000 -0700
@@ -24410,13 +24410,13 @@
 
   fi
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for DSA_get_default_method in -lssl" >&5
-$as_echo_n "checking for DSA_get_default_method in -lssl... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for DSA_get_default_method in -lssl.35" >&5
+$as_echo_n "checking for DSA_get_default_method in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_DSA_get_default_method+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -24452,13 +24452,13 @@
 
 fi
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for X509_free in -lcrypto" >&5
-$as_echo_n "checking for X509_free in -lcrypto... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for X509_free in -lcrypto.35" >&5
+$as_echo_n "checking for X509_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_X509_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -24712,7 +24712,8 @@
     if test "$found_openssl" = "no"; then
 
     if test "$PHP_OPENSSL_DIR" = "yes"; then
-      PHP_OPENSSL_DIR="/usr/local/ssl /usr/local /usr /usr/local/openssl"
+      OPENSSL_LIBDIR="/usr/lib"
+      PHP_OPENSSL_DIR="/usr/local/libressl /usr/local/ssl /usr/local /usr /usr/local/openssl"
     fi
 
     for i in $PHP_OPENSSL_DIR; do
@@ -24726,7 +24727,7 @@
     done
 
     if test -z "$OPENSSL_INCDIR"; then
-      as_fn_error $? "Cannot find OpenSSL's <evp.h>" "$LINENO" 5
+      as_fn_error $? "Cannot find OpenSSL's <evp.h> in: $OPENSSL_INCDIR" "$LINENO" 5
     fi
 
     if test -z "$OPENSSL_LIBDIR"; then
@@ -24890,13 +24891,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto" >&5
-$as_echo_n "checking for CRYPTO_free in -lcrypto... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto.35" >&5
+$as_echo_n "checking for CRYPTO_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_CRYPTO_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -24937,14 +24938,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        OPENSSL_SHARED_LIBADD="-lcrypto $OPENSSL_SHARED_LIBADD"
+        OPENSSL_SHARED_LIBADD="-lcrypto.35 $OPENSSL_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -24969,7 +24970,7 @@
 
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
 
   save_old_LDFLAGS=$LDFLAGS
   ac_stuff="
@@ -25067,13 +25068,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl" >&5
-$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl.35" >&5
+$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_SSL_CTX_set_ssl_version+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -25129,14 +25130,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        OPENSSL_SHARED_LIBADD="-lssl $OPENSSL_SHARED_LIBADD"
+        OPENSSL_SHARED_LIBADD="-lssl.35 $OPENSSL_SHARED_LIBADD"
     else
 
 
   case ssl in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lssl $LIBS"
+      LIBS="-lssl.35 $LIBS"
    ;;
   esac
 
@@ -25152,14 +25153,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        OPENSSL_SHARED_LIBADD="-lcrypto $OPENSSL_SHARED_LIBADD"
+        OPENSSL_SHARED_LIBADD="-lcrypto.35 $OPENSSL_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -37985,7 +37986,8 @@
     if test "$found_openssl" = "no"; then
 
     if test "$PHP_OPENSSL_DIR" = "yes"; then
-      PHP_OPENSSL_DIR="/usr/local/ssl /usr/local /usr /usr/local/openssl"
+      OPENSSL_LIBDIR="/usr/lib"
+      PHP_OPENSSL_DIR="/usr/local/libressl /usr/local/ssl /usr/local /usr /usr/local/openssl"
     fi
 
     for i in $PHP_OPENSSL_DIR; do
@@ -38163,13 +38165,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto" >&5
-$as_echo_n "checking for CRYPTO_free in -lcrypto... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto.35" >&5
+$as_echo_n "checking for CRYPTO_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_CRYPTO_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -38210,14 +38212,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        FTP_SHARED_LIBADD="-lcrypto $FTP_SHARED_LIBADD"
+        FTP_SHARED_LIBADD="-lcrypto.35 $FTP_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -38242,7 +38244,7 @@
 
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
 
   save_old_LDFLAGS=$LDFLAGS
   ac_stuff="
@@ -38340,13 +38342,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl" >&5
-$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl.35" >&5
+$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_SSL_CTX_set_ssl_version+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -38402,14 +38404,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        FTP_SHARED_LIBADD="-lssl $FTP_SHARED_LIBADD"
+        FTP_SHARED_LIBADD="-lssl.35 $FTP_SHARED_LIBADD"
     else
 
 
   case ssl in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lssl $LIBS"
+      LIBS="-lssl.35 $LIBS"
    ;;
   esac
 
@@ -38425,14 +38427,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        FTP_SHARED_LIBADD="-lcrypto $FTP_SHARED_LIBADD"
+        FTP_SHARED_LIBADD="-lcrypto.35 $FTP_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -49597,7 +49599,8 @@
     if test "$found_openssl" = "no"; then
 
     if test "$PHP_OPENSSL_DIR" = "yes"; then
-      PHP_OPENSSL_DIR="/usr/local/ssl /usr/local /usr /usr/local/openssl"
+      OPENSSL_LIBDIR="/usr/lib"
+      PHP_OPENSSL_DIR="/usr/local/libressl /usr/local/ssl /usr/local /usr /usr/local/openssl"
     fi
 
     for i in $PHP_OPENSSL_DIR; do
@@ -49775,13 +49778,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto" >&5
-$as_echo_n "checking for CRYPTO_free in -lcrypto... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto.35" >&5
+$as_echo_n "checking for CRYPTO_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_CRYPTO_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -49822,14 +49825,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        IMAP_SHARED_LIBADD="-lcrypto $IMAP_SHARED_LIBADD"
+        IMAP_SHARED_LIBADD="-lcrypto.35 $IMAP_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -49854,7 +49857,7 @@
 
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
 
   save_old_LDFLAGS=$LDFLAGS
   ac_stuff="
@@ -49952,13 +49955,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl" >&5
-$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl.35" >&5
+$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_SSL_CTX_set_ssl_version+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -50014,14 +50017,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        IMAP_SHARED_LIBADD="-lssl $IMAP_SHARED_LIBADD"
+        IMAP_SHARED_LIBADD="-lssl.35 $IMAP_SHARED_LIBADD"
     else
 
 
   case ssl in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lssl $LIBS"
+      LIBS="-lssl.35 $LIBS"
    ;;
   esac
 
@@ -50037,14 +50040,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        IMAP_SHARED_LIBADD="-lcrypto $IMAP_SHARED_LIBADD"
+        IMAP_SHARED_LIBADD="-lcrypto.35 $IMAP_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -103364,13 +103367,13 @@
   test -z "$PHP_OPENSSL" && PHP_OPENSSL=no
 
   if test "$PHP_OPENSSL" != "no" || test "$PHP_OPENSSL_DIR" != "no"; then
-    { $as_echo "$as_me:${as_lineno-$LINENO}: checking for DSA_get_default_method in -lssl" >&5
-$as_echo_n "checking for DSA_get_default_method in -lssl... " >&6; }
+    { $as_echo "$as_me:${as_lineno-$LINENO}: checking for DSA_get_default_method in -lssl.35" >&5
+$as_echo_n "checking for DSA_get_default_method in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_DSA_get_default_method+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -103406,13 +103409,13 @@
 
 fi
 
-    { $as_echo "$as_me:${as_lineno-$LINENO}: checking for X509_free in -lcrypto" >&5
-$as_echo_n "checking for X509_free in -lcrypto... " >&6; }
+    { $as_echo "$as_me:${as_lineno-$LINENO}: checking for X509_free in -lcrypto.35" >&5
+$as_echo_n "checking for X509_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_X509_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -103655,7 +103658,8 @@
     if test "$found_openssl" = "no"; then
 
     if test "$PHP_OPENSSL_DIR" = "yes"; then
-      PHP_OPENSSL_DIR="/usr/local/ssl /usr/local /usr /usr/local/openssl"
+      OPENSSL_LIBDIR="/usr/lib"
+      PHP_OPENSSL_DIR="/usr/local/libressl /usr/local/ssl /usr/local /usr /usr/local/openssl"
     fi
 
     for i in $PHP_OPENSSL_DIR; do
@@ -103833,13 +103837,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto" >&5
-$as_echo_n "checking for CRYPTO_free in -lcrypto... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CRYPTO_free in -lcrypto.35" >&5
+$as_echo_n "checking for CRYPTO_free in -lcrypto.35... " >&6; }
 if test "${ac_cv_lib_crypto_CRYPTO_free+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lcrypto  $LIBS"
+LIBS="-lcrypto.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -103880,14 +103884,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        MYSQLND_SHARED_LIBADD="-lcrypto $MYSQLND_SHARED_LIBADD"
+        MYSQLND_SHARED_LIBADD="-lcrypto.35 $MYSQLND_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
@@ -103912,7 +103916,7 @@
 
 
     old_LIBS=$LIBS
-    LIBS="$LIBS -lcrypto"
+    LIBS="$LIBS -lcrypto.35"
 
   save_old_LDFLAGS=$LDFLAGS
   ac_stuff="
@@ -104010,13 +104014,13 @@
     esac
   done
 
-  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl" >&5
-$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl... " >&6; }
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for SSL_CTX_set_ssl_version in -lssl.35" >&5
+$as_echo_n "checking for SSL_CTX_set_ssl_version in -lssl.35... " >&6; }
 if test "${ac_cv_lib_ssl_SSL_CTX_set_ssl_version+set}" = set; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lssl  $LIBS"
+LIBS="-lssl.35  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -104072,14 +104076,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        MYSQLND_SHARED_LIBADD="-lssl $MYSQLND_SHARED_LIBADD"
+        MYSQLND_SHARED_LIBADD="-lssl.35 $MYSQLND_SHARED_LIBADD"
     else
 
 
   case ssl in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lssl $LIBS"
+      LIBS="-lssl.35 $LIBS"
    ;;
   esac
 
@@ -104095,14 +104099,14 @@
   c|c_r|pthread*) ;;
   *)
     if test "$ext_shared" = "yes"; then
-        MYSQLND_SHARED_LIBADD="-lcrypto $MYSQLND_SHARED_LIBADD"
+        MYSQLND_SHARED_LIBADD="-lcrypto.35 $MYSQLND_SHARED_LIBADD"
     else
 
 
   case crypto in
   c|c_r|pthread*) ;;
   *)
-      LIBS="-lcrypto $LIBS"
+      LIBS="-lcrypto.35 $LIBS"
    ;;
   esac
 
diff -ur a/ext/mbstring/oniguruma/regint.h b/ext/mbstring/oniguruma/regint.h
--- a/ext/mbstring/oniguruma/regint.h	2016-06-21 18:50:40.000000000 -0700
+++ b/ext/mbstring/oniguruma/regint.h	2016-07-08 14:31:36.000000000 -0700
@@ -196,6 +196,7 @@
 #endif
 
 #include "regenc.h"
+#include <main/php_config.h>
 
 #ifdef MIN
 #undef MIN

