diff -Naur php-5.4.22/TSRM/build.mk php/TSRM/build.mk
--- php-5.4.22/TSRM/build.mk	1999-10-09 19:01:20.000000000 -0700
+++ php/TSRM/build.mk	2009-01-03 20:35:18.000000000 -0800
@@ -25,7 +25,7 @@
 
 $(LT_TARGETS):
 	rm -f $(LT_TARGETS)
-	libtoolize --automake $(AMFLAGS) -f
+	glibtoolize --automake $(AMFLAGS) -f
 
 $(makefile_in_files): $(makefile_am_files)
 	automake -a -i $(AMFLAGS) $(makefile_files)
diff -Naur php-5.4.22/TSRM/buildconf php/TSRM/buildconf
--- php-5.4.22/TSRM/buildconf	1999-10-05 06:47:16.000000000 -0700
+++ php/TSRM/buildconf	2009-01-03 20:35:18.000000000 -0800
@@ -7,7 +7,7 @@
 ;;
 esac
 
-libtoolize --force --automake $automake_flags
+glibtoolize --force --automake $automake_flags
 
 mv aclocal.m4 aclocal.m4.old 2>/dev/null
 aclocal
--- php/Zend/zend_language_parser.y.orig	2016-12-20 08:05:48.000000000 -0800
+++ php/Zend/zend_language_parser.y	2016-12-20 08:34:07.000000000 -0800
@@ -46,12 +46,6 @@
 %pure_parser
 %expect 0
 
-%code requires {
-}
-
-%destructor { zend_ast_destroy($$); } <ast>
-%destructor { if ($$) zend_string_release_ex($$, 0); } <str>
-
 %left T_INCLUDE T_INCLUDE_ONCE T_EVAL T_REQUIRE T_REQUIRE_ONCE
 %left ','
 %left T_LOGICAL_OR

--- php/ext/json/json_parser.y.orig	2016-12-21 17:02:09.000000000 -0800
+++ php/ext/json/json_parser.y	2016-12-21 17:17:15.000000000 -0800
@@ -1,4 +1,4 @@
-%code top {
+%{
 /*
   +----------------------------------------------------------------------+
   | PHP Version 7                                                        |
@@ -36,10 +36,10 @@
 #define PHP_JSON_USE_1(uvr, uv1) PHP_JSON_USE(uvr); PHP_JSON_USE(uv1)
 #define PHP_JSON_USE_2(uvr, uv1, uv2) PHP_JSON_USE(uvr); PHP_JSON_USE(uv1); PHP_JSON_USE(uv2)
 
-}
+%}
 
 %pure-parser
-%name-prefix "php_json_yy"
+%name-prefix = "php_json_yy"
 %lex-param  { php_json_parser *parser  }
 %parse-param { php_json_parser *parser }
 
@@ -74,14 +74,11 @@
 %type <value> members member elements element
 %type <pair> pair
 
-%destructor { zval_ptr_dtor_nogc(&$$); } <value>
-%destructor { zend_string_release_ex($$.key, 0); zval_ptr_dtor_nogc(&$$.val); } <pair>
-
-%code {
+%{
 static int php_json_yylex(union YYSTYPE *value, php_json_parser *parser);
 static void php_json_yyerror(php_json_parser *parser, char const *msg);
 
-}
+%}
 
 %% /* Rules */
 
diff -Naur php-5.4.22/Zend/build.mk php/Zend/build.mk
--- php-5.4.22/Zend/build.mk	1999-10-09 19:02:13.000000000 -0700
+++ php/Zend/build.mk	2009-01-03 20:35:18.000000000 -0800
@@ -25,7 +25,7 @@
 
 $(LT_TARGETS):
 	rm -f $(LT_TARGETS)
-	libtoolize --automake $(AMFLAGS) -f
+	glibtoolize --automake $(AMFLAGS) -f
 
 $(makefile_in_files): $(makefile_am_files)
 	automake -a -i $(AMFLAGS) $(makefile_files)
diff -Naur php-5.4.22/Zend/buildconf php/Zend/buildconf
--- php-5.4.22/Zend/buildconf	1999-10-04 08:20:12.000000000 -0700
+++ php/Zend/buildconf	2009-01-03 20:35:18.000000000 -0800
@@ -7,7 +7,7 @@
 ;;
 esac
 
-libtoolize --force --automake $automake_flags
+glibtoolize --force --automake $automake_flags
 
 mv aclocal.m4 aclocal.m4.old 2>/dev/null
 aclocal
diff -Naur php-5.4.22/ext/mbstring/libmbfl/config.h php/ext/mbstring/libmbfl/config.h
--- php-5.4.22/ext/mbstring/libmbfl/config.h	1969-12-31 16:00:00.000000000 -0800
+++ php/ext/mbstring/libmbfl/config.h	2009-01-03 20:35:19.000000000 -0800
@@ -0,0 +1 @@
+#include "php_config.h"
diff -Naur php/configure.orig php/configure
--- php/configure.orig	2014-01-16 15:42:07.000000000 -0800
+++ php/configure	2014-01-16 15:37:54.000000000 -0800
@@ -8350,10 +8350,9 @@
 if ac_fn_c_try_compile "$LINENO"; then :
 
 
-$as_echo "#define HAVE_PORT 1" >>confdefs.h
 
-        { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-$as_echo "yes" >&6; }
+        { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes but overridden to no by patch" >&5
+$as_echo "no" >&6; }
 
 else
 
@@ -39759,7 +39759,7 @@
 
 $as_echo "#define ENABLE_GD_TTF 1" >>confdefs.h
 
-    GDLIB_CFLAGS="$GDLIB_CFLAGS -DHAVE_LIBFREETYPE -DENABLE_GD_TTF"
+    GDLIB_CFLAGS="$GDLIB_CFLAGS -DENABLE_GD_TTF"
   fi
 
   if test -n "$USE_GD_JIS_CONV"; then
