--- php-5.3.6/php.ini-development	2011-02-08 16:25:44.000000000 -0800
+++ php/php.ini-development	2011-03-18 14:07:34.000000000 -0700
@@ -1061,7 +1061,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1210,12 +1210,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1266,7 +1266,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
--- php-5.3.6/php.ini-production	2011-02-08 16:25:44.000000000 -0800
+++ php/php.ini-production	2011-03-18 14:07:34.000000000 -0700
@@ -1060,7 +1060,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1209,12 +1209,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1273,7 +1273,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
