--- php-5.3.4/php.ini-development	2010-12-07 15:53:51.000000000 -0800
+++ php/php.ini-development	2010-12-10 14:41:17.000000000 -0800
@@ -1057,7 +1057,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1206,12 +1206,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1262,7 +1262,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
--- php-5.3.4/php.ini-production	2010-12-07 15:53:51.000000000 -0800
+++ php/php.ini-production	2010-12-10 14:41:17.000000000 -0800
@@ -1056,7 +1056,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1205,12 +1205,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1269,7 +1269,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
