--- php-5.3.1RC3/php.ini-development	2009-10-28 04:03:07.000000000 -0700
+++ php/php.ini-development	2009-11-09 13:37:18.000000000 -0700
@@ -1060,7 +1060,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1208,12 +1208,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1264,7 +1264,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
--- php-5.3.1RC3/php.ini-production	2009-10-28 04:03:07.000000000 -0700
+++ php/php.ini-production	2009-11-09 13:37:18.000000000 -0700
@@ -1060,7 +1060,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/pdo_mysql.default-socket
-pdo_mysql.default_socket=
+pdo_mysql.default_socket=/var/mysql/mysql.sock
 
 [Phar]
 ; http://php.net/phar.readonly
@@ -1208,12 +1208,12 @@
 ; compile-time value defined MYSQL_PORT (in that order).  Win32 will only look
 ; at MYSQL_PORT.
 ; http://php.net/mysql.default-port
-mysql.default_port =
+mysql.default_port = 3306
 
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysql.default-socket
-mysql.default_socket =
+mysql.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysql.default-host
@@ -1272,7 +1272,7 @@
 ; Default socket name for local MySQL connects.  If empty, uses the built-in
 ; MySQL defaults.
 ; http://php.net/mysqli.default-socket
-mysqli.default_socket =
+mysqli.default_socket = /var/mysql/mysql.sock
 
 ; Default host for mysql_connect() (doesn't apply in safe mode).
 ; http://php.net/mysqli.default-host
