diff -Naur configure.orig configure
--- configure.orig      2007-05-06 08:05:26.000000000 -0700
+++ configure   2007-05-06 08:06:05.000000000 -0700
@@ -3679,10 +3679,6 @@
         CPPFLAGS="$CPPFLAGS -no-cpp-precomp"
       fi
     fi
-    cat >> confdefs.h <<\EOF
-#define BIND_8_COMPAT 1
-EOF
-
     php_multiple_shlib_versions_ok=yes
     ;;
   *beos*)
diff -Naur /tmp/apache_mod_php/php-4.3.8/ext/standard/dns.c ./ext/standard/dns.c
--- /tmp/apache_mod_php/php-4.3.8/ext/standard/dns.c    2003-09-01 15:37:46.000000000 -0700
+++ ./ext/standard/dns.c        2004-12-14 12:37:22.000000000 -0800
@@ -55,6 +55,9 @@
 #undef STATUS
 #undef T_UNSPEC
 #endif
+#if !defined BIND_8_COMPAT
+#include <arpa/nameser_compat.h>
+#endif
 #if HAVE_ARPA_NAMESER_H
 #include <arpa/nameser.h>
 #endif
