--- Completion/Unix/Type/_groups	2005-06-15 04:04:51.000000000 -0700
+++ Completion/Unix/Type/_groups	2007-05-07 19:56:07.000000000 -0700
@@ -7,7 +7,7 @@
 if ! zstyle -a ":completion:${curcontext}:" groups groups; then
   (( $+_cache_groups )) ||
       if [[ $OSTYPE = darwin* ]]; then
-        : ${(A)_cache_groups:=${${(M)${(f)"$(_call_program groups lookupd -q group)"}:#name*}##*: }}
+        : ${(A)_cache_groups:=${${(M)${(f)"$(_call_program groups dscacheutil -q group)"}:#name*}##*: }}
       elif (( ${+commands[getent]} )); then
         : ${(A)_cache_groups:=${${(s: :)$(_call_program groups getent group 2>/dev/null)}%%:*}}
       else
