--- configure.orig	2008-08-23 01:54:27.000000000 -0700
+++ configure	2008-08-23 01:54:47.000000000 -0700
@@ -4260,10 +4260,6 @@
 
 fi
 
-case "$host_os" in
-  darwin*) CPP="$CPP -traditional-cpp" ;;
-esac
-
 { echo "$as_me:$LINENO: checking for ${CC-cc} option to accept ANSI C" >&5
 echo $ECHO_N "checking for ${CC-cc} option to accept ANSI C... $ECHO_C" >&6; }
 if test "${fp_cv_prog_cc_stdc+set}" = set; then
