--- Src/builtin.c.orig	2015-10-05 12:29:56.000000000 -0700
+++ Src/builtin.c	2015-10-05 12:57:15.000000000 -0700
@@ -3009,13 +3009,12 @@
     }
 
     if ((off & PM_UNDEFINED) || (OPT_ISSET(ops,'k') && OPT_ISSET(ops,'z')) ||
-	(OPT_ISSET(ops,'x') && !OPT_HASARG(ops,'x')) ||
 	(OPT_MINUS(ops,'X') && (OPT_ISSET(ops,'m') || *argv || !scriptname))) {
 	zwarnnam(name, "invalid option(s)");
 	return 1;
     }
 
-    if (OPT_ISSET(ops,'x')) {
+    if (OPT_ISSET(ops,'x') && OPT_HASARG(ops,'x')) {
 	char *eptr;
 	expand = (int)zstrtol(OPT_ARG(ops,'x'), &eptr, 10);
 	if (*eptr) {
