--- Doc/Makefile.in	2006-12-17 08:02:02.000000000 -0800
+++ Doc/Makefile.in	2007-05-17 14:05:21.000000000 -0700
@@ -45,10 +45,9 @@
 .SUFFIXES: .yo .1
 
 # man pages to install
-MAN = zsh.1 zshbuiltins.1 zshcalsys.1 zshcompctl.1 zshcompwid.1 zshcompsys.1 \
+MAN = zsh.1 zshbuiltins.1 zshcompctl.1 zshcompwid.1 zshcompsys.1 \
 zshcontrib.1 zshexpn.1 zshmisc.1 zshmodules.1 \
-zshoptions.1 zshparam.1 zshroadmap.1 zshtcpsys.1 zshzftpsys.1 zshzle.1 \
-zshall.1
+zshoptions.1 zshparam.1 zshroadmap.1 zshtcpsys.1 zshzftpsys.1 zshzle.1
 
 # yodl documentation
 
@@ -70,7 +69,6 @@
 Zsh/mod_zutil.yo
 
 YODLSRC = zmacros.yo zman.yo ztexi.yo Zsh/arith.yo Zsh/builtins.yo \
-Zsh/calsys.yo \
 Zsh/compat.yo Zsh/compctl.yo Zsh/compsys.yo Zsh/compwid.yo Zsh/cond.yo \
 Zsh/contrib.yo Zsh/exec.yo Zsh/expn.yo \
 Zsh/filelist.yo Zsh/files.yo Zsh/func.yo Zsh/grammar.yo Zsh/manual.yo \
@@ -180,8 +178,6 @@
 
 zshbuiltins.1: Zsh/builtins.yo
 
-zshcalsys.1: Zsh/calsys.yo
-
 zshcompctl.1: Zsh/compctl.yo
 
 zshcompwid.1: Zsh/compwid.yo
--- Doc/zsh.1	2007-04-19 02:35:57.000000000 -0700
+++ Doc/zsh.1	2007-05-17 14:05:01.000000000 -0700
@@ -32,15 +32,11 @@
 .TP
 \fIzshmodules\fP   Zsh loadable modules
 .TP
-\fIzshcalsys\fP    Zsh built\-in calendar functions
-.TP
 \fIzshtcpsys\fP    Zsh built\-in TCP functions
 .TP
 \fIzshzftpsys\fP   Zsh built\-in FTP client
 .TP
 \fIzshcontrib\fP   Additional zsh functions and utilities
-.TP
-\fIzshall\fP       Meta\-man page containing all of the above
 .PD
 .SH "DESCRIPTION"
 Zsh is a UNIX command interpreter (shell) usable as an interactive
