--- mu-conference_0.8.orig/src/conference.c	2009-09-30 12:06:03.000000000 -0700
+++ mu-conference_0.8/src/conference.c	2009-09-30 12:11:49.000000000 -0700
@@ -24,8 +24,8 @@
 {
   cnr room = (cnr)data;
   jpacket jp = (jpacket)arg;
-  char users[10];
-  char maxu[10];
+  char users[16];
+  char maxu[16];
   xmlnode x;
 
   if(room == NULL)
@@ -68,8 +68,8 @@
 {
   cnr room = (cnr)data;
   jpacket jp = (jpacket)arg;
-  char users[10];
-  char maxu[10];
+  char users[16];
+  char maxu[16];
   xmlnode x;
 
   if(room == NULL)
