--- configure.orig	2005-10-05 17:33:16.000000000 -0700
+++ configure	2005-10-05 17:33:43.000000000 -0700
@@ -6145,7 +6145,7 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lmd  $LIBS"
+LIBS="-lmd -lcrypto  $LIBS"
 cat >conftest.$ac_ext <<_ACEOF
 #line $LINENO "configure"
 /* confdefs.h.  */
@@ -6198,7 +6198,7 @@
 #define HAVE_LIBMD 1
 _ACEOF
 
-  LIBS="-lmd $LIBS"
+  LIBS="-lmd -lcrypto $LIBS"
 
 fi
 
