diff --exclude=cscope.out --exclude='.*' -Naur ../ntp-4.2.0/ntpdate/ntpdate.c ./ntpdate/ntpdate.c
--- ../ntp-4.2.0/ntpdate/ntpdate.c	2005-09-09 13:08:25.000000000 -0700
+++ ./ntpdate/ntpdate.c	2005-09-09 14:34:03.000000000 -0700
@@ -209,8 +209,8 @@
 static	void	server_data P((struct server *, s_fp, l_fp *, u_fp));
 static	void	clock_filter	P((struct server *));
 static	struct server *clock_select P((void));
+void	addserver	P((char *));
 static	int clock_adjust	P((void));
-static	void	addserver	P((char *));
 static	struct server *findserver P((struct sockaddr_storage *));
 		void	timer		P((void));
 static	void	init_alarm	P((void));
@@ -1336,7 +1336,7 @@
  * addserver - determine a server's address and allocate a new structure
  *		   for it.
  */
-static void
+void
 addserver(
 	char *serv
 	)
diff --exclude=cscope.out --exclude='.*' -Naur ../ntp-4.2.0/ntpdate/ntptime_config.c ./ntpdate/ntptime_config.c
--- ../ntp-4.2.0/ntpdate/ntptime_config.c	2005-09-09 13:08:25.000000000 -0700
+++ ./ntpdate/ntptime_config.c	2005-09-09 14:52:31.000000000 -0700
@@ -135,6 +135,7 @@
 			   int complain));
 
 
+extern void	addserver	P((char *));
 /*
  * loadservers - load list of NTP servers from configuration file
  */
@@ -236,7 +237,9 @@
 					tokens[0]);
 				break;
 			}
-			
+#ifdef __APPLE__
+			addserver(tokens[1]);
+#else			
 			if (!getnetnum(tokens[1], &peeraddr, 1)) {
 				/* Resolve now, or lose! */
 				break;
@@ -368,7 +371,7 @@
 				srvcnt++;
 			}
 			break;
-			
+#endif			
 			case CONFIG_KEYS:
 			if (ntokens >= 2) {
 				key_file = (char *) emalloc(strlen(tokens[1]) + 1);
@@ -380,6 +383,7 @@
 	(void) fclose(fp);
 
 	/* build final list */
+#ifndef __APPLE__
 	sys_numservers = srvcnt;
 	sys_servers = (struct server **) 
 	    emalloc(sys_numservers * sizeof(struct server *));
@@ -388,6 +392,7 @@
 		sys_servers[i]->event_time = i+1;
 		srvlist = srvlist->next_server;
 	}
+#endif
 }
 
 
