--- ntpd/ntp_io.c+orig	2008-04-30 19:19:08.000000000 -0700
+++ ntpd/ntp_io.c	2008-04-30 19:23:16.000000000 -0700
@@ -2472,6 +2472,7 @@ sendpkt(
 	)
 {
 	int cc, slot;
+	static int last_errno;
 #ifdef SYS_WINNT
 	DWORD err;
 #endif /* SYS_WINNT */
@@ -2646,13 +2647,16 @@ sendpkt(
 			default:  /* don't care if not supported */
 				break;
 			}
-
-			netsyslog(LOG_ERR, "sendto(%s) (fd=%d): %m",
-				  stoa(dest), inter->fd);
+			if (last_errno != errno) {
+				last_errno = errno;
+				netsyslog(LOG_ERR, "sendto(%s) (fd=%d): %m",
+					  stoa(dest), inter->fd);
+			}
 		}
 	}
 	else
 	{
+		last_errno = 0;
 		inter->sent++;
 		packets_sent++;
 		/*
