--- ntpd/ntp_io.c~	2008-07-08 11:02:25.000000000 -0700
+++ ntpd/ntp_io.c	2008-07-08 18:00:30.000000000 -0700
@@ -70,8 +70,8 @@ extern const char *specific_interface;
 
 #if defined(SO_TIMESTAMP) && defined(SCM_TIMESTAMP)
 #if defined(CMSG_FIRSTHDR)
-#define HAVE_TIMESTAMP
-#define USE_TIMESTAMP_CMSG
+/*#define HAVE_TIMESTAMP*/
+/*#define USE_TIMESTAMP_CMSG*/
 #ifndef TIMESTAMP_CTLMSGBUF_SIZE
 #define TIMESTAMP_CTLMSGBUF_SIZE 1536 /* moderate default */
 #endif
