diff -up -r /SourceCache/ntp/ntp-44/ntp/ntpd/ntpd.c ./ntp/ntpd/ntpd.c
--- /SourceCache/ntp/ntp-44/ntp/ntpd/ntpd.c	2007-09-10 16:55:16.000000000 -0700
+++ ntpd/ntpd.c	2008-09-16 16:25:01.000000000 -0700
@@ -252,9 +252,6 @@ init_logging(char const *name)
 		setlogmask(LOG_UPTO(LOG_DEBUG)); /* @@@ was INFO */
 # endif /* LOG_DAEMON */
 #endif	/* !SYS_WINNT && !VMS */
-
-	NLOG(NLOG_SYSINFO) /* conditional if clause for conditional syslog */
-		msyslog(LOG_NOTICE, "%s", Version);
 }
 
 
@@ -844,6 +841,8 @@ ntpdmain(
 	 */
 
 	getconfig(argc, argv);
+	NLOG(NLOG_SYSINFO) /* conditional if clause for conditional syslog */
+		msyslog(LOG_NOTICE, "%s", Version);
 
 	loop_config(LOOP_DRIFTCOMP, old_drift / 1e6);
 #ifdef OPENSSL
