--- gzexe.in.orig	2006-02-03 12:20:56.000000000 -0800
+++ gzexe.in	2006-02-03 12:21:54.000000000 -0800
@@ -121,7 +121,7 @@
 if tail +$skip "$0" | "BINDIR"/gzip -cd > /tmp/gztmp$$; then
   umask $umask
   /bin/chmod 700 /tmp/gztmp$$
-  prog="`echo $0 | /bin/sed 's|^.*/||'`"
+  prog="`echo $0 | /usr/bin/sed 's|^.*/||'`"
   if /bin/ln /tmp/gztmp$$ "/tmp/$prog" 2>/dev/null; then
     trap '/bin/rm -f /tmp/gztmp$$ "/tmp/$prog"; exit $res' 0
     (/bin/sleep 5; /bin/rm -f /tmp/gztmp$$ "/tmp/$prog") 2>/dev/null &
