--- Makefile.in.orig	2012-01-04 15:45:35.000000000 -0800
+++ Makefile.in	2012-01-04 15:46:36.000000000 -0800
@@ -269,7 +269,7 @@
 ACLOCAL_AMFLAGS = -I m4
 AM_CPPFLAGS = -I$(top_srcdir)/lib
 man_MANS = gunzip.1 gzexe.1 gzip.1 \
-  zcat.1 zcmp.1 zdiff.1 zforce.1 zgrep.1 zless.1 zmore.1 znew.1
+  zcat.1 zcmp.1 zdiff.1 zforce.1 zless.1 zmore.1 znew.1
 
 EXTRA_DIST = $(ACINCLUDE_INPUTS) $(man_MANS) \
   algorithm.doc amiga/match.a amiga/Makefile.gcc amiga/Makefile.sasc \
@@ -287,7 +287,7 @@
   zegrep.in zfgrep.in zforce.in zgrep.in zless.in zmore.in znew.in
 
 bin_SCRIPTS = gunzip gzexe zcat zcmp zdiff \
-  zegrep zfgrep zforce zgrep zless zmore znew
+  zforce zless zmore znew
 
 gzip_SOURCES = \
   bits.c crypt.c deflate.c gzip.c inflate.c lzw.c \
