--- asctime.c.bsdnew	2009-11-14 13:55:43.000000000 -0800
+++ asctime.c	2009-11-14 14:04:38.000000000 -0800
@@ -79,9 +79,7 @@ static char	buf_asctime[MAX_ASCTIME_BUF_
 */
 
 char *
-asctime_r(timeptr, buf)
-const struct tm *	timeptr;
-char *			buf;
+asctime_r(const struct tm * __restrict timeptr, char * __restrict buf)
 {
 	static const char	wday_name[][3] = {
 		"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
