--- Makefile.in.orig	2004-08-11 15:12:37.000000000 -0700
+++ Makefile.in	2004-08-11 15:12:48.000000000 -0700
@@ -313,7 +313,7 @@
 $($(PACKAGE)stub_LIB_FILE): $($(PACKAGE)stub_OBJECTS)
 	-rm -f $($(PACKAGE)stub_LIB_FILE)
 	${MAKE_STUB_LIB}
-	$(RANLIB) $($(PACKAGE)stub_LIB_FILE)
+	ranlib $($(PACKAGE)stub_LIB_FILE)
 
 #========================================================================
 # We need to enumerate the list of .c to .o lines here.
