--- base5067/innobase/buf/buf0buf.c	2008-08-04 05:19:12.000000000 -0700
+++ m5067/innobase/buf/buf0buf.c	2008-09-08 06:53:12.000000000 -0700
@@ -1226,7 +1226,7 @@
 
 	if (mode == BUF_GET_NOWAIT) {
 		if (rw_latch == RW_S_LATCH) {
-			success = rw_lock_s_lock_func_nowait(&(block->lock),
+			success = rw_lock_s_lock_nowait(&(block->lock),
 								file, line);
 			fix_type = MTR_MEMO_PAGE_S_FIX;
 		} else {
@@ -1359,8 +1359,7 @@
 	ut_ad(!ibuf_inside() || ibuf_page(block->space, block->offset));
 
 	if (rw_latch == RW_S_LATCH) {
-		success = rw_lock_s_lock_func_nowait(&(block->lock),
-								file, line);
+		success = rw_lock_s_lock_nowait(&(block->lock),	file, line);
 		fix_type = MTR_MEMO_PAGE_S_FIX;
 	} else {
 		success = rw_lock_x_lock_func_nowait(&(block->lock),
@@ -1507,8 +1506,7 @@
 	ut_ad(!ibuf_inside() || (mode == BUF_KEEP_OLD));
 
 	if (rw_latch == RW_S_LATCH) {
-		success = rw_lock_s_lock_func_nowait(&(block->lock),
-								file, line);
+		success = rw_lock_s_lock_nowait(&(block->lock), file, line);
 		fix_type = MTR_MEMO_PAGE_S_FIX;
 	} else {
 		success = rw_lock_x_lock_func_nowait(&(block->lock),
