--- DateTime/Build.PL	2010-04-01 07:01:15.000000000 -0500
+++ DateTime/Build.PL	2010-04-01 07:01:25.000000000 -0500
@@ -5,6 +5,14 @@
 
 use Module::Build;
 
+my @compiler_flags;
+my @linker_flags;
+if ($ENV{RC_CFLAGS}) {
+    my @extra_cflags = split " ", $ENV{RC_CFLAGS};
+    push @compiler_flags, @extra_cflags;
+    push @linker_flags, @extra_cflags;
+}
+
 my $force_pp = grep { $_ eq '--pp' } @ARGV;
 
 my @pp = _make_pp_tests();
@@ -36,6 +44,8 @@
                 'https://perl-date-time.svn.sourceforge.net/svnroot/perl-date-time',
         },
     },
+   extra_compiler_flags => \@compiler_flags,
+   extra_linker_flags => \@linker_flags,
 );
 
 if ($force_pp) {
