--- DBD-SQLite/Makefile.PL	2004-10-12 15:17:59.000000000 -0700
+++ /tmp/Makefile.PL	2004-10-14 14:29:07.000000000 -0700
@@ -29,7 +29,7 @@
 corruption.
 
 EOT
-        if (prompt("Continue?", "N") !~ /^y/i) {
+        if (prompt("Continue?", "Y") !~ /^y/i) {
             print "Exiting\n";
             exit -1;
         }
--- DBD-SQLite/dbdimp.c 2004-11-05 15:43:36.000000000 -0800
+++ /tmp/dbdimp.c       2004-11-21 15:01:02.000000000 -0800
@@ -328,7 +328,8 @@
         int sql_type = SvIV(sql_type_sv);
 
         sqlite_trace(4, "params left in 0x%p: %d", imp_sth->params, 1+av_len(imp_sth->params));
-        sqlite_trace(4, "bind %d type %d as %s", i, sql_type, SvPV_nolen(value));
+        sqlite_trace(4, "bind %d type %d as %s", i, sql_type, 
+                     SvOK(value) ? SvPV_nolen(value) : "NULL");
         
         if (!SvOK(value)) {
             sqlite_trace(5, "binding null");
@@ -431,7 +432,8 @@
     }
     pos = 2 * (SvIV(param) - 1);
     sqlite_trace(3, "bind into 0x%p: %d => %s (%d) pos %d\n",
-      imp_sth->params, SvIV(param), SvPV_nolen(value), sql_type, pos);
+      imp_sth->params, SvIV(param), 
+      SvOK(value) ? SvPV_nolen(value) : "NULL", sql_type, pos);
     av_store(imp_sth->params, pos, SvREFCNT_inc(value));
     av_store(imp_sth->params, pos+1, newSViv(sql_type));
     
