*** Lingua-EN-Inflect/lib/Lingua/EN/Inflect.pm.orig   Mon Jun  4 17:38:33 2007
--- Lingua-EN-Inflect/lib/Lingua/EN/Inflect.pm        Mon Jun  4 17:38:47 2007
***************
*** 692,698 ****
  local $SIG{__WARN__} = sub {0};
  my $rcfile;
  
! $rcfile = $INC{'Lingua//EN/Inflect.pm'} || '';
  $rcfile =~ s/Inflect.pm$/.inflectrc/;
  do $rcfile or die "\nBad .inflectrc file ($rcfile):\n\t$@\n"
  if $rcfile && -r $rcfile && -s $rcfile;
--- 692,698 ----
  local $SIG{__WARN__} = sub {0};
  my $rcfile;
  
! $rcfile = $INC{'Lingua/EN/Inflect.pm'} || '';
  $rcfile =~ s/Inflect.pm$/.inflectrc/;
  do $rcfile or die "\nBad .inflectrc file ($rcfile):\n\t$@\n"
  if $rcfile && -r $rcfile && -s $rcfile;
