--- Net-Daemon/lib/Net/Daemon/Log.pm	2009-08-21 12:45:39.000000000 -0700
+++ Net-Daemon/lib/Net/Daemon/Log.pm.hacked	2009-08-21 12:36:56.000000000 -0700
@@ -59,7 +59,8 @@
 	}
 	if ($^O ne 'solaris'  &&  $^O ne 'freebsd'  &&
 	    defined(&Sys::Syslog::setlogsock)  &&
-	    eval { &Sys::Syslog::_PATH_LOG() }) {
+        eval { &Sys::Syslog::_PATH_LOG() } && 
+        ( -S &Sys::Syslog::_PATH_LOG() ) ) {
 	    &Sys::Syslog::setlogsock('unix');
 	}
 	&Sys::Syslog::openlog($self->{'logname'} || ref($self), 'pid',
