--- JSON-Syck/syck.h	2006-02-14 18:58:33.000000000 -0600
+++ /tmp/syck.h	2006-03-20 09:29:26.000000000 -0600
@@ -16,9 +16,17 @@
 #define SYCK_VERSION    "0.61"
 #define YAML_DOMAIN     "yaml.org,2002"
 
-#ifdef HAVE_STDLIB_H
+/* #######################################################################
+ * Paul Steinkamp  3/2006 
+ * The flag HAVE_STDLIB_H is not defined on Mac OS, even though 
+ * the library is available. Commented out the ifdef
+ * 		#ifdef HAVE_STDLIB_H
+ * 		# include <stdlib.h>
+ * 		#endif
+ * Since the code will not build without this library, always include it.
+ * ####################################################################### */
+
 # include <stdlib.h>
-#endif
 
 #ifdef HAVE_STRING_H
 # include <string.h>
