--- Makefile.bak        2007-02-13 18:47:15.000000000 -0800
+++ Makefile    2007-02-13 19:00:42.000000000 -0800
@@ -113,7 +113,7 @@
 INSTALL_STRIP_PROGRAM = ${SHELL} $(install_sh) -c -s
 LDFLAGS = -L/tmp/ChatServer.roots/ChatServer~obj/jabber_od_auth.build/lib -L/usr/lib -L/usr/lib/mysql -L/tmp/ChatServer.roots/ChatServer~obj/libidn.build/lib 
 LIBOBJS =  lstat$U.o
-LIBS = -lsqlite3 -lssl -lcrypto -lsasl2 -lidn 
+LIBS = -lsqlite3 -lssl -lcrypto -lsasl2 -lidn -lresolv
 LIBTOOL = $(SHELL) $(top_builddir)/libtool
 LN_S = ln -s
 LTLIBOBJS =  lstat$U.lo

