--- a/lib/ExtUtils/Embed.pm	2014-01-06 14:46:45.000000000 -0800
+++ b/lib/ExtUtils/Embed.pm	2018-05-11 00:10:36.000000000 -0700
@@ -266,7 +266,7 @@ sub ccdlflags {
 sub perl_inc {
     my $dir = File::Spec->catdir($Config{archlibexp}, 'CORE');
     $dir = qq["$dir"] if $^O eq 'MSWin32';
-    my_return(" -I$dir ");
+    my_return(" -iwithsysroot $dir ");
 }
 
 sub ccopts {
--- perl-5.28.1/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Unix.pm	2018-11-01 15:07:28.000000000 -0700
+++ perl-5.28.1-new/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Unix.pm	2019-03-26 16:49:55.000000000 -0700
@@ -461,10 +461,17 @@
 MAN3PODS = ".$self->wraplist(sort keys %{$self->{MAN3PODS}})."
 ";
 
+if ($self->{'PERL_INC'} =~ /^\/System\/Library\/Perl\//) {
+    push @m, q{
+SDKROOT := $(shell xcrun --show-sdk-path)
+PERL_SYSROOT = $(SDKROOT)
+};
+}
+
 
     push @m, q{
 # Where is the Config information that we are using/depend on
-CONFIGDEP = $(PERL_ARCHLIBDEP)$(DFSEP)Config.pm $(PERL_INCDEP)$(DFSEP)config.h
+CONFIGDEP = $(PERL_ARCHLIBDEP)$(DFSEP)Config.pm $(PERL_SYSROOT)$(PERL_INCDEP)$(DFSEP)config.h
 } if -e $self->catfile( $self->{PERL_INC}, 'config.h' );
 
 
@@ -1034,9 +1041,9 @@
         # platforms.  We peek at lddlflags to see if we need -Wl,-R
         # or -R to add paths to the run-time library search path.
         if ($Config{'lddlflags'} =~ /-Wl,-R/) {
-            $libs .= ' "-L$(PERL_INC)" "-Wl,-R$(INSTALLARCHLIB)/CORE" "-Wl,-R$(PERL_ARCHLIB)/CORE" -lperl';
+            $libs .= ' "-L$(PERL_SYSROOT)$(PERL_INC)" "-Wl,-R$(INSTALLARCHLIB)/CORE" "-Wl,-R$(PERL_ARCHLIB)/CORE" -lperl';
         } elsif ($Config{'lddlflags'} =~ /-R/) {
-            $libs .= ' "-L$(PERL_INC)" "-R$(INSTALLARCHLIB)/CORE" "-R$(PERL_ARCHLIB)/CORE" -lperl';
+            $libs .= ' "-L$(PERL_SYSROOT)$(PERL_INC)" "-R$(INSTALLARCHLIB)/CORE" "-R$(PERL_ARCHLIB)/CORE" -lperl';
         } elsif ( $Is{Android} ) {
             # The Android linker will not recognize symbols from
             # libperl unless the module explicitly depends on it.
@@ -3963,7 +3970,7 @@
 .xs$(OBJ_EXT) :
 	$(XSUBPPRUN) $(XSPROTOARG) $(XSUBPPARGS) $*.xs > $*.xsc
 	$(MV) $*.xsc $*.c
-	$(CCCMD) $(CCCDLFLAGS) "-I$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.c %s
+	$(CCCMD) $(CCCDLFLAGS) -iwithsysroot "-I$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.c %s
 EOF
     if ($self->{XSMULTI}) {
 	for my $ext ($self->_xs_list_basenames) {
--- a/dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder/Base.pm	2014-01-06 14:46:44.000000000 -0800
+++ b/dist/ExtUtils-CBuilder/lib/ExtUtils/CBuilder/Base.pm	2018-07-27 22:45:32.000000000 -0700
@@ -111,7 +111,12 @@ sub object_file {
 
 sub arg_include_dirs {
   my $self = shift;
-  return map {"-I$_"} @_;
+  my @args;
+  foreach (@_) {
+     push (@args, "-iwithsysroot");
+     push (@args, $_);
+  }
+  return @args;
 }
 
 sub arg_nolink { '-c' }
