--- python.man.orig	2005-03-20 06:16:03.000000000 -0800
+++ python.man	2007-02-22 23:05:24.000000000 -0800
@@ -374,6 +374,49 @@
 If this is set to a non-empty string it is equivalent to specifying
 the \fB\-v\fP option. If set to an integer, it is equivalent to
 specifying \fB\-v\fP multiple times. 
+.SH INTERACTIVE INPUT EDITING AND HISTORY SUBSTITUTION
+The Python inteterpreter supports editing of the current input line and
+history substitution, similar to facilities found in the Korn shell and
+the GNU Bash shell.
+However, rather than being implemented using the GNU Readline library, this
+Python interpreter uses the BSD EditLine library
+.BR editline (3)
+with a GNU Readline emulation layer.
+.PP
+The
+.I readline
+module provides the access to the EditLine library, but there are a few major
+differences compared to a traditional implementation using the Readline library.
+The command language used in the preference files is that of EditLine,
+as described in
+.BR editrc (5)
+and not that used by the Readline library.
+This also means that the
+.IR parse_and_bind ()
+routines uses EditLine commands.
+And the preference file itself is
+.I ~/.editrc
+instead of
+.IR ~/.inputrc .
+.PP
+For example, the
+.I rlcompleter
+module, which defines a completion function for the
+.I readline
+modules, works correctly with the EditLine libraries, but needs to be
+initialized somewhat differently:
+.IP
+import rlcompleter
+.br
+import readline
+.br
+readline.parse_and_bind("bind ^I rl_complete")
+.PP
+For
+.I vi
+mode, one needs:
+.IP
+readline.parse_and_bind("bind -v")
 .SH AUTHOR
 The Python Software Foundation: http://www.python.org/psf
 .SH INTERNET RESOURCES
