diff -udbNpr ./TwistedXish-0.1.0/twisted/xish/_version.py ./twisted/xish/_version.py
--- ./TwistedXish-0.1.0/twisted/xish/_version.py	1969-12-31 16:00:00.000000000 -0800
+++ ./twisted/xish/_version.py	2006-05-21 17:01:55.000000000 -0700
@@ -0,0 +1,3 @@
+# This is an auto-generated file. Use admin/change-versions to update.
+from twisted.python import versions
+version = versions.Version(__name__[:__name__.rfind('.')], 0, 4, 0)
