--- jabberd	2006-06-14 15:13:51.000000000 -0700
+++ ../../jabberd	1970-01-01 00:00:00.000000000 -0800
@@ -27,7 +27,7 @@
 #-----------------------------------------------------------------------------
 my $Bin = "/usr/bin";
 my $VERSION = "2.0s9";
-my $config_dir = "/usr/local/etc/jabberd";
+my $config_dir = "/private/etc/jabberd";
 my $config = $config_dir."/jabberd.cfg";
 $config = "internal" unless (-e $config);
 my $debug = 0;
@@ -37,7 +37,8 @@
 my %jobs;
 my %fhs;
 my @programs;
-
+my $t1 = 0;
+my $t2 = 0;
