--- export/src/fgetln.c	2006-08-29 21:11:07.000000000 +0200
+++ ../src/fgetln.c	2006-08-29 21:18:07.000000000 +0200
@@ -32,17 +32,13 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-#include "nbtool_config.h"
-
 #if !HAVE_FGETLN
+#include <config.h>
 #include <stdlib.h>
-#ifdef notdef
-/* These headers are required, but included from nbtool_config.h */
 #include <stdio.h>
 #include <unistd.h>
 #include <errno.h>
 #include <string.h>
-#endif
 
 char *
 fgetln(FILE *fp, size_t *len)
