--- export/examples/test.c	2006-08-29 21:11:07.000000000 +0200
+++ ../examples/test.c	2006-08-29 21:18:07.000000000 +0200
@@ -32,20 +32,6 @@
  * SUCH DAMAGE.
  */
 
-#include "config.h"
-#ifndef lint
-__COPYRIGHT("@(#) Copyright (c) 1992, 1993\n\
-	The Regents of the University of California.  All rights reserved.\n");
-#endif /* not lint */
-
-#if !defined(lint) && !defined(SCCSID)
-#if 0
-static char sccsid[] = "@(#)test.c	8.1 (Berkeley) 6/4/93";
-#else
-__RCSID("$NetBSD: test.c,v 1.18 2005/06/01 11:37:52 lukem Exp $");
-#endif
-#endif /* not lint && not SCCSID */
-
 /*
  * test.c: A little test program
  */
@@ -60,6 +46,11 @@
 
 #include "histedit.h"
 
+/* from src/sys/sys/cdefs.h */
+#ifndef __UNCONST
+# define __UNCONST(a) ((void *)(unsigned long)(const void *)(a))
+#endif
+
 static int continuation = 0;
 volatile sig_atomic_t gotsig = 0;
 
