--- ../../../mfpatch/gettext/gettext-tools/examples/Makefile	2006-09-19 15:46:34.000000000 -0700
+++ Makefile	2006-09-19 15:48:37.000000000 -0700
@@ -1439,7 +1439,7 @@
 distdir: $(DISTFILES)
 	$(am__remove_distdir)
 	mkdir $(distdir)
-$(MAKE) distdir1
+	$(MAKE) distdir1
 	$(mkdir_p) $(distdir)/. $(distdir)/../../build-aux $(distdir)/../../gettext-runtime/m4 $(distdir)/../../m4
 	@srcdirstrip=`echo "$(srcdir)" | sed 's|.|.|g'`; \
 	topsrcdirstrip=`echo "$(top_srcdir)" | sed 's|.|.|g'`; \
