--- Makefile.bak	2006-10-18 10:40:26.000000000 -0700
+++ Makefile	2006-10-18 10:40:26.000000000 -0700
@@ -108,7 +108,7 @@
 INSTALL_STRIP_PROGRAM = ${SHELL} $(install_sh) -c -s
 LDFLAGS = -L/tmp/ChatServer.roots/ChatServer~obj/jabber_od_auth.build/lib -L/usr/lib -L/usr/lib/mysql -L/tmp/ChatServer.roots/ChatServer~obj/libidn.build/lib
 LIBOBJS =  lstat$U.o
-LIBS = -lsqlite3 -lssl -lcrypto -lsasl2 -lidn 
+LIBS = -lsqlite3 -lxmppodauth -lssl -lcrypto -lsasl2 -lidn -lresolv
 LIBTOOL = $(SHELL) $(top_builddir)/libtool
 LN_S = ln -s
 LTLIBOBJS =  lstat$U.lo
@@ -182,14 +182,14 @@
 	  case '$(am__configure_deps)' in \
 	    *$$dep*) \
 	      echo ' cd $(srcdir) && $(AUTOMAKE) --gnu '; \
-	      cd $(srcdir) && $(AUTOMAKE) --gnu  \
+	      cd $(srcdir) && echo ' SKIPPING AUTOMAKE --gnu'  \
 		&& exit 0; \
 	      exit 1;; \
 	  esac; \
 	done; \
 	echo ' cd $(top_srcdir) && $(AUTOMAKE) --gnu  Makefile'; \
 	cd $(top_srcdir) && \
-	  $(AUTOMAKE) --gnu  Makefile
+	  echo ' SKIPPING AUTOMAKE --gnu  Makefile'
 .PRECIOUS: Makefile
 Makefile: $(srcdir)/Makefile.in $(top_builddir)/config.status
 	@case '$?' in \
@@ -205,7 +205,7 @@
 	$(SHELL) ./config.status --recheck
 
 $(top_srcdir)/configure:  $(am__configure_deps)
-	cd $(srcdir) && $(AUTOCONF)
+	cd $(srcdir) && echo ' SKIPPING AUTOCONF'
 $(ACLOCAL_M4):  $(am__aclocal_m4_deps)
 	cd $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
 
@@ -219,9 +219,9 @@
 	@rm -f stamp-h1
 	cd $(top_builddir) && $(SHELL) ./config.status config.h
 $(srcdir)/config.h.in:  $(am__configure_deps) 
-	cd $(top_srcdir) && $(AUTOHEADER)
-	rm -f stamp-h1
-	touch $@
+	cd $(top_srcdir) && echo ' SKIPPING AUTOHEADER'
+	echo 'SKIPPING rm -f stamp-h1'
+	echo 'SKIPPING touch $@'
 
 distclean-hdr:
 	-rm -f config.h stamp-h1
