--- command.c.orig	2004-07-28 11:13:04.000000000 -0700
+++ command.c	2004-07-28 11:14:44.000000000 -0700
@@ -38,6 +38,7 @@
 extern int secure;
 extern int hshift;
 extern int show_attn;
+extern int more_mode;
 extern char *every_first_cmd;
 extern char *curr_altfilename;
 extern char version[];
@@ -476,12 +477,16 @@
 		flag = 0;
 		switch (c)
 		{
-		case CONTROL('E'): /* ignore END of file */
 		case '*':
+			if (more_mode)
+				break;
+		case CONTROL('E'): /* ignore END of file */
 			flag = SRCH_PAST_EOF;
 			break;
-		case CONTROL('F'): /* FIRST file */
 		case '@':
+			if (more_mode)
+				break;
+		case CONTROL('F'): /* FIRST file */
 			flag = SRCH_FIRST_FILE;
 			break;
 		case CONTROL('K'): /* KEEP position */
