--- ifile.c.orig	2005-01-06 17:32:18.000000000 -0800
+++ ifile.c	2005-02-17 14:11:54.000000000 -0800
@@ -23,6 +23,7 @@
 #include "less.h"
 
 extern IFILE	curr_ifile;
+extern int	unix2003_compat;
 
 struct ifile {
 	struct ifile *h_next;		/* Links for command line list */
@@ -229,8 +230,13 @@
 {
 	register struct ifile *p;
 
-	if ((p = find_ifile(filename)) == NULL)
+	if (unix2003_compat) {
+		/* don't exclude duplicates in list */
 		p = new_ifile(filename, int_ifile(prev));
+	} else {
+		if ((p = find_ifile(filename)) == NULL)
+			p = new_ifile(filename, int_ifile(prev));
+	}
 	return (ext_ifile(p));
 }
 
