--- optfunc.c.orig	2005-01-06 17:32:19.000000000 -0800
+++ optfunc.c	2005-02-21 10:44:05.000000000 -0800
@@ -45,6 +45,7 @@
 extern char *wproto;
 extern IFILE curr_ifile;
 extern char version[];
+extern char *dashp_commands;
 #if LOGFILE
 extern char *namelogfile;
 extern int force_logfile;
@@ -591,3 +592,13 @@
 	return (sc_height + swindow);
 }
 
+/* Following handler function is used in Unix 2003 override of -p option.  */
+
+	public void
+opt_dashp(type,s)
+	int type;
+	char *s;
+{
+	dashp_commands = s;
+}
+
