--- position.c	2004-02-03 20:52:27.000000000 -0800
+++ /Volumes/dee/Len/Build/less.roots/less/less/position.c	2005-09-27 20:09:54.000000000 -0700
@@ -186,7 +186,7 @@ get_scrpos(scrpos)
 	 * Find the first line on the screen which has something on it,
 	 * and return the screen line number and the file position.
 	 */
-	for (i = 0; i < sc_height;  i++)
+	for (i = 0; i < table_size;  i++)
 		if (table[i] != NULL_POSITION)
 		{
 			scrpos->ln = i+1;
