--- search.c.orig	2004-07-28 11:25:16.000000000 -0700
+++ search.c	2004-07-28 11:25:56.000000000 -0700
@@ -22,7 +22,7 @@
 #if HAVE_POSIX_REGCOMP
 #include <regex.h>
 #ifdef REG_EXTENDED
-#define	REGCOMP_FLAG	REG_EXTENDED
+#define	REGCOMP_FLAG	(more_mode ? 0 : REG_EXTENDED)
 #else
 #define	REGCOMP_FLAG	0
 #endif
@@ -52,6 +52,7 @@
 extern int sc_height;
 extern int jump_sline;
 extern int bs_mode;
+extern int more_mode;
 extern int ctldisp;
 extern int status_col;
 extern POSITION start_attnpos;
