--- signal.c.orig	2005-12-03 12:20:32.000000000 -0800
+++ signal.c	2006-01-25 20:35:41.000000000 -0800
@@ -34,6 +34,7 @@ extern int linenums;
 extern int wscroll;
 extern int reading;
 extern int quit_on_intr;
+extern int more_mode;
 
 /*
  * Interrupt signal handler.
@@ -57,6 +58,8 @@ u_interrupt(type)
 	if (kbhit())
 		getkey();
 #endif
+	if (more_mode)
+		quit(0);
 	if (reading)
 		intread();
 }
